{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>.container { width:100% !important; }</style>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import math\n",
    "from collections import defaultdict\n",
    "import matplotlib.pyplot as plt\n",
    "import pickle\n",
    "from IPython.core.display import HTML\n",
    "HTML(\"<style>.container { width:100% !important; }</style>\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "#Reading the Zinc-finger domain data\n",
    "curr_dir = !pwd\n",
    "my_path = curr_dir[0]+\"/hmm_domains/\"\n",
    "filename = \"zf-C2H2.csv\"\n",
    "zinc_finger = pd.read_csv(my_path+filename, sep='\\t', index_col=0)\n",
    "#Sort the zinc finger data\n",
    "sorted_zinc = zinc_finger.sort_values(by=[\"chrom_num\", \"ensembl_id\", \"TargetStart\"])\n",
    "sorted_zinc = sorted_zinc.reset_index(drop=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "#Get the canonic protein id for Zinc domain\n",
    "with open(my_path+'zinc_canonic_prot.pik', 'rb') as handle:\n",
    "    canonic_protein = pickle.load(handle)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "chromosome_names = [\"1\", \"2\", \"3\", \"4\", \"5\", \"6\", \"7\", \"8\", \"9\", \"10\", \"11\", \"12\", \"13\", \"14\", \"15\", \"16\", \"17\", \"18\", \"19\", \"20\", \"21\", \"22\", \"X\", \"Y\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Leaving outside 74 out of 10492\n"
     ]
    }
   ],
   "source": [
    "chrom_names_list = sorted_zinc[\"chrom_num\"].tolist()\n",
    "starnge_chrom_sum = 0\n",
    "for name in chrom_names_list:\n",
    "    if (name not in chromosome_names):\n",
    "        starnge_chrom_sum += 1\n",
    "print \"Leaving outside \"+str(starnge_chrom_sum)+\" out of \"+str(len(chrom_names_list))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "#A function that get chromosome raw data from the hmmer results and return a table of the exons\n",
    "def create_exon_pos_table(chrom_raw):\n",
    "    exons_raw = chrom_raw\n",
    "    \n",
    "    #Removing the complement bracates if exist\n",
    "    if (exons_raw.find(\"complement(\") >= 0):\n",
    "        exons_raw = exons_raw[exons_raw.find(\"complement(\")+11:-1]\n",
    "    \n",
    "    #Removing the join bracates if exist\n",
    "    if (exons_raw.find(\"join(\") >= 0):\n",
    "        exons_raw = exons_raw[exons_raw.find(\"join(\")+5:-1]\n",
    "        \n",
    "    #In case there's only one exon, take everything after the second \":\"\n",
    "    else:\n",
    "        exons_raw = exons_raw[exons_raw.find(\":\", chrom_raw.find(\":\")+1)+1:]\n",
    "    \n",
    "    exons_list = exons_raw.split(\",\")\n",
    "    exon_pos = []\n",
    "    for ex in exons_list:\n",
    "        exon_pos.append(ex.split(\"..\"))\n",
    "    exon_df = pd.DataFrame(exon_pos)\n",
    "    exon_df.columns = [\"start_pos\", \"end_pos\"]\n",
    "    exon_len = []\n",
    "    for index, exon in exon_df.iterrows():\n",
    "        exon_len.append(int(exon[1]) - int(exon[0])+1)\n",
    "    exon_df[\"length\"] = exon_len\n",
    "    first_bp_count = 1\n",
    "    first_bp_list = []\n",
    "    for index, exon in exon_df.iterrows():\n",
    "        first_bp_list.append(first_bp_count)\n",
    "        first_bp_count += int(exon[2])\n",
    "    exon_df[\"first_bp_count\"] = first_bp_list\n",
    "    return(exon_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "#A function that get chromosome position and table of exons, and return the protein position or -1 if it's not within any exon\n",
    "def find_protein_pos(chrom_pos, exon_df, chrom_raw):\n",
    "    for index, exon in exon_df.iterrows():\n",
    "        start_pos = int(exon[0])\n",
    "        end_pos = int(exon[1])\n",
    "        first_bp_count = int(exon[3])\n",
    "        if (chrom_pos >= start_pos and chrom_pos <= end_pos):\n",
    "            \n",
    "            #Calculate position for reverse complement strand: the protein is translated from the end position towards the start position of the exon\n",
    "            if (chrom_raw.find(\"complement\")):\n",
    "                len_from_exon_start = end_pos - chrom_pos\n",
    "            #Calculate position for forward starnd\n",
    "            else:\n",
    "                len_from_exon_start = chrom_pos - start_pos\n",
    "            \n",
    "            #Calculate the position on the mRNA transcript\n",
    "            transcript_pos = len_from_exon_start + first_bp_count\n",
    "            \n",
    "            #Calculate the position on the protein sequence\n",
    "            protein_pos = int(math.ceil(float(transcript_pos)/3))\n",
    "            return protein_pos\n",
    "    \n",
    "    #If the position wasn't in the regions of any exon\n",
    "    return -1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "#A function that get protein position and table of exons, and return the chromosome position of the 1st bp of the codon\n",
    "def find_chrom_bps(protein_pos, exon_table, chrom_raw_data):\n",
    "    \n",
    "    #calculate the mRNA transcript index of this protein position (the 1st bp in the triplet)\n",
    "    transcript_pos = (protein_pos*3)-2\n",
    "    \n",
    "    #Iterating over all the gene exons\n",
    "    for index, exon in exon_table.iterrows():  \n",
    "        first_bp_count = int(exon[\"first_bp_count\"])\n",
    "        exon_length = int(exon[\"length\"])\n",
    "        last_bp_count = first_bp_count + exon_length - 1\n",
    "        \n",
    "        #Checking if the transcript position is within this exon\n",
    "        if (first_bp_count <= transcript_pos and transcript_pos <= last_bp_count):\n",
    "            \n",
    "            start_pos = int(exon[\"start_pos\"])\n",
    "            end_pos = int(exon[\"end_pos\"])\n",
    "            \n",
    "            len_from_exon_start = transcript_pos - first_bp_count\n",
    "            \n",
    "            #Calculate bps position for reverse complement strand: the protein is translated from the end position towards the start position of the exon\n",
    "            if (chrom_raw_data.find(\"complement\")):\n",
    "                chrom_pos_1st = end_pos - len_from_exon_start\n",
    "                \n",
    "                chrom_pos_2nd = chrom_pos_1st - 1\n",
    "                #If the exons ends here: move to the next exon\n",
    "                if (chrom_pos_2nd < start_pos):\n",
    "                    index += 1\n",
    "                    chrom_pos_2nd = int(exon_table[\"end_pos\"][index])\n",
    "                    start_pos = int(exon[\"start_pos\"][index])\n",
    "                    end_pos = int(exon[\"end_pos\"][index])\n",
    "                \n",
    "                #If the exons ends here: move to the next exon\n",
    "                chrom_pos_3rd = chrom_pos_2nd - 1\n",
    "                if (chrom_pos_3rd < start_pos):\n",
    "                    index += 1\n",
    "                    chrom_pos_3rd = int(exon_table[\"end_pos\"][index])\n",
    "                    \n",
    "            #Calculate position for forward strand\n",
    "            else:\n",
    "                chrom_pos_1st = start_pos + len_from_exon_start\n",
    "                \n",
    "                chrom_pos_2nd = chrom_pos_1st + 1\n",
    "                 #If the exons ends here: move to the next exon\n",
    "                if (chrom_pos_2nd > end_pos):\n",
    "                    index += 1\n",
    "                    chrom_pos_2nd = int(exon_table[\"start_pos\"][index])\n",
    "                    start_pos = int(exon[\"start_pos\"][index])\n",
    "                    end_pos = int(exon[\"end_pos\"][index])\n",
    "                \n",
    "                #If the exons ends here: move to the next exon\n",
    "                chrom_pos_3rd = chrom_pos_2nd + 1\n",
    "                if (chrom_pos_3rd > end_pos):\n",
    "                    index += 1\n",
    "                    chrom_pos_3rd = int(exon_table[\"start_pos\"][index])\n",
    "            \n",
    "            return (chrom_pos_1st, chrom_pos_2nd, chrom_pos_3rd)\n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "#Boolean function - determine if a given text can be converted to a number\n",
    "def is_number(s):\n",
    "    try:\n",
    "        float(s)\n",
    "        return True\n",
    "    except ValueError:\n",
    "        return False"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 94,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "#A function that return the hmm state of that protein position\n",
    "# return -1 for positions outside of domains regions, -2 for matching insertion\n",
    "#TODO: do we need also transcript id? do we want to consider more than 1 transcript per gene?\n",
    "def protein_pos_to_hmm_state_and_aa(protein_pos, domain_gene_table):\n",
    "    for index, row in domain_gene_table.iterrows():\n",
    "        target_start = row[\"TargetStart\"]\n",
    "        target_end = row[\"TargetEnd\"]\n",
    "        \n",
    "        #Check if the position is inside this domain instance of the gene\n",
    "        if (protein_pos >= target_start and protein_pos <= target_end):\n",
    "            \n",
    "            #Find the HMM match state\n",
    "            hmm_pos = (row[\"HMM_Pos\"]).split(\",\")\n",
    "            index_inside_match = (protein_pos - target_start)\n",
    "            hmm_state_text = hmm_pos[index_inside_match]\n",
    "            if is_number(hmm_state_text):\n",
    "                hmm_state = int(hmm_state_text)\n",
    "            else:\n",
    "                #the position match insertion\n",
    "                hmm_state = -2\n",
    "                \n",
    "            #Find the amino acid\n",
    "            target_seq = list(row[\"Target_Seq\"])\n",
    "            aa = target_seq[index_inside_match]\n",
    "            \n",
    "            return(hmm_state, aa)\n",
    "            \n",
    "    #The protein position isn't in any domain region        \n",
    "    return (-1,'-')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 203,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "#A function that create the new codon for the alteration\n",
    "def create_alt_codon(curr_alt_bp, ref_codon, alt_codon_pos, chrom_raw_data):\n",
    "    #Complement strand - transversing the bp to base-complement\n",
    "    if (chrom_raw_data.find(\"complement\")):\n",
    "        if (curr_alt_bp == 'A'):\n",
    "            new_bp = 'T'\n",
    "        elif (curr_alt_bp == 'T'):\n",
    "            new_bp = 'A'\n",
    "        elif (curr_alt_bp == 'G'):\n",
    "            new_bp = 'C'\n",
    "        else:\n",
    "            new_bp = 'G'\n",
    "    #Regular starnd\n",
    "    else:\n",
    "        new_bp = curr_alt_bp\n",
    "        \n",
    "    new_alt_codon = ref_codon[:alt_codon_pos]+new_bp+ref_codon[alt_codon_pos+1:]\n",
    "    \n",
    "    return new_alt_codon"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 212,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "codon_table = {\n",
    "    'ATA':'I', 'ATC':'I', 'ATT':'I', 'ATG':'M',\n",
    "    'ACA':'T', 'ACC':'T', 'ACG':'T', 'ACT':'T',\n",
    "    'AAC':'N', 'AAT':'N', 'AAA':'K', 'AAG':'K',\n",
    "    'AGC':'S', 'AGT':'S', 'AGA':'R', 'AGG':'R',\n",
    "    'CTA':'L', 'CTC':'L', 'CTG':'L', 'CTT':'L',\n",
    "    'CCA':'P', 'CCC':'P', 'CCG':'P', 'CCT':'P',\n",
    "    'CAC':'H', 'CAT':'H', 'CAA':'Q', 'CAG':'Q',\n",
    "    'CGA':'R', 'CGC':'R', 'CGG':'R', 'CGT':'R',\n",
    "    'GTA':'V', 'GTC':'V', 'GTG':'V', 'GTT':'V',\n",
    "    'GCA':'A', 'GCC':'A', 'GCG':'A', 'GCT':'A',\n",
    "    'GAC':'D', 'GAT':'D', 'GAA':'E', 'GAG':'E',\n",
    "    'GGA':'G', 'GGC':'G', 'GGG':'G', 'GGT':'G',\n",
    "    'TCA':'S', 'TCC':'S', 'TCG':'S', 'TCT':'S',\n",
    "    'TTC':'F', 'TTT':'F', 'TTA':'L', 'TTG':'L',\n",
    "    'TAC':'Y', 'TAT':'Y', 'TAA':'_', 'TAG':'_',\n",
    "    'TGC':'C', 'TGT':'C', 'TGA':'_', 'TGG':'W',\n",
    "    }"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 264,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "#A function that return a dict with the MAF info for the protein position and corresponding chromosomal location\n",
    "def calc_exac_maf_data(chrom_pos_list, chrom_gene_table, protein_pos, aa, chrom_raw_data):\n",
    "    \n",
    "    res_dict = {}\n",
    "    res_dict[\"chrom\"] = chrom\n",
    "    res_dict[\"chrom_pos\"] = (chrom_pos_1st, chrom_pos_2nd, chrom_pos_3rd)\n",
    "    res_dict[\"prot_pos\"] = protein_pos\n",
    "    res_dict[\"aa_ref\"] = aa\n",
    "    \n",
    "    alterations_af_dict = defaultdict(list)\n",
    "    alterations_af_adj_dict = defaultdict(list)\n",
    "    for chrom_pos in chrom_pos_list:\n",
    "        #Retreiving relevant ExAC entry\n",
    "        chrom_alter = chrom_gene_table[chrom_gene_table[\"pos\"] == chrom_pos]\n",
    "        chrom_alter = chrom_alter.reset_index(drop=True)\n",
    "                \n",
    "        if (chrom_alter.shape[0] == 0):\n",
    "            #No ExAC entry for this chromosome position\n",
    "            continue\n",
    "        else:\n",
    "            #Validation: the ExAC chromosome position is within the protein\n",
    "            if (chrom_alter[\"prot_pos\"][0] == \"\"):\n",
    "                print \"Error: ExAC chromosome position doesn't correspond to a protein\"\n",
    "                #Not adding any alteration data\n",
    "                continue\n",
    "                    \n",
    "            #The ExAC entry has protein details\n",
    "            else:\n",
    "                #Validation: the ExAC protein position match the HMMER protein position\n",
    "                exac_prot_pos = int(chrom_alter_1st[\"prot_pos\"][0])\n",
    "                if (exac_prot_pos != protein_pos):\n",
    "                    print \"Error: ExAC protein position mismatch\"\n",
    "\n",
    "                #Validation: the ExAC aa match the HMMER aa\n",
    "                exac_aa = chrom_alter[\"amino_acids\"][0]\n",
    "                exac_aa_ref = exac_aa[:1]\n",
    "                exac_aa_alt = exac_aa[exac_aa.find(\"/\")+1:]\n",
    "                if (exac_aa[:1] != aa):\n",
    "                    print \"Error: ExAC amino acid identity mismatch\"\n",
    "                \n",
    "                #Creating lists in case there's more than one alteration in this chromosomal position\n",
    "                alt_bp_list = (chrom_alter[\"alt\"][0]).split(\",\")\n",
    "                af_list = (chrom_alter[\"AF\"][0]).split(\",\")\n",
    "                an_adj = int(chrom_alter[\"AN_Adj\"][0])\n",
    "                ac_adj_list = (chrom_alter[\"AC_Adj\"][0]).split(\",\")\n",
    "                codons = chrom_alter[\"codons\"][0]\n",
    "                ref_codon = codons[:codons.find(\"/\")]\n",
    "                alt_codon_pos = 0\n",
    "                for c in ref_codon:\n",
    "                    if c.isupper():\n",
    "                        break\n",
    "                    alt_codon_pos += 1\n",
    "                \n",
    "                for j in range(len(alt_bp_list)):\n",
    "                    #The first alteration has ExAC mata-data\n",
    "                    if (j == 0):\n",
    "                        #Checking if it's synonymous\n",
    "                        if (exac_aa_alt == exac_aa_ref):\n",
    "                            #Not adding any alteration data\n",
    "                            continue\n",
    "                        #Non-synonymous\n",
    "                        else:\n",
    "                            alterations_af_dict[exac_aa_alt].append(af_list[0])\n",
    "                            af_adj = float(ac_adj_list[0])/float(an_adj)\n",
    "                            af_adj_format = float('{:.3e}'.format(float(af_adj)))\n",
    "                            alterations_af_adj_dict[exac_aa_alt].append(af_adj_format)\n",
    "                    #Additional alterations don't have ExAC meta-data\n",
    "                    else:\n",
    "                        curr_alt_bp = alt_bp_list[j]\n",
    "                        alt_codon = create_alt_codon(curr_alt_bp, ref_codon, alt_codon_pos, chrom_raw_data)\n",
    "                        alt_aa = codon_table[alt_codon.upper()]\n",
    "                        #Checking if it's synonymous\n",
    "                        if (alt_aa == exac_aa_ref):\n",
    "                            #Not adding any alteration data\n",
    "                            continue\n",
    "                        #Non-synonymous\n",
    "                        else:\n",
    "                            alterations_af_dict[alt_aa].append(af_list[j])\n",
    "                            af_adj = float(ac_adj_list[j])/float(an_adj)\n",
    "                            af_adj_format = float('{:.3e}'.format(float(af_adj)))\n",
    "                            alterations_af_adj_dict[alt_aa].append(af_adj_format)\n",
    "    \n",
    "    #Calculating the overall MAF from the alteration dicts\n",
    "    res_dict[\"af\"] = 0\n",
    "    res_dict[\"af_adj\"] = 0\n",
    "    \n",
    "    #Checking if any alteration is above 0.5, and changing the ref accordingly\n",
    "    for aa in alterations_af_dict.keys():\n",
    "        aa_sum = sum(alterations_af_dict[aa])\n",
    "        aa_adj_sum = sum(alterations_af_adj_dict[aa])\n",
    "        if (aa_sum > 0.5):\n",
    "            res_dict[\"aa_ref\"] = aa\n",
    "            res_dict[\"af\"] = (1 - aa_sum)\n",
    "            res_dict[\"af_adj\"] = (1 - aa_adj_sum)\n",
    "        else:\n",
    "            res_dict[\"af\"] += aa_sum\n",
    "            res_dict[\"af_adj\"] += aa_adj_sum\n",
    "    \n",
    "    return res_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 267,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>#TargetID</th>\n",
       "      <th>Hugo_symbol</th>\n",
       "      <th>pfam_id</th>\n",
       "      <th>domain_name</th>\n",
       "      <th>E-value</th>\n",
       "      <th>BitScore</th>\n",
       "      <th>TargetStart</th>\n",
       "      <th>TargetEnd</th>\n",
       "      <th>HMM_Seq</th>\n",
       "      <th>Target_Seq</th>\n",
       "      <th>HMM_Pos</th>\n",
       "      <th>prot_id</th>\n",
       "      <th>ensembl_id</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>chromosome_id</th>\n",
       "      <th>chrom_num</th>\n",
       "      <th>length</th>\n",
       "      <th>refseq</th>\n",
       "      <th>hmm_start</th>\n",
       "      <th>hmm_end</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>ZBTB17.002</td>\n",
       "      <td>ZBTB17</td>\n",
       "      <td>PF00096</td>\n",
       "      <td>zf-C2H2</td>\n",
       "      <td>2.700000e-10</td>\n",
       "      <td>35.1</td>\n",
       "      <td>390</td>\n",
       "      <td>412</td>\n",
       "      <td>ykCpdCgksFkrksnLkrHirtH</td>\n",
       "      <td>YRCEDCGKLFTTSGNLKRHQLVH</td>\n",
       "      <td>1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,1...</td>\n",
       "      <td>ENSP00000364895</td>\n",
       "      <td>ENSG00000116809</td>\n",
       "      <td>ENST00000375743</td>\n",
       "      <td>GRCh37.75:1:complement(join(16274786..16274990...</td>\n",
       "      <td>1</td>\n",
       "      <td>803</td>\n",
       "      <td>NP_003434</td>\n",
       "      <td>1</td>\n",
       "      <td>23</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>ZBTB17.002</td>\n",
       "      <td>ZBTB17</td>\n",
       "      <td>PF00096</td>\n",
       "      <td>zf-C2H2</td>\n",
       "      <td>9.100000e-10</td>\n",
       "      <td>33.4</td>\n",
       "      <td>474</td>\n",
       "      <td>496</td>\n",
       "      <td>ykCpdCgksFkrksnLkrHirtH</td>\n",
       "      <td>LKCRECGKQFTTSGNLKRHLRIH</td>\n",
       "      <td>1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,1...</td>\n",
       "      <td>ENSP00000364895</td>\n",
       "      <td>ENSG00000116809</td>\n",
       "      <td>ENST00000375743</td>\n",
       "      <td>GRCh37.75:1:complement(join(16274786..16274990...</td>\n",
       "      <td>1</td>\n",
       "      <td>803</td>\n",
       "      <td>NP_003434</td>\n",
       "      <td>1</td>\n",
       "      <td>23</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ZBTB17.002</td>\n",
       "      <td>ZBTB17</td>\n",
       "      <td>PF00096</td>\n",
       "      <td>zf-C2H2</td>\n",
       "      <td>1.500000e-09</td>\n",
       "      <td>32.7</td>\n",
       "      <td>306</td>\n",
       "      <td>328</td>\n",
       "      <td>ykCpdCgksFkrksnLkrHirtH</td>\n",
       "      <td>HKCEDCGKEFTHTGNFKRHIRIH</td>\n",
       "      <td>1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,1...</td>\n",
       "      <td>ENSP00000364895</td>\n",
       "      <td>ENSG00000116809</td>\n",
       "      <td>ENST00000375743</td>\n",
       "      <td>GRCh37.75:1:complement(join(16274786..16274990...</td>\n",
       "      <td>1</td>\n",
       "      <td>803</td>\n",
       "      <td>NP_003434</td>\n",
       "      <td>1</td>\n",
       "      <td>23</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>ZBTB17.002</td>\n",
       "      <td>ZBTB17</td>\n",
       "      <td>PF00096</td>\n",
       "      <td>zf-C2H2</td>\n",
       "      <td>1.600000e-09</td>\n",
       "      <td>32.5</td>\n",
       "      <td>558</td>\n",
       "      <td>580</td>\n",
       "      <td>ykCpdCgksFkrksnLkrHirtH</td>\n",
       "      <td>YVCERCGKRFVQSSQLANHIRHH</td>\n",
       "      <td>1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,1...</td>\n",
       "      <td>ENSP00000364895</td>\n",
       "      <td>ENSG00000116809</td>\n",
       "      <td>ENST00000375743</td>\n",
       "      <td>GRCh37.75:1:complement(join(16274786..16274990...</td>\n",
       "      <td>1</td>\n",
       "      <td>803</td>\n",
       "      <td>NP_003434</td>\n",
       "      <td>1</td>\n",
       "      <td>23</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>ZBTB17.002</td>\n",
       "      <td>ZBTB17</td>\n",
       "      <td>PF00096</td>\n",
       "      <td>zf-C2H2</td>\n",
       "      <td>4.800000e-09</td>\n",
       "      <td>30.9</td>\n",
       "      <td>418</td>\n",
       "      <td>440</td>\n",
       "      <td>ykCpdCgksFkrksnLkrHirtH</td>\n",
       "      <td>YQCDYCGRSFSDPTSKMRHLETH</td>\n",
       "      <td>1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,1...</td>\n",
       "      <td>ENSP00000364895</td>\n",
       "      <td>ENSG00000116809</td>\n",
       "      <td>ENST00000375743</td>\n",
       "      <td>GRCh37.75:1:complement(join(16274786..16274990...</td>\n",
       "      <td>1</td>\n",
       "      <td>803</td>\n",
       "      <td>NP_003434</td>\n",
       "      <td>1</td>\n",
       "      <td>23</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>ZBTB17.002</td>\n",
       "      <td>ZBTB17</td>\n",
       "      <td>PF00096</td>\n",
       "      <td>zf-C2H2</td>\n",
       "      <td>9.600000e-09</td>\n",
       "      <td>30.0</td>\n",
       "      <td>446</td>\n",
       "      <td>468</td>\n",
       "      <td>ykCpdCgksFkrksnLkrHirtH</td>\n",
       "      <td>HKCPHCDKKFNQVGNLKAHLKIH</td>\n",
       "      <td>1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,1...</td>\n",
       "      <td>ENSP00000364895</td>\n",
       "      <td>ENSG00000116809</td>\n",
       "      <td>ENST00000375743</td>\n",
       "      <td>GRCh37.75:1:complement(join(16274786..16274990...</td>\n",
       "      <td>1</td>\n",
       "      <td>803</td>\n",
       "      <td>NP_003434</td>\n",
       "      <td>1</td>\n",
       "      <td>23</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>ZBTB17.002</td>\n",
       "      <td>ZBTB17</td>\n",
       "      <td>PF00096</td>\n",
       "      <td>zf-C2H2</td>\n",
       "      <td>1.400000e-08</td>\n",
       "      <td>29.4</td>\n",
       "      <td>530</td>\n",
       "      <td>552</td>\n",
       "      <td>ykCpdCgksFkrksnLkrHirtH</td>\n",
       "      <td>CQCVMCGKAFTQASSLIAHVRQH</td>\n",
       "      <td>1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,1...</td>\n",
       "      <td>ENSP00000364895</td>\n",
       "      <td>ENSG00000116809</td>\n",
       "      <td>ENST00000375743</td>\n",
       "      <td>GRCh37.75:1:complement(join(16274786..16274990...</td>\n",
       "      <td>1</td>\n",
       "      <td>803</td>\n",
       "      <td>NP_003434</td>\n",
       "      <td>1</td>\n",
       "      <td>23</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>ZBTB17.002</td>\n",
       "      <td>ZBTB17</td>\n",
       "      <td>PF00096</td>\n",
       "      <td>zf-C2H2</td>\n",
       "      <td>4.500000e-08</td>\n",
       "      <td>27.7</td>\n",
       "      <td>614</td>\n",
       "      <td>637</td>\n",
       "      <td>ykCpdCgksFkrksnLkrHirt.H</td>\n",
       "      <td>YLCDKCGRGFNRVDNLRSHVKTvH</td>\n",
       "      <td>1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,1...</td>\n",
       "      <td>ENSP00000364895</td>\n",
       "      <td>ENSG00000116809</td>\n",
       "      <td>ENST00000375743</td>\n",
       "      <td>GRCh37.75:1:complement(join(16274786..16274990...</td>\n",
       "      <td>1</td>\n",
       "      <td>803</td>\n",
       "      <td>NP_003434</td>\n",
       "      <td>1</td>\n",
       "      <td>23</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     #TargetID Hugo_symbol  pfam_id domain_name       E-value  BitScore  \\\n",
       "7   ZBTB17.002      ZBTB17  PF00096     zf-C2H2  2.700000e-10      35.1   \n",
       "15  ZBTB17.002      ZBTB17  PF00096     zf-C2H2  9.100000e-10      33.4   \n",
       "2   ZBTB17.002      ZBTB17  PF00096     zf-C2H2  1.500000e-09      32.7   \n",
       "21  ZBTB17.002      ZBTB17  PF00096     zf-C2H2  1.600000e-09      32.5   \n",
       "10  ZBTB17.002      ZBTB17  PF00096     zf-C2H2  4.800000e-09      30.9   \n",
       "12  ZBTB17.002      ZBTB17  PF00096     zf-C2H2  9.600000e-09      30.0   \n",
       "18  ZBTB17.002      ZBTB17  PF00096     zf-C2H2  1.400000e-08      29.4   \n",
       "23  ZBTB17.002      ZBTB17  PF00096     zf-C2H2  4.500000e-08      27.7   \n",
       "\n",
       "    TargetStart  TargetEnd                   HMM_Seq  \\\n",
       "7           390        412   ykCpdCgksFkrksnLkrHirtH   \n",
       "15          474        496   ykCpdCgksFkrksnLkrHirtH   \n",
       "2           306        328   ykCpdCgksFkrksnLkrHirtH   \n",
       "21          558        580   ykCpdCgksFkrksnLkrHirtH   \n",
       "10          418        440   ykCpdCgksFkrksnLkrHirtH   \n",
       "12          446        468   ykCpdCgksFkrksnLkrHirtH   \n",
       "18          530        552   ykCpdCgksFkrksnLkrHirtH   \n",
       "23          614        637  ykCpdCgksFkrksnLkrHirt.H   \n",
       "\n",
       "                  Target_Seq  \\\n",
       "7    YRCEDCGKLFTTSGNLKRHQLVH   \n",
       "15   LKCRECGKQFTTSGNLKRHLRIH   \n",
       "2    HKCEDCGKEFTHTGNFKRHIRIH   \n",
       "21   YVCERCGKRFVQSSQLANHIRHH   \n",
       "10   YQCDYCGRSFSDPTSKMRHLETH   \n",
       "12   HKCPHCDKKFNQVGNLKAHLKIH   \n",
       "18   CQCVMCGKAFTQASSLIAHVRQH   \n",
       "23  YLCDKCGRGFNRVDNLRSHVKTvH   \n",
       "\n",
       "                                              HMM_Pos          prot_id  \\\n",
       "7   1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,1...  ENSP00000364895   \n",
       "15  1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,1...  ENSP00000364895   \n",
       "2   1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,1...  ENSP00000364895   \n",
       "21  1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,1...  ENSP00000364895   \n",
       "10  1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,1...  ENSP00000364895   \n",
       "12  1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,1...  ENSP00000364895   \n",
       "18  1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,1...  ENSP00000364895   \n",
       "23  1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,1...  ENSP00000364895   \n",
       "\n",
       "         ensembl_id    transcript_id  \\\n",
       "7   ENSG00000116809  ENST00000375743   \n",
       "15  ENSG00000116809  ENST00000375743   \n",
       "2   ENSG00000116809  ENST00000375743   \n",
       "21  ENSG00000116809  ENST00000375743   \n",
       "10  ENSG00000116809  ENST00000375743   \n",
       "12  ENSG00000116809  ENST00000375743   \n",
       "18  ENSG00000116809  ENST00000375743   \n",
       "23  ENSG00000116809  ENST00000375743   \n",
       "\n",
       "                                        chromosome_id chrom_num  length  \\\n",
       "7   GRCh37.75:1:complement(join(16274786..16274990...         1     803   \n",
       "15  GRCh37.75:1:complement(join(16274786..16274990...         1     803   \n",
       "2   GRCh37.75:1:complement(join(16274786..16274990...         1     803   \n",
       "21  GRCh37.75:1:complement(join(16274786..16274990...         1     803   \n",
       "10  GRCh37.75:1:complement(join(16274786..16274990...         1     803   \n",
       "12  GRCh37.75:1:complement(join(16274786..16274990...         1     803   \n",
       "18  GRCh37.75:1:complement(join(16274786..16274990...         1     803   \n",
       "23  GRCh37.75:1:complement(join(16274786..16274990...         1     803   \n",
       "\n",
       "        refseq  hmm_start  hmm_end  \n",
       "7   NP_003434           1       23  \n",
       "15  NP_003434           1       23  \n",
       "2   NP_003434           1       23  \n",
       "21  NP_003434           1       23  \n",
       "10  NP_003434           1       23  \n",
       "12  NP_003434           1       23  \n",
       "18  NP_003434           1       23  \n",
       "23  NP_003434           1       23  "
      ]
     },
     "execution_count": 267,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "domain_gene_table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 268,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Error: ExAC chromosome position doesn't correspond to a protein\n"
     ]
    }
   ],
   "source": [
    "protein_pos = 550\n",
    "(hmm_state, aa) = protein_pos_to_hmm_state_and_aa(protein_pos, domain_gene_table)\n",
    "chrom_pos_list = find_chrom_bps(protein_pos, exon_table, chrom_raw_data)\n",
    "info_dict = calc_exac_maf_data(chrom_pos_list, chrom_gene_table, protein_pos, aa, chrom_raw_data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 282,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'R'"
      ]
     },
     "execution_count": 282,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "aa"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 274,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>start_pos</th>\n",
       "      <th>end_pos</th>\n",
       "      <th>length</th>\n",
       "      <th>first_bp_count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>16274786</td>\n",
       "      <td>16274990</td>\n",
       "      <td>205</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>16273430</td>\n",
       "      <td>16273618</td>\n",
       "      <td>189</td>\n",
       "      <td>206</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>16272649</td>\n",
       "      <td>16272789</td>\n",
       "      <td>141</td>\n",
       "      <td>395</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>16272210</td>\n",
       "      <td>16272335</td>\n",
       "      <td>126</td>\n",
       "      <td>536</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>16271432</td>\n",
       "      <td>16271697</td>\n",
       "      <td>266</td>\n",
       "      <td>662</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>16271192</td>\n",
       "      <td>16271334</td>\n",
       "      <td>143</td>\n",
       "      <td>928</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>16270795</td>\n",
       "      <td>16271095</td>\n",
       "      <td>301</td>\n",
       "      <td>1071</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>16270303</td>\n",
       "      <td>16270390</td>\n",
       "      <td>88</td>\n",
       "      <td>1372</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>16270094</td>\n",
       "      <td>16270210</td>\n",
       "      <td>117</td>\n",
       "      <td>1460</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>16269894</td>\n",
       "      <td>16270014</td>\n",
       "      <td>121</td>\n",
       "      <td>1577</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>16269559</td>\n",
       "      <td>16269689</td>\n",
       "      <td>131</td>\n",
       "      <td>1698</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>16269024</td>\n",
       "      <td>16269233</td>\n",
       "      <td>210</td>\n",
       "      <td>1829</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>16268826</td>\n",
       "      <td>16268915</td>\n",
       "      <td>90</td>\n",
       "      <td>2039</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>16268464</td>\n",
       "      <td>16268747</td>\n",
       "      <td>284</td>\n",
       "      <td>2129</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   start_pos   end_pos  length  first_bp_count\n",
       "0   16274786  16274990     205               1\n",
       "1   16273430  16273618     189             206\n",
       "2   16272649  16272789     141             395\n",
       "3   16272210  16272335     126             536\n",
       "4   16271432  16271697     266             662\n",
       "5   16271192  16271334     143             928\n",
       "6   16270795  16271095     301            1071\n",
       "7   16270303  16270390      88            1372\n",
       "8   16270094  16270210     117            1460\n",
       "9   16269894  16270014     121            1577\n",
       "10  16269559  16269689     131            1698\n",
       "11  16269024  16269233     210            1829\n",
       "12  16268826  16268915      90            2039\n",
       "13  16268464  16268747     284            2129"
      ]
     },
     "execution_count": 274,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "exon_table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 270,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(16269943, 16269942, 16269941)"
      ]
     },
     "execution_count": 270,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "chrom_pos_list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 281,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>pos</th>\n",
       "      <th>id</th>\n",
       "      <th>ref</th>\n",
       "      <th>alt</th>\n",
       "      <th>qual</th>\n",
       "      <th>filter</th>\n",
       "      <th>AC</th>\n",
       "      <th>AC_Adj</th>\n",
       "      <th>AF</th>\n",
       "      <th>...</th>\n",
       "      <th>DP</th>\n",
       "      <th>gene</th>\n",
       "      <th>conseq</th>\n",
       "      <th>prot_pos</th>\n",
       "      <th>amino_acids</th>\n",
       "      <th>codons</th>\n",
       "      <th>SWISSPROT</th>\n",
       "      <th>SIFT</th>\n",
       "      <th>domains</th>\n",
       "      <th>clin_sig</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>175</th>\n",
       "      <td>1</td>\n",
       "      <td>16269941</td>\n",
       "      <td>.</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>1631.48</td>\n",
       "      <td>PASS</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>9.416e-06</td>\n",
       "      <td>...</td>\n",
       "      <td>1363977</td>\n",
       "      <td>ENSG00000116809</td>\n",
       "      <td>non_coding_transcript_exon_variant&amp;non_coding_...</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1 rows × 22 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     chrom       pos id ref alt     qual filter AC AC_Adj         AF   ...     \\\n",
       "175      1  16269941  .   G   C  1631.48   PASS  1      1  9.416e-06   ...      \n",
       "\n",
       "          DP             gene  \\\n",
       "175  1363977  ENSG00000116809   \n",
       "\n",
       "                                                conseq prot_pos amino_acids  \\\n",
       "175  non_coding_transcript_exon_variant&non_coding_...                        \n",
       "\n",
       "    codons SWISSPROT SIFT domains clin_sig  \n",
       "175                                         \n",
       "\n",
       "[1 rows x 22 columns]"
      ]
     },
     "execution_count": 281,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "chrom_gene_table[chrom_gene_table[\"pos\"] == 16269941]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 279,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "problem = chrom_gene_table[chrom_gene_table[\"pos\"] == 16269941]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 280,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'AC': {175: '1'},\n",
       " 'AC_Adj': {175: '1'},\n",
       " 'AF': {175: '9.416e-06'},\n",
       " 'AN': {175: 106202},\n",
       " 'AN_Adj': {175: 101894},\n",
       " 'DP': {175: 1363977},\n",
       " 'SIFT': {175: ''},\n",
       " 'SWISSPROT': {175: ''},\n",
       " 'alt': {175: 'C'},\n",
       " 'amino_acids': {175: ''},\n",
       " 'chrom': {175: 1},\n",
       " 'clin_sig': {175: ''},\n",
       " 'codons': {175: ''},\n",
       " 'conseq': {175: 'non_coding_transcript_exon_variant&non_coding_transcript_variant'},\n",
       " 'domains': {175: ''},\n",
       " 'filter': {175: 'PASS'},\n",
       " 'gene': {175: 'ENSG00000116809'},\n",
       " 'id': {175: '.'},\n",
       " 'pos': {175: 16269941},\n",
       " 'prot_pos': {175: ''},\n",
       " 'qual': {175: 1631.48},\n",
       " 'ref': {175: 'G'}}"
      ]
     },
     "execution_count": 280,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "problem.to_dict()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 255,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "res_dict = {}\n",
    "res_dict[\"chrom\"] = chrom\n",
    "res_dict[\"chrom_pos\"] = (chrom_pos_1st, chrom_pos_2nd, chrom_pos_3rd)\n",
    "res_dict[\"prot_pos\"] = protein_pos\n",
    "res_dict[\"aa_ref\"] = aa\n",
    "\n",
    "alterations_af_dict = defaultdict(list)\n",
    "alterations_af_adj_dict = defaultdict(list)\n",
    "\n",
    "for chrom_pos in chrom_pos_list:\n",
    "        #Retreiving relevant ExAC entry\n",
    "        chrom_alter = chrom_gene_table[chrom_gene_table[\"pos\"] == chrom_pos]\n",
    "        chrom_alter = chrom_alter_1st.reset_index(drop=True)\n",
    "                \n",
    "        if (chrom_alter.shape[0] == 0):\n",
    "            #No ExAC entry for this chromosome position\n",
    "            continue\n",
    "        else:\n",
    "            #Validation: the ExAC chromosome position is within the protein\n",
    "            if (chrom_alter[\"prot_pos\"][0] == \"\"):\n",
    "                print \"Error: ExAC chromosome position doesn't correspond to a protein\"\n",
    "                #Not adding any alteration data\n",
    "                continue"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 253,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "No ExAC\n"
     ]
    }
   ],
   "source": [
    "chrom_pos = chrom_pos_list[2]\n",
    "chrom_alter = chrom_gene_table[chrom_gene_table[\"pos\"] == chrom_pos]\n",
    "\n",
    "if (chrom_alter.shape[0] == 0):\n",
    "    print \"No ExAC\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 252,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "res_dict[\"af\"] = 0\n",
    "res_dict[\"af_adj\"] = 0\n",
    "\n",
    "for aa in alterations_af_dict.keys():\n",
    "    aa_sum = sum(alterations_af_dict[aa])\n",
    "    aa_adj_sum = sum(alterations_af_adj_dict[aa])\n",
    "    if (aa_sum > 0.5):\n",
    "        res_dict[\"aa_ref\"] = aa\n",
    "        res_dict[\"af\"] = (1 - aa_sum)\n",
    "        res_dict[\"af_adj\"] = (1 - aa_adj_sum)\n",
    "    else:\n",
    "        res_dict[\"af\"] += aa_sum\n",
    "        res_dict[\"af_adj\"] += aa_adj_sum"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "chrom_path = curr_dir[0]+\"/parsed/\"\n",
    "chrom_filename = \"parsed_chrom\"\n",
    "states_dict = defaultdict(list)\n",
    "\n",
    "for chrom in chromosome_names:\n",
    "    \n",
    "    #Filtering the domain data relevant to this chromosome\n",
    "    domain_chrom_data = sorted_zinc[sorted_zinc[\"chrom_num\"] == chrom]\n",
    "    \n",
    "    #Loading the ExAC parsed data of this chromosome\n",
    "    chrom_csv = pd.read_csv(chrom_path+chrom_filename+chrom+\".csv\", sep='\\t', index_col=0)\n",
    "    chrom_csv = chrom_csv.sort_values(by=[\"pos\"])\n",
    "    chrom_csv = chrom_csv.reset_index(drop=True)\n",
    "    chrom_csv.fillna('', inplace=True)\n",
    "    \n",
    "    #Getting a list of all the relevant ensembl gene ids for this chromosome\n",
    "    domain_ens_genes = (domain_chrom_data[\"ensembl_id\"]).unique()\n",
    "    \n",
    "    #For each ensembl gene in the domain data - finding all the ExAC alterations\n",
    "    for ens_gene in domain_ens_genes:\n",
    "        \n",
    "        #Filtering the domain data for this gene according to the canonical protein id\n",
    "        canonic_prot = canonic_protein[ens_gene]\n",
    "        domain_gene_table = domain_chrom_data[domain_chrom_data[\"prot_id\"] == canonic_prot]\n",
    "        #Making sure that if two HMM-matches overlaps, the higher bit score will come first in the table\n",
    "        domain_gene_table = domain_gene_table.sort_values(by=\"BitScore\", ascending=False)\n",
    "        \n",
    "        #Creating a table of the exons for this gene, according to the canonical protein\n",
    "        chrom_raw_data = domain_gene_table[\"chromosome_id\"].unique()[0] #there should be only one element here\n",
    "        if (len(domain_gene_table[\"chromosome_id\"].unique()) > 1):\n",
    "            print \"Error: \"+ens_gene+\": more than one chromosome raw data\" #sanity check\n",
    "        exon_table = create_exon_pos_table(chrom_raw_data)\n",
    "        \n",
    "        #Filtering the chromosome data to the exons region\n",
    "        exons_start_pos = min(exon_table[\"start_pos\"][0],exon_table[\"start_pos\"][len(exon_table)-1]) #in case of complelemt, the minimal position could be at the last row\n",
    "        exons_end_pos = max(exon_table[\"end_pos\"][0],exon_table[\"end_pos\"][len(exon_table)-1]) #in case of complelemt, the maximal position could be at the first row\n",
    "        chrom_gene_table = chrom_csv[chrom_csv[\"pos\"] >= int(exons_start_pos)][chrom_csv[\"pos\"] <= int(exons_end_pos)]\n",
    "        chrom_gene_table = chrom_gene_table.reset_index(drop=True)\n",
    "        chrom_gene_size = chrom_gene_table.shape[0]\n",
    "        \n",
    "        #Iterating over the amino-acids of the protein\n",
    "        prot_len = domain_gene_table[\"length\"].unique()[0]\n",
    "        for protein_pos in range(1,prot_len+1):\n",
    "    \n",
    "            #Trying to match HMM-state, and retreive the aa from HMMER results\n",
    "            (hmm_state, aa) = protein_pos_to_hmm_state_and_aa(protein_pos, domain_gene_table) #TODO: what happens when two matches overlap? maybe sort to the best bit score?\n",
    "                \n",
    "            #If there's a match to HMM-state: find the corresponding codon bps chromosome positions\n",
    "            if (hmm_state > 0):\n",
    "                chrom_pos_list =find_chrom_bps(protein_pos, exon_table, chrom_raw_data)\n",
    "                \n",
    "                #Analysis of the amino-acid MAF and realted data, returned in a dictionary\n",
    "                info_dict = calc_exac_maf_data(chrom_pos_list, chrom_gene_table, protein_pos, aa, chrom_raw_data)\n",
    "                \n",
    "                #Adding the dictionary to the HMM-state list\n",
    "                states_dict[hmm_state].append(info_dict)\n",
    "                    \n",
    "    print \"Finished chromosome \"+chrom"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "with open(my_path+'zinc_hmm_states_dict.pik', 'wb') as handle:\n",
    "    pickle.dump(states_dict, handle, protocol=pickle.HIGHEST_PROTOCOL)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
