{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>.container { width:100% !important; }</style>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import math\n",
    "from collections import defaultdict\n",
    "import matplotlib.pyplot as plt\n",
    "import pickle\n",
    "import sys\n",
    "from IPython.core.display import HTML\n",
    "HTML(\"<style>.container { width:100% !important; }</style>\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "#Reading the Zinc-finger domain data\n",
    "curr_dir = !pwd\n",
    "my_path = curr_dir[0]+\"/hmm_domains/\"\n",
    "filename = \"zf-C2H2.csv\"\n",
    "#filename = \"Homeobox.csv\"\n",
    "domain_data = pd.read_csv(my_path+filename, sep='\\t', index_col=0)\n",
    "#Sort the zinc finger data\n",
    "sorted_domain_data = domain_data.sort_values(by=[\"chrom_num\", \"ensembl_id\", \"TargetStart\"])\n",
    "sorted_domain_data = sorted_domain_data.reset_index(drop=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "#Get the canonic protein id for Zinc domain\n",
    "with open(my_path+'zinc_canonic_prot.pik', 'rb') as handle:\n",
    "    canonic_protein = pickle.load(handle)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "#Get the canonic protein id for Homeobox domain\n",
    "with open(my_path+'homeobox_canonic_prot.pik', 'rb') as handle:\n",
    "    canonic_protein = pickle.load(handle)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "chromosome_names = [\"1\", \"2\", \"3\", \"4\", \"5\", \"6\", \"7\", \"8\", \"9\", \"10\", \"11\", \"12\", \"13\", \"14\", \"15\", \"16\", \"17\", \"18\", \"19\", \"20\", \"21\", \"22\", \"X\", \"Y\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Leaving outside 74 out of 10492\n"
     ]
    }
   ],
   "source": [
    "chrom_names_list = sorted_domain_data[\"chrom_num\"].tolist()\n",
    "starnge_chrom_sum = 0\n",
    "for name in chrom_names_list:\n",
    "    if (name not in chromosome_names):\n",
    "        starnge_chrom_sum += 1\n",
    "print \"Leaving outside \"+str(starnge_chrom_sum)+\" out of \"+str(len(chrom_names_list))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "#A function that get chromosome raw data from the hmmer results and return a table of the exons\n",
    "def create_exon_pos_table(chrom_raw):\n",
    "    exons_raw = chrom_raw\n",
    "    \n",
    "    #Removing the complement bracates if exist\n",
    "    if (exons_raw.find(\"complement(\") >= 0):\n",
    "        exons_raw = exons_raw[exons_raw.find(\"complement(\")+11:-1]\n",
    "    \n",
    "    #Removing the join bracates if exist\n",
    "    if (exons_raw.find(\"join(\") >= 0):\n",
    "        exons_raw = exons_raw[exons_raw.find(\"join(\")+5:-1]\n",
    "        \n",
    "    #In case there's only one exon, take everything after the second \":\"\n",
    "    else:\n",
    "        exons_raw = exons_raw[exons_raw.find(\":\", chrom_raw.find(\":\")+1)+1:]\n",
    "    \n",
    "    exons_list = exons_raw.split(\",\")\n",
    "    exon_pos = []\n",
    "    for ex in exons_list:\n",
    "        exon_pos.append(ex.split(\"..\"))\n",
    "    exon_df = pd.DataFrame(exon_pos)\n",
    "    exon_df.columns = [\"start_pos\", \"end_pos\"]\n",
    "    exon_len = []\n",
    "    for index, exon in exon_df.iterrows():\n",
    "        exon_len.append(int(exon[1]) - int(exon[0])+1)\n",
    "    exon_df[\"length\"] = exon_len\n",
    "    first_bp_count = 1\n",
    "    first_bp_list = []\n",
    "    for index, exon in exon_df.iterrows():\n",
    "        first_bp_list.append(first_bp_count)\n",
    "        first_bp_count += int(exon[2])\n",
    "    exon_df[\"first_bp_count\"] = first_bp_list\n",
    "    return(exon_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "#A function that get chromosome position and table of exons, and return the protein position or -1 if it's not within any exon\n",
    "def find_protein_pos(chrom_pos, exon_df, chrom_raw):\n",
    "    for index, exon in exon_df.iterrows():\n",
    "        start_pos = int(exon[0])\n",
    "        end_pos = int(exon[1])\n",
    "        first_bp_count = int(exon[3])\n",
    "        if (chrom_pos >= start_pos and chrom_pos <= end_pos):\n",
    "            \n",
    "            #Calculate position for reverse complement strand: the protein is translated from the end position towards the start position of the exon\n",
    "            if (chrom_raw.find(\"complement\") >= 0):\n",
    "                len_from_exon_start = end_pos - chrom_pos\n",
    "            #Calculate position for forward starnd\n",
    "            else:\n",
    "                len_from_exon_start = chrom_pos - start_pos\n",
    "            \n",
    "            #Calculate the position on the mRNA transcript\n",
    "            transcript_pos = len_from_exon_start + first_bp_count\n",
    "            \n",
    "            #Calculate the position on the protein sequence\n",
    "            protein_pos = int(math.ceil(float(transcript_pos)/3))\n",
    "            return protein_pos\n",
    "    \n",
    "    #If the position wasn't in the regions of any exon\n",
    "    return -1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "#A function that get protein position and table of exons, and return the chromosome positions of the corresponding codon\n",
    "def find_chrom_bps(protein_pos, exon_table, chrom_raw_data):\n",
    "    \n",
    "    #calculate the mRNA transcript index of this protein position (the 1st bp in the triplet)\n",
    "    transcript_pos = (protein_pos*3)-2\n",
    "    \n",
    "    #Iterating over all the gene exons\n",
    "    for index, exon in exon_table.iterrows():  \n",
    "        first_bp_count = int(exon[\"first_bp_count\"])\n",
    "        exon_length = int(exon[\"length\"])\n",
    "        last_bp_count = first_bp_count + exon_length - 1\n",
    "        \n",
    "        #Checking if the transcript position is within this exon\n",
    "        if (first_bp_count <= transcript_pos and transcript_pos <= last_bp_count):\n",
    "            \n",
    "            start_pos = int(exon[\"start_pos\"])\n",
    "            end_pos = int(exon[\"end_pos\"])\n",
    "            \n",
    "            len_from_exon_start = transcript_pos - first_bp_count\n",
    "            \n",
    "            #Calculate bps position for reverse complement strand: the protein is translated from the end position towards the start position of the exon\n",
    "            if (chrom_raw_data.find(\"complement\") >= 0):\n",
    "                chrom_pos_1st = end_pos - len_from_exon_start\n",
    "                \n",
    "                chrom_pos_2nd = chrom_pos_1st - 1\n",
    "                #If the exons ends here: move to the next exon\n",
    "                if (chrom_pos_2nd < start_pos):\n",
    "                    index += 1\n",
    "                    chrom_pos_2nd = int(exon_table[\"end_pos\"][index])\n",
    "                    start_pos = int(exon_table[\"start_pos\"][index])\n",
    "                    end_pos = int(exon_table[\"end_pos\"][index])\n",
    "                \n",
    "                #If the exons ends here: move to the next exon\n",
    "                chrom_pos_3rd = chrom_pos_2nd - 1\n",
    "                if (chrom_pos_3rd < start_pos):\n",
    "                    index += 1\n",
    "                    chrom_pos_3rd = int(exon_table[\"end_pos\"][index])\n",
    "                    \n",
    "            #Calculate position for forward strand\n",
    "            else:\n",
    "                chrom_pos_1st = start_pos + len_from_exon_start\n",
    "                \n",
    "                chrom_pos_2nd = chrom_pos_1st + 1\n",
    "                 #If the exons ends here: move to the next exon\n",
    "                if (chrom_pos_2nd > end_pos):\n",
    "                    index += 1\n",
    "                    chrom_pos_2nd = int(exon_table[\"start_pos\"][index])\n",
    "                    start_pos = int(exon_table[\"start_pos\"][index])\n",
    "                    end_pos = int(exon_table[\"end_pos\"][index])\n",
    "                \n",
    "                #If the exons ends here: move to the next exon\n",
    "                chrom_pos_3rd = chrom_pos_2nd + 1\n",
    "                if (chrom_pos_3rd > end_pos):\n",
    "                    index += 1\n",
    "                    chrom_pos_3rd = int(exon_table[\"start_pos\"][index])\n",
    "            \n",
    "            return (chrom_pos_1st, chrom_pos_2nd, chrom_pos_3rd)\n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 285,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "False"
      ]
     },
     "execution_count": 285,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "is_number(\"-\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "#Boolean function - determine if a given text can be converted to a number\n",
    "def is_number(s):\n",
    "    try:\n",
    "        float(s)\n",
    "        return True\n",
    "    except ValueError:\n",
    "        return False"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 313,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "#A function that return the hmm state of that protein position\n",
    "# return -1 for positions outside of domains regions, -2 for matching insertion\n",
    "#TODO: do we need also transcript id? do we want to consider more than 1 transcript per gene?\n",
    "def protein_pos_to_hmm_state_and_aa(protein_pos, domain_gene_table):\n",
    "    for index, row in domain_gene_table.iterrows():\n",
    "        target_start = row[\"TargetStart\"]\n",
    "        target_end = row[\"TargetEnd\"]\n",
    "        aa = \"-\"\n",
    "        \n",
    "        #Check if the position is inside this domain instance of the gene\n",
    "        if (protein_pos >= target_start and protein_pos <= target_end):\n",
    "            \n",
    "            hmm_pos = (row[\"HMM_Pos\"]).split(\",\")\n",
    "            target_seq = list(row[\"Target_Seq\"])\n",
    "            index_inside_match = (protein_pos - target_start)\n",
    "            \n",
    "            #Correct index_inside_match for previous deletions '-'\n",
    "            for j in range(index_inside_match):\n",
    "                if (target_seq[j] == \"-\"):\n",
    "                    index_inside_match += 1\n",
    "            \n",
    "            #Find the HMM match state\n",
    "            while (aa == \"-\"):\n",
    "                hmm_state_text = hmm_pos[index_inside_match]\n",
    "                if (is_number(hmm_state_text) == True):\n",
    "                    hmm_state = int(hmm_state_text)\n",
    "                else:\n",
    "                    #the position match insertion\n",
    "                    hmm_state = -2\n",
    "\n",
    "                #Find the amino acid\n",
    "                aa = (target_seq[index_inside_match]).upper()\n",
    "            \n",
    "                #Handling a deletion in the domain instance in comparison to the HMM profile\n",
    "                if (aa == \"-\"):\n",
    "                    index_inside_match += 1\n",
    "                \n",
    "            return(hmm_state, aa)\n",
    "            \n",
    "    #The protein position isn't in any domain region        \n",
    "    return (-1,'-')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "#A function that create the new codon for the alteration\n",
    "def create_alt_codon(exac_ref_bp, curr_alt_bp, ref_codon, alt_codon_pos, chrom_raw_data):\n",
    "    \n",
    "    #For error logging\n",
    "    functionNameAsString = sys._getframe().f_code.co_name\n",
    "    \n",
    "    #Complement strand - transversing the bp to base-complement\n",
    "    if (chrom_raw_data.find(\"complement\") >= 0):\n",
    "        new_bp = \"\"\n",
    "        for c in curr_alt_bp:\n",
    "            if (c.upper() == 'A'):\n",
    "                new_bp = new_bp+'T'\n",
    "            elif (c.upper() == 'T'):\n",
    "                new_bp = new_bp+'A'\n",
    "            elif (c.upper() == 'G'):\n",
    "                new_bp = new_bp+'C'\n",
    "            else:\n",
    "                new_bp = new_bp+'G'\n",
    "            new_bp = new_bp[::-1] #TODO: is the reverse needed?\n",
    "        \n",
    "        exac_ref_bp_adj = \"\"\n",
    "        for c in exac_ref_bp:\n",
    "            if (c.upper() == 'A'):\n",
    "                exac_ref_bp_adj = exac_ref_bp_adj+'T'\n",
    "            elif (c.upper() == 'T'):\n",
    "                exac_ref_bp_adj = exac_ref_bp_adj+'A'\n",
    "            elif (c.upper() == 'G'):\n",
    "                exac_ref_bp_adj = exac_ref_bp_adj+'C'\n",
    "            else:\n",
    "                exac_ref_bp_adj = exac_ref_bp_adj+'G'\n",
    "            exac_ref_bp_adj = exac_ref_bp_adj[::-1] #TODO: is the reverse needed?\n",
    "        \n",
    "    #Regular strand\n",
    "    else:\n",
    "        new_bp = curr_alt_bp\n",
    "        exac_ref_bp_adj = exac_ref_bp\n",
    "        \n",
    "    #Validation: making sure the ref bp from ExAC is inside the ref codon sequence retrieved from hg19 or the other way around (at least one contain the other)\n",
    "    if (ref_codon.find(exac_ref_bp_adj) == -1 and exac_ref_bp_adj.find(ref_codon) == -1):\n",
    "        print functionNameAsString+\" Error: ExAC ref sequence \"+exac_ref_bp_adj+\" isn't found in hg19 retrieved codon sequence \"+ref_codon\n",
    "        \n",
    "    new_alt_codon = ref_codon[:alt_codon_pos]+new_bp+ref_codon[alt_codon_pos+len(exac_ref_bp_adj):]\n",
    "    \n",
    "    return new_alt_codon"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "codon_table = {\n",
    "    'ATA':'I', 'ATC':'I', 'ATT':'I', 'ATG':'M',\n",
    "    'ACA':'T', 'ACC':'T', 'ACG':'T', 'ACT':'T',\n",
    "    'AAC':'N', 'AAT':'N', 'AAA':'K', 'AAG':'K',\n",
    "    'AGC':'S', 'AGT':'S', 'AGA':'R', 'AGG':'R',\n",
    "    'CTA':'L', 'CTC':'L', 'CTG':'L', 'CTT':'L',\n",
    "    'CCA':'P', 'CCC':'P', 'CCG':'P', 'CCT':'P',\n",
    "    'CAC':'H', 'CAT':'H', 'CAA':'Q', 'CAG':'Q',\n",
    "    'CGA':'R', 'CGC':'R', 'CGG':'R', 'CGT':'R',\n",
    "    'GTA':'V', 'GTC':'V', 'GTG':'V', 'GTT':'V',\n",
    "    'GCA':'A', 'GCC':'A', 'GCG':'A', 'GCT':'A',\n",
    "    'GAC':'D', 'GAT':'D', 'GAA':'E', 'GAG':'E',\n",
    "    'GGA':'G', 'GGC':'G', 'GGG':'G', 'GGT':'G',\n",
    "    'TCA':'S', 'TCC':'S', 'TCG':'S', 'TCT':'S',\n",
    "    'TTC':'F', 'TTT':'F', 'TTA':'L', 'TTG':'L',\n",
    "    'TAC':'Y', 'TAT':'Y', 'TAA':'*', 'TAG':'*',\n",
    "    'TGC':'C', 'TGT':'C', 'TGA':'*', 'TGG':'W',\n",
    "    }"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "#Retrieve the codon base-parirs from the ref sequence\n",
    "def retrieve_codon_seq(chrom_pos_list, chrom_raw_data, chrom):\n",
    "    chromsome_name = \"chr\"+chrom\n",
    "    chrom_pos_1st = chrom_pos_list[0]\n",
    "    chrom_pos_2nd = chrom_pos_list[1]\n",
    "    chrom_pos_3rd = chrom_pos_list[2]\n",
    "    \n",
    "    if (chrom_pos_1st < chrom_pos_3rd):\n",
    "        chrom_pos_list_adj = chrom_pos_list\n",
    "    else:\n",
    "        chrom_pos_list_adj = reversed(chrom_pos_list) #For reverse strand - order is reveresed\n",
    "        \n",
    "    seq = \"\"\n",
    "    for chrom_pos in chrom_pos_list:\n",
    "        seq_start = chrom_pos - 1\n",
    "        query = !./twoBitToFa hg19.2bit stdout -seq=$chromsome_name -start=$seq_start -end=$chrom_pos\n",
    "        seq = seq+query[1]\n",
    "    \n",
    "    #Complement strand - transversing the bp to base-complement\n",
    "    if (chrom_raw_data.find(\"complement\") >= 0):\n",
    "        complement_seq = []\n",
    "        for c in seq:\n",
    "            if (c.upper() == 'A'):\n",
    "                complement_seq.append('T')\n",
    "            elif (c.upper() == 'T'):\n",
    "                complement_seq.append('A')\n",
    "            elif (c.upper() == 'G'):\n",
    "                complement_seq.append('C')\n",
    "            else:\n",
    "                complement_seq.append('G')\n",
    "        seq = complement_seq[0]+complement_seq[1]+complement_seq[2]\n",
    "        \n",
    "    return seq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "FRAME_SHIFT_INDEL = 1\n",
    "IN_FRAME_INDEL = 2\n",
    "NO_INDEL = 0"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "#A boolean function that check according to ref and alt if it's frameshift indel or not\n",
    "def is_indel(ref, alt, chrom_alter):\n",
    "    #For error logging\n",
    "    functionNameAsString = sys._getframe().f_code.co_name\n",
    "    \n",
    "    if (len(ref) != len(alt)):\n",
    "        if (len(ref) > len(alt)):\n",
    "            indel_len = len(ref) - len(alt)\n",
    "        else: \n",
    "            indel_len = len(ref) - len(alt)\n",
    "        if ((indel_len % 3) > 0):\n",
    "            \n",
    "            #Validation: ExAC identify this as a frameshift too\n",
    "            conseq = chrom_alter[\"conseq\"]\n",
    "            #if (\"frameshift\" not in conseq):\n",
    "             #   print functionNameAsString+\" Error: ExAC doesn't recognize a frameshift indel - \"+str(chrom_alter[\"pos\"])\n",
    "            \n",
    "            #frameshift indel    \n",
    "            return FRAME_SHIFT_INDEL\n",
    "        \n",
    "        #In frame indel\n",
    "        else:\n",
    "            return IN_FRAME_INDEL\n",
    "        \n",
    "    #Not an indel\n",
    "    else:\n",
    "        return NO_INDEL "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 113,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "# A function that returns the indices of the strings differences\n",
    "def diff(a, b):\n",
    "    return [i for i in range(len(a)) if a[i] != b[i]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 218,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "#A function that adds more lines for all the effected chromosomal positions of each indel\n",
    "def table_editing(chrom_gene_table):\n",
    "    i_cnt = 0\n",
    "    d_cnt = 0\n",
    "    indels_table = pd.DataFrame(columns = chrom_gene_table.columns)\n",
    "    indels_table_i = 0\n",
    "    comments_col = []\n",
    "    \n",
    "    for index, line in chrom_gene_table.iterrows():\n",
    "        ref = line[\"ref\"]\n",
    "        alt = line[\"alt\"]\n",
    "        pos = line[\"pos\"]\n",
    "        feature = line[\"feature\"]\n",
    "        conseq = line[\"conseq\"]\n",
    "        prot_pos = line[\"prot_pos\"]\n",
    "        strand = line[\"strand\"]\n",
    "        \n",
    "        #Handling deletion\n",
    "        if (len(ref) > len(alt)):\n",
    "            d_cnt += 1\n",
    "            comments_col.append(\"ignore for this position: d-\"+str(d_cnt))\n",
    "            #Adding to indels table only the inframe ones\n",
    "            if (is_indel(ref, alt, line) == IN_FRAME_INDEL):\n",
    "                init_pos = pos +  len(alt)\n",
    "                deletion_len = len(ref) - len(alt)\n",
    "                for j in range(deletion_len):\n",
    "                    new_line = line.copy(deep=True)\n",
    "                    new_line[\"pos\"] = init_pos + j\n",
    "                    new_line[\"ref\"] = ref[j+1]\n",
    "                    new_line[\"alt\"] = \"-\"\n",
    "                    new_line[\"comments\"] = \"d-\"+str(d_cnt)\n",
    "                    indels_table.loc[indels_table_i] = new_line\n",
    "                    indels_table_i += 1\n",
    "            \n",
    "\n",
    "        #Handling insertion\n",
    "        elif (len(alt) > len(ref)):\n",
    "            i_cnt += 1\n",
    "            comments_col.append(\"ignore for this position: i-\"+str(i_cnt))\n",
    "            \n",
    "            #Adding to indels table only the inframe ones\n",
    "            if (is_indel(ref, alt, line) == IN_FRAME_INDEL):\n",
    "                init_pos = pos + len(ref)\n",
    "                #insertion_len = len(alt) - len(ref)\n",
    "                new_line = line.copy(deep=True)\n",
    "                if (strand == 1):\n",
    "                    new_line[\"pos\"] = init_pos\n",
    "                else:\n",
    "                    new_line[\"pos\"] = init_pos - 1\n",
    "                new_line[\"ref\"] = \"-\"\n",
    "                new_line[\"alt\"] = alt[len(ref):]\n",
    "                new_line[\"comments\"] = \"i-\"+str(i_cnt)\n",
    "                indels_table.loc[indels_table_i] = new_line\n",
    "                indels_table_i += 1\n",
    "            \n",
    "        #Handling mismtach written with redundant bps\n",
    "        elif (len(ref) > 1):\n",
    "            diff_idx = diff(ref, alt)\n",
    "            #A case when only the first bp is the alteration\n",
    "            if (diff_idx == [0]):\n",
    "                #Fix ref and alt fields\n",
    "                chrom_gene_table.set_value(index, \"ref\", ref[0])\n",
    "                chrom_gene_table.set_value(index, \"alt\", alt[0])\n",
    "                \n",
    "                strand = line[\"strand\"]\n",
    "                \n",
    "                #Fix amino_acids field\n",
    "                aa = line[\"amino_acids\"]\n",
    "                #If aa field is not empty\n",
    "                if (aa != \"\"):\n",
    "                    if (strand == 1):\n",
    "                        if (aa.find(\"/\")!= -1):\n",
    "                            new_aa = aa[0] + aa[aa.find(\"/\"):aa.find(\"/\")+2]\n",
    "                        else:\n",
    "                            new_aa = aa[0]\n",
    "                    else:\n",
    "                        if (aa.find(\"/\")!= -1):\n",
    "                            new_aa = aa[aa.find(\"/\")-1:aa.find(\"/\")+1]+aa[-1]\n",
    "                        else:\n",
    "                            new_aa = aa[-1]\n",
    "                    chrom_gene_table.set_value(index, \"amino_acids\", new_aa)\n",
    "                \n",
    "                #Fix prot_pos field\n",
    "                prot_pos = line[\"prot_pos\"]\n",
    "                if (prot_pos != \"\"):\n",
    "                    if (prot_pos.find(\"-\") != -1):\n",
    "                        if (strand == 1):\n",
    "                            chrom_gene_table.set_value(index, \"prot_pos\",  prot_pos[:prot_pos.find(\"-\")])\n",
    "                        else:\n",
    "                            chrom_gene_table.set_value(index, \"prot_pos\",  prot_pos[prot_pos.find(\"-\")+1:])\n",
    "                \n",
    "                comments_col.append(\"removed redundant bps\")\n",
    "            else:\n",
    "                comments_col.append(\"\")\n",
    "                \n",
    "        #No Indel\n",
    "        else:\n",
    "            comments_col.append(\"\")\n",
    "    \n",
    "    chrom_gene_table[\"comments\"] = comments_col\n",
    "    #print \"Number of insertions = \"+str(i_cnt)\n",
    "    #print \"number of deletion = \"+str(d_cnt)\n",
    "    \n",
    "    return indels_table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 181,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def exac_validation_checks(chrom_alter, protein_pos, aa, alt_codon_pos, chrom_pos, bp_ref):\n",
    "    \n",
    "    #For error logging\n",
    "    functionNameAsString = sys._getframe().f_code.co_name\n",
    "    \n",
    "    error_flag = False\n",
    "    \n",
    "    #Validation: the ExAC chromosome position is within a protein\n",
    "    exac_prot_data = True\n",
    "    if (chrom_alter[\"prot_pos\"] == \"\"):\n",
    "        print functionNameAsString+\" Error: ExAC chromosome position \"+str(chrom_pos)+\" doesn't correspond to a protein\"\n",
    "        #We assume it's an error in ExAC and logging alteration anyway.\n",
    "        exac_prot_data = False\n",
    "        \n",
    "    else:\n",
    "        #Validation: the ExAC protein position match the HMMER protein position\n",
    "        exac_prot_pos = chrom_alter[\"prot_pos\"]\n",
    "         #in case there's more than one position listed\n",
    "        if (exac_prot_pos.find(\"-\") != -1):\n",
    "            first_exac_prot_pos = int(exac_prot_pos[:exac_prot_pos.find(\"-\")])\n",
    "            last_exac_prot_pos = int(exac_prot_pos[exac_prot_pos.find(\"-\")+1:])\n",
    "        else:\n",
    "            first_exac_prot_pos = int(exac_prot_pos)\n",
    "            last_exac_prot_pos = first_exac_prot_pos\n",
    "        #Checking of the protein position isn't within the range described by ExAC\n",
    "        if not(first_exac_prot_pos <= protein_pos <= last_exac_prot_pos):\n",
    "            print functionNameAsString+\" \"+ str(chrom_pos)+\" Error: ExAC protein position \"+str(first_exac_prot_pos)+\" doesn't match HMMER protein position \"+str(protein_pos)\n",
    "            error_flag = True\n",
    "\n",
    "        #Validation: the ExAC aa match the HMMER aa\n",
    "        exac_aa = chrom_alter[\"amino_acids\"]\n",
    "        if (exac_aa.find(\"/\") != -1):\n",
    "            exac_ref_aa = exac_aa[:exac_aa.find(\"/\")]\n",
    "        else:\n",
    "            exac_ref_aa = exac_aa\n",
    "        exac_alt_aa = exac_aa[exac_aa.find(\"/\")+1:]\n",
    "        if (exac_ref_aa != aa):\n",
    "            print functionNameAsString+\" \"+ str(chrom_pos)+\" Error: ExAC amino acid identity \"+exac_ref_aa+\" doesn't match HMMER amino-acid \"+aa\n",
    "            error_flag = True\n",
    "\n",
    "        #Extracting aa codon data if exist\n",
    "        exac_codons = chrom_alter[\"codons\"]\n",
    "        exac_ref_codon = exac_codons[:exac_codons.find(\"/\")]\n",
    "        exac_alt_codon = exac_codons[exac_codons.find(\"/\")+1:]\n",
    "        #Validation: the ExAC codon match the returned codon sequence from hg19, or at least one contain the other\n",
    "        if ((exac_ref_codon.upper().find(bp_ref.upper()) == -1) and (bp_ref.upper().find(exac_ref_codon.upper()) == -1)):\n",
    "            print functionNameAsString+\" \"+ str(chrom_pos)+\" Error: ExAC bp codon \"+exac_ref_codon.upper()+\" doesn't match hg19 codon sequence retrieved \"+bp_ref.upper()\n",
    "        \n",
    "        #Getting the changed position inside the ExAC codon - ignore for now\n",
    "        #exac_alt_codon_pos = 0\n",
    "        #for c in exac_ref_codon:\n",
    "            #if c.isupper():\n",
    "                #break\n",
    "            #exac_alt_codon_pos += 1\n",
    "        #Validation: the ExAC alt position match my alt codon position calculation\n",
    "        #if (exac_alt_codon_pos != alt_codon_pos):\n",
    "            #print functionNameAsString+\" \"+ str(chrom_pos)+\" Error: the ExAC alt position in codon \"+str(exac_alt_codon_pos)+\" doesn't match my codon position calculation \"+str(alt_codon_pos)\n",
    "    \n",
    "        return (exac_prot_data, exac_alt_aa, exac_alt_codon, error_flag)\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 279,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "#A function that return a dict with the MAF info for the protein position and corresponding chromosomal location\n",
    "def calc_exac_maf_data(chrom_pos_list, chrom_gene_table, indels_table, protein_pos, aa, chrom_raw_data, chrom):\n",
    "    \n",
    "    res_dict = {}\n",
    "    res_dict[\"chrom\"] = chrom\n",
    "    res_dict[\"chrom_pos\"] = chrom_pos_list\n",
    "    res_dict[\"prot_pos\"] = protein_pos\n",
    "    res_dict[\"aa_ref\"] = aa\n",
    "    res_dict[\"bp_ref\"] = retrieve_codon_seq(chrom_pos_list, chrom_raw_data, chrom)\n",
    "    frameshift_cnt = 0\n",
    "    errors_cnt = 0\n",
    "    inframe_ids = []\n",
    "    \n",
    "    #For error logging\n",
    "    functionNameAsString = sys._getframe().f_code.co_name\n",
    "    \n",
    "    #Validation: checking that the returned codon sequence from hg19 match the HMMER amino-acid\n",
    "    translated_aa = codon_table[(res_dict[\"bp_ref\"]).upper()]\n",
    "    if (translated_aa != aa):\n",
    "        print functionNameAsString+\" Error: hg19 codon sequence retrieved \"+(res_dict[\"bp_ref\"]).upper()+\"=\"+translated_aa+\" doesn't match HMMER amino-acid \"+aa\n",
    "        print chrom_pos_list\n",
    "        print \"prot_pos = \"+str(protein_pos)\n",
    "        print \"chrom_raw_data = \"+chrom_raw_data\n",
    "    \n",
    "    alterations_af_dict = defaultdict(list)\n",
    "    alterations_af_adj_dict = defaultdict(list)\n",
    "    \n",
    "    for i in range(len(chrom_pos_list)):\n",
    "        chrom_pos = chrom_pos_list[i]\n",
    "        alt_codon_pos = i\n",
    "            \n",
    "        #Retreiving relevant ExAC entry\n",
    "        chrom_alter_table = chrom_gene_table[chrom_gene_table[\"pos\"] == chrom_pos]\n",
    "        chrom_alter_table = chrom_alter_table.reset_index(drop=True)\n",
    "                \n",
    "        if (chrom_alter_table.shape[0] == 0):\n",
    "            #No ExAC entry for this chromosome position - not adding alteration data\n",
    "            continue\n",
    "        \n",
    "        else:\n",
    "            #In case there are several alterations for that position, iterating\n",
    "            for index, line in chrom_alter_table.iterrows():\n",
    "                chrom_alter = line\n",
    "                \n",
    "                #Extracting ref and alt\n",
    "                exac_ref_bp = chrom_alter[\"ref\"]\n",
    "                exac_alt_bp = chrom_alter[\"alt\"]\n",
    "                \n",
    "                #Check if frameshift indel - skip (we assume the whole protein may not function and don't add those to the MAF count)\n",
    "                if (is_indel(exac_ref_bp, exac_alt_bp, chrom_alter) == FRAME_SHIFT_INDEL):\n",
    "                    frameshift_cnt += 1\n",
    "                    continue\n",
    "                    \n",
    "                #Check if inframe indel - take data from the indels table later\n",
    "                elif (is_indel(exac_ref_bp, exac_alt_bp, chrom_alter) == IN_FRAME_INDEL):\n",
    "                    indel_id = chrom_alter[\"comments\"][chrom_alter[\"comments\"].find(\"-\"):]\n",
    "                    inframe_ids.append(indel_id)\n",
    "                    continue     \n",
    "                \n",
    "                #Perform validation checks (comparing ExAC and HMMER data)\n",
    "                (exac_prot_data, exac_alt_aa, exac_alt_codon, errors) = exac_validation_checks(chrom_alter, protein_pos, aa, alt_codon_pos, chrom_pos, res_dict[\"bp_ref\"])\n",
    "                if (errors):\n",
    "                    errors_cnt += 1\n",
    "                    #Skipping if there are validation errors\n",
    "                    continue\n",
    "\n",
    "                #Extracting ExAC allele frequency data\n",
    "                af = chrom_alter[\"AF\"]\n",
    "                an_adj = int(chrom_alter[\"AN_Adj\"])\n",
    "                ac_adj = chrom_alter[\"AC_Adj\"]\n",
    "                \n",
    "                #Calculating the alteration relevant data\n",
    "                alt_codon = create_alt_codon(exac_ref_bp, exac_alt_bp, res_dict[\"bp_ref\"], alt_codon_pos, chrom_raw_data)\n",
    "                if (len(alt_codon) != 3):\n",
    "                    alt_aa = \"indel\"\n",
    "                    print \"indel!!!? \"+str(chrom_pos)\n",
    "                    continue #TODO: handle inframe indels\n",
    "                else:\n",
    "                    alt_aa = codon_table[alt_codon.upper()]\n",
    "\n",
    "                #Validation: ExAC alt codon and aa for the first alteration match the calculated alt\n",
    "                #if (exac_prot_data and exac_alt_codon.upper() != alt_codon):\n",
    "                    #print functionNameAsString+\" \"+ str(chrom_pos)+\" Error: the ExAC alt codon \"+exac_alt_codon.upper()+\" doesn't match my alt codon calculation \"+alt_codon\n",
    "                if (exac_prot_data and exac_alt_aa != alt_aa):\n",
    "                    print functionNameAsString+\" \"+ str(chrom_pos)+\" Error: the ExAC alt aa \"+exac_alt_aa+\" doesn't match my alt aa calculation \"+alt_aa\n",
    "\n",
    "                if (alt_aa == res_dict[\"aa_ref\"]):\n",
    "                    #Not logging alteration for synonymous mutations\n",
    "                    continue\n",
    "\n",
    "                #Non-synonymous(!!!) - logging the alteration\n",
    "                else:\n",
    "                    alterations_af_dict[alt_aa].append(float(af))\n",
    "                    af_adj = float(ac_adj)/float(an_adj)\n",
    "                    af_adj_format = float('{:.3e}'.format(float(af_adj)))\n",
    "                    alterations_af_adj_dict[alt_aa].append(af_adj_format)\n",
    "\n",
    "    #Calculating the overall MAF from the alteration dicts\n",
    "    res_dict[\"af\"] = 0\n",
    "    res_dict[\"af_adj\"] = 0\n",
    "    \n",
    "    #Checking if any alteration is above 0.5, and changing the ref accordingly\n",
    "    for aa in alterations_af_dict.keys():\n",
    "        aa_sum = sum(alterations_af_dict[aa])\n",
    "        aa_adj_sum = sum(alterations_af_adj_dict[aa])\n",
    "        if (aa != \"indel\" and aa_sum > 0.5):\n",
    "            res_dict[\"aa_ref\"] = aa\n",
    "            res_dict[\"af\"] =(1 - aa_sum)\n",
    "            res_dict[\"af_adj\"] = (1 - aa_adj_sum)\n",
    "            break\n",
    "        else:\n",
    "            res_dict[\"af\"] += aa_sum\n",
    "            res_dict[\"af_adj\"] += aa_adj_sum\n",
    "        \n",
    "        #Fix the AF format\n",
    "        res_dict[\"af\"] = float('{:.3e}'.format(float(res_dict[\"af\"])))\n",
    "        res_dict[\"af_adj\"] = float('{:.3e}'.format(float(res_dict[\"af_adj\"])))\n",
    "    \n",
    "    return (res_dict, frameshift_cnt, errors_cnt)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 281,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def correct_zinc_domain_data(sorted_domain_data):\n",
    "    sorted_domain_data.set_value(8877, \"TargetStart\", 442)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 282,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "correct_zinc_domain_data(sorted_domain_data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 276,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "441"
      ]
     },
     "execution_count": 276,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sorted_domain_data.iloc[8877][\"TargetStart\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 315,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Finished chromosome 1\n",
      "Finished chromosome 2\n",
      "Finished chromosome 3\n",
      "Finished chromosome 4\n",
      "Finished chromosome 5\n",
      "Finished chromosome 6\n",
      "Finished chromosome 7\n",
      "Finished chromosome 8\n",
      "Finished chromosome 9\n",
      "Finished chromosome 10\n",
      "Finished chromosome 11\n",
      "calc_exac_maf_data Error: hg19 codon sequence retrieved GTT=V doesn't match HMMER amino-acid F\n",
      "(57860061, 57860062, 57860063)\n",
      "prot_pos = 140\n",
      "chrom_raw_data = GRCh37.75:12:join(-527..-527,57858647..57858651,57858894..57859038,57859390..57859479,57859571..57859708,57860023..57860170)\n",
      "calc_exac_maf_data Error: hg19 codon sequence retrieved CGT=R doesn't match HMMER amino-acid V\n",
      "(57860064, 57860065, 57860066)\n",
      "prot_pos = 141\n",
      "chrom_raw_data = GRCh37.75:12:join(-527..-527,57858647..57858651,57858894..57859038,57859390..57859479,57859571..57859708,57860023..57860170)\n",
      "exac_validation_checks 57860064 Error: ExAC protein position 227 doesn't match HMMER protein position 141\n",
      "exac_validation_checks 57860064 Error: ExAC amino acid identity F doesn't match HMMER amino-acid V\n",
      "exac_validation_checks 57860064 Error: ExAC bp codon TTC doesn't match hg19 codon sequence retrieved CGT\n",
      "exac_validation_checks 57860065 Error: ExAC protein position 228 doesn't match HMMER protein position 141\n",
      "exac_validation_checks 57860065 Error: ExAC bp codon GTG doesn't match hg19 codon sequence retrieved CGT\n",
      "calc_exac_maf_data Error: hg19 codon sequence retrieved GTG=V doesn't match HMMER amino-acid C\n",
      "(57860067, 57860068, 57860069)\n",
      "prot_pos = 142\n",
      "chrom_raw_data = GRCh37.75:12:join(-527..-527,57858647..57858651,57858894..57859038,57859390..57859479,57859571..57859708,57860023..57860170)\n",
      "calc_exac_maf_data Error: hg19 codon sequence retrieved CCA=P doesn't match HMMER amino-acid H\n",
      "(57860070, 57860071, 57860072)\n",
      "prot_pos = 143\n",
      "chrom_raw_data = GRCh37.75:12:join(-527..-527,57858647..57858651,57858894..57859038,57859390..57859479,57859571..57859708,57860023..57860170)\n",
      "exac_validation_checks 57860080 Error: ExAC protein position 233 doesn't match HMMER protein position 146\n",
      "exac_validation_checks 57860080 Error: ExAC bp codon GGC doesn't match hg19 codon sequence retrieved GGG\n",
      "exac_validation_checks 57860080 Error: ExAC protein position 233 doesn't match HMMER protein position 146\n",
      "exac_validation_checks 57860080 Error: ExAC bp codon GGC doesn't match hg19 codon sequence retrieved GGG\n",
      "exac_validation_checks 57860080 Error: ExAC protein position 233 doesn't match HMMER protein position 146\n",
      "exac_validation_checks 57860080 Error: ExAC bp codon GGC doesn't match hg19 codon sequence retrieved GGG\n",
      "calc_exac_maf_data Error: hg19 codon sequence retrieved CTG=L doesn't match HMMER amino-acid C\n",
      "(57860082, 57860083, 57860084)\n",
      "prot_pos = 147\n",
      "chrom_raw_data = GRCh37.75:12:join(-527..-527,57858647..57858651,57858894..57859038,57859390..57859479,57859571..57859708,57860023..57860170)\n",
      "exac_validation_checks 57860084 Error: ExAC protein position 234 doesn't match HMMER protein position 147\n",
      "exac_validation_checks 57860084 Error: ExAC bp codon TGC doesn't match hg19 codon sequence retrieved CTG\n",
      "calc_exac_maf_data Error: hg19 codon sequence retrieved CTC=L doesn't match HMMER amino-acid S\n",
      "(57860085, 57860086, 57860087)\n",
      "prot_pos = 148\n",
      "chrom_raw_data = GRCh37.75:12:join(-527..-527,57858647..57858651,57858894..57859038,57859390..57859479,57859571..57859708,57860023..57860170)\n",
      "exac_validation_checks 57860085 Error: ExAC protein position 234 doesn't match HMMER protein position 148\n",
      "exac_validation_checks 57860085 Error: ExAC amino acid identity C doesn't match HMMER amino-acid S\n",
      "exac_validation_checks 57860085 Error: ExAC bp codon TGC doesn't match hg19 codon sequence retrieved CTC\n",
      "calc_exac_maf_data Error: hg19 codon sequence retrieved GAG=E doesn't match HMMER amino-acid R\n",
      "(57860097, 57860098, 57860099)\n",
      "prot_pos = 152\n",
      "chrom_raw_data = GRCh37.75:12:join(-527..-527,57858647..57858651,57858894..57859038,57859390..57859479,57859571..57859708,57860023..57860170)\n",
      "calc_exac_maf_data Error: hg19 codon sequence retrieved GCC=A doesn't match HMMER amino-acid P\n",
      "(57860100, 57860101, 57860102)\n",
      "prot_pos = 153\n",
      "chrom_raw_data = GRCh37.75:12:join(-527..-527,57858647..57858651,57858894..57859038,57859390..57859479,57859571..57859708,57860023..57860170)\n",
      "calc_exac_maf_data Error: hg19 codon sequence retrieved CTT=L doesn't match HMMER amino-acid F\n",
      "(57860103, 57860104, 57860105)\n",
      "prot_pos = 154\n",
      "chrom_raw_data = GRCh37.75:12:join(-527..-527,57858647..57858651,57858894..57859038,57859390..57859479,57859571..57859708,57860023..57860170)\n",
      "calc_exac_maf_data Error: hg19 codon sequence retrieved CAA=Q doesn't match HMMER amino-acid K\n",
      "(57860106, 57860107, 57860108)\n",
      "prot_pos = 155\n",
      "chrom_raw_data = GRCh37.75:12:join(-527..-527,57858647..57858651,57858894..57859038,57859390..57859479,57859571..57859708,57860023..57860170)\n",
      "calc_exac_maf_data Error: hg19 codon sequence retrieved AGC=S doesn't match HMMER amino-acid A\n",
      "(57860109, 57860110, 57860111)\n",
      "prot_pos = 156\n",
      "chrom_raw_data = GRCh37.75:12:join(-527..-527,57858647..57858651,57858894..57859038,57859390..57859479,57859571..57859708,57860023..57860170)\n",
      "calc_exac_maf_data Error: hg19 codon sequence retrieved CCA=P doesn't match HMMER amino-acid Q\n",
      "(57860112, 57860113, 57860114)\n",
      "prot_pos = 157\n",
      "chrom_raw_data = GRCh37.75:12:join(-527..-527,57858647..57858651,57858894..57859038,57859390..57859479,57859571..57859708,57860023..57860170)\n",
      "calc_exac_maf_data Error: hg19 codon sequence retrieved GTA=V doesn't match HMMER amino-acid Y\n",
      "(57860115, 57860116, 57860117)\n",
      "prot_pos = 158\n",
      "chrom_raw_data = GRCh37.75:12:join(-527..-527,57858647..57858651,57858894..57859038,57859390..57859479,57859571..57859708,57860023..57860170)\n",
      "calc_exac_maf_data Error: hg19 codon sequence retrieved CAT=H doesn't match HMMER amino-acid M\n",
      "(57860118, 57860119, 57860120)\n",
      "prot_pos = 159\n",
      "chrom_raw_data = GRCh37.75:12:join(-527..-527,57858647..57858651,57858894..57859038,57859390..57859479,57859571..57859708,57860023..57860170)\n",
      "calc_exac_maf_data Error: hg19 codon sequence retrieved GCT=A doesn't match HMMER amino-acid L\n",
      "(57860121, 57860122, 57860123)\n",
      "prot_pos = 160\n",
      "chrom_raw_data = GRCh37.75:12:join(-527..-527,57858647..57858651,57858894..57859038,57859390..57859479,57859571..57859708,57860023..57860170)\n",
      "calc_exac_maf_data Error: hg19 codon sequence retrieved GGT=G doesn't match HMMER amino-acid V\n",
      "(57860124, 57860125, 57860126)\n",
      "prot_pos = 161\n",
      "chrom_raw_data = GRCh37.75:12:join(-527..-527,57858647..57858651,57858894..57859038,57859390..57859479,57859571..57859708,57860023..57860170)\n",
      "exac_validation_checks 57860125 Error: ExAC protein position 248 doesn't match HMMER protein position 161\n",
      "exac_validation_checks 57860125 Error: ExAC bp codon GTG doesn't match hg19 codon sequence retrieved GGT\n",
      "calc_exac_maf_data Error: hg19 codon sequence retrieved GGT=G doesn't match HMMER amino-acid V\n",
      "(57860127, 57860128, 57860129)\n",
      "prot_pos = 162\n",
      "chrom_raw_data = GRCh37.75:12:join(-527..-527,57858647..57858651,57858894..57859038,57859390..57859479,57859571..57859708,57860023..57860170)\n",
      "calc_exac_maf_data Error: hg19 codon sequence retrieved TCA=S doesn't match HMMER amino-acid H\n",
      "(57860130, 57860131, 57860132)\n",
      "prot_pos = 163\n",
      "chrom_raw_data = GRCh37.75:12:join(-527..-527,57858647..57858651,57858894..57859038,57859390..57859479,57859571..57859708,57860023..57860170)\n",
      "calc_exac_maf_data Error: hg19 codon sequence retrieved CAT=H doesn't match HMMER amino-acid M\n",
      "(57860133, 57860134, 57860135)\n",
      "prot_pos = 164\n",
      "chrom_raw_data = GRCh37.75:12:join(-527..-527,57858647..57858651,57858894..57859038,57859390..57859479,57859571..57859708,57860023..57860170)\n",
      "exac_validation_checks 57860133 Error: ExAC protein position 250 doesn't match HMMER protein position 164\n",
      "exac_validation_checks 57860133 Error: ExAC amino acid identity H doesn't match HMMER amino-acid M\n",
      "exac_validation_checks 57860133 Error: ExAC bp codon CAC doesn't match hg19 codon sequence retrieved CAT\n",
      "calc_exac_maf_data Error: hg19 codon sequence retrieved GCG=A doesn't match HMMER amino-acid R\n",
      "(57860136, 57860137, 57860138)\n",
      "prot_pos = 165\n",
      "chrom_raw_data = GRCh37.75:12:join(-527..-527,57858647..57858651,57858894..57859038,57859390..57859479,57859571..57859708,57860023..57860170)\n",
      "calc_exac_maf_data Error: hg19 codon sequence retrieved CAG=Q doesn't match HMMER amino-acid R\n",
      "(57860139, 57860140, 57860141)\n",
      "prot_pos = 166\n",
      "chrom_raw_data = GRCh37.75:12:join(-527..-527,57858647..57858651,57858894..57859038,57859390..57859479,57859571..57859708,57860023..57860170)\n",
      "calc_exac_maf_data Error: hg19 codon sequence retrieved ACA=T doesn't match HMMER amino-acid H\n",
      "(57860142, 57860143, 57860144)\n",
      "prot_pos = 167\n",
      "chrom_raw_data = GRCh37.75:12:join(-527..-527,57858647..57858651,57858894..57859038,57859390..57859479,57859571..57859708,57860023..57860170)\n",
      "Finished chromosome 12\n",
      "Finished chromosome 13\n",
      "Finished chromosome 14\n",
      "Finished chromosome 15\n",
      "Finished chromosome 16\n",
      "Finished chromosome 17\n",
      "Finished chromosome 18\n",
      "create_alt_codon Error: ExAC ref sequence G isn't found in hg19 retrieved codon sequence CAg\n",
      "Finished chromosome 19\n",
      "Finished chromosome 20\n",
      "Finished chromosome 21\n",
      "Finished chromosome 22\n",
      "Finished chromosome X\n",
      "Finished chromosome Y\n"
     ]
    }
   ],
   "source": [
    "chrom_path = curr_dir[0]+\"/parsed/\"\n",
    "chrom_filename = \"parsed_chrom\"\n",
    "states_dict = defaultdict(list)\n",
    "\n",
    "#For error logging\n",
    "functionNameAsString = sys._getframe().f_code.co_name\n",
    "\n",
    "#A list to count frameshifts per gene\n",
    "domain_ens_genes_frameshifts = []\n",
    "\n",
    "#A list to count validation errors per gene\n",
    "domain_ens_genes_errors = []\n",
    "\n",
    "for chrom in chromosome_names:\n",
    "    \n",
    "    #Filtering the domain data relevant to this chromosome\n",
    "    domain_chrom_data = sorted_domain_data[sorted_domain_data[\"chrom_num\"] == chrom]\n",
    "    \n",
    "    #Loading the ExAC parsed data of this chromosome\n",
    "    fields = ['chrom', 'pos', 'ref', 'alt', 'AC', 'AC_Adj', 'AF', 'AN', 'AN_Adj', 'gene', 'feature', \n",
    "              'feature_type', 'conseq', 'prot_pos', 'amino_acids', 'codons', 'strand', 'ENSP', 'exon', \n",
    "              'intron', 'domains']\n",
    "    chrom_csv = pd.read_csv(chrom_path+chrom_filename+chrom+\".csv\", sep='\\t', index_col=0, usecols=fields)\n",
    "    chrom_csv = chrom_csv.sort_values(by=[\"pos\"])\n",
    "    chrom_csv = chrom_csv.reset_index(drop=True)\n",
    "    chrom_csv.fillna('', inplace=True)\n",
    "    chrom_csv[\"comments\"] = \"\"\n",
    "    \n",
    "    #Getting a list of all the relevant ensembl gene ids for this chromosome\n",
    "    domain_ens_genes = (domain_chrom_data[\"ensembl_id\"]).unique()\n",
    "    \n",
    "    #For each ensembl gene in the domain data - finding all the ExAC alterations\n",
    "    for ens_gene in domain_ens_genes:\n",
    "        \n",
    "        #Filtering the domain data for this gene according to the canonical protein id\n",
    "        canonic_prot = canonic_protein[ens_gene]\n",
    "        domain_gene_table = domain_chrom_data[domain_chrom_data[\"prot_id\"] == canonic_prot]\n",
    "        #Making sure that if two HMM-matches overlaps, the higher bit score will come first in the table\n",
    "        domain_gene_table = domain_gene_table.sort_values(by=\"BitScore\", ascending=False)\n",
    "        \n",
    "        #Getting the chosen protein transcript id for the ExAC filtering\n",
    "        canonic_transcript_id = domain_gene_table[\"transcript_id\"].tolist()[0]\n",
    "        \n",
    "        #Creating a table of the exons for this gene, according to the canonical protein\n",
    "        chrom_raw_data = domain_gene_table[\"chromosome_id\"].unique()[0] #there should be only one element here\n",
    "        if (len(domain_gene_table[\"chromosome_id\"].unique()) > 1):\n",
    "            print functionNameAsString+\" Error: \"+ens_gene+\": more than one chromosome raw data\" #sanity check\n",
    "        exon_table = create_exon_pos_table(chrom_raw_data)\n",
    "        \n",
    "        #Filtering the chromosome data to the gene exons region\n",
    "        exons_start_pos = min(exon_table[\"start_pos\"][0],exon_table[\"start_pos\"][len(exon_table)-1]) #in case of complelemt, the minimal position could be at the last row\n",
    "        exons_end_pos = max(exon_table[\"end_pos\"][0],exon_table[\"end_pos\"][len(exon_table)-1]) #in case of complelemt, the maximal position could be at the first row\n",
    "        chrom_gene_table = chrom_csv[chrom_csv[\"pos\"] >= int(exons_start_pos)][chrom_csv[\"pos\"] <= int(exons_end_pos)][chrom_csv[\"feature\"] == canonic_transcript_id]\n",
    "        chrom_gene_table = chrom_gene_table.reset_index(drop=True)\n",
    "        \n",
    "        #Handling indels\n",
    "        indels_table = table_editing(chrom_gene_table)\n",
    "        \n",
    "        #A counter for frameshifts inside the domain\n",
    "        protein_frameshifts_cnt = 0\n",
    "        #A counter for validation errors inside the domain\n",
    "        protein_errors_cnt = 0\n",
    "        \n",
    "        #Iterating over the amino-acids of the protein\n",
    "        prot_len = domain_gene_table[\"length\"].unique()[0]\n",
    "        for protein_pos in range(1,prot_len+1):\n",
    "    \n",
    "            #Trying to match HMM-state, and retreive the aa from HMMER results\n",
    "            (hmm_state, aa) = protein_pos_to_hmm_state_and_aa(protein_pos, domain_gene_table) #TODO: what happens when two matches overlap? maybe sort to the best bit score?\n",
    "                \n",
    "            #If there's a match to HMM-state: find the corresponding codon bps chromosome positions\n",
    "            if (hmm_state > 0):\n",
    "                chrom_pos_list =find_chrom_bps(protein_pos, exon_table, chrom_raw_data)\n",
    "                \n",
    "                #Analysis of the amino-acid MAF and realted data, returned in a dictionary\n",
    "                (info_dict, frameshift_cnt, errors_cnt) = calc_exac_maf_data(chrom_pos_list, chrom_gene_table, indels_table, protein_pos, aa, chrom_raw_data, chrom)\n",
    "                \n",
    "                #Adding the dictionary to the HMM-state list\n",
    "                states_dict[hmm_state].append(info_dict)\n",
    "                \n",
    "                #Adding the frameshifts to the global counter\n",
    "                protein_frameshifts_cnt += frameshift_cnt\n",
    "                \n",
    "                #Adding the errors to the global counter\n",
    "                protein_errors_cnt += errors_cnt\n",
    "        \n",
    "        domain_ens_genes_frameshifts.append(protein_frameshifts_cnt)\n",
    "        domain_ens_genes_errors.append(protein_errors_cnt)\n",
    "                                \n",
    "    print \"Finished chromosome \"+chrom\n",
    "\n",
    "with open(my_path+'zinc_hmm_states_dict.pik', 'wb') as handle:\n",
    "    pickle.dump(states_dict, handle, protocol=pickle.HIGHEST_PROTOCOL)\n",
    "\n",
    "#with open(my_path+'homeobox_hmm_states_dict.pik', 'wb') as handle:\n",
    "#    pickle.dump(states_dict, handle, protocol=pickle.HIGHEST_PROTOCOL)    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 322,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "2038"
      ]
     },
     "execution_count": 322,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sum(domain_ens_genes_frameshifts)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 287,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "chrom = \"6\"\n",
    "\n",
    "#For error logging\n",
    "functionNameAsString = sys._getframe().f_code.co_name\n",
    "\n",
    "#A list to count frameshifts per gene\n",
    "domain_ens_genes_frameshifts = []\n",
    "\n",
    "#A list to count validation errors per gene\n",
    "domain_ens_genes_errors = []\n",
    "\n",
    "#Filtering the domain data relevant to this chromosome\n",
    "domain_chrom_data = sorted_domain_data[sorted_domain_data[\"chrom_num\"] == chrom]\n",
    "\n",
    "#Loading the ExAC parsed data of this chromosome\n",
    "fields = ['chrom', 'pos', 'ref', 'alt', 'AC', 'AC_Adj', 'AF', 'AN', 'AN_Adj', 'gene', 'feature', \n",
    "          'feature_type', 'conseq', 'prot_pos', 'amino_acids', 'codons', 'strand', 'ENSP', 'exon', \n",
    "          'intron', 'domains']\n",
    "chrom_csv = pd.read_csv(chrom_path+chrom_filename+chrom+\".csv\", sep='\\t', index_col=0, usecols=fields)\n",
    "chrom_csv = chrom_csv.sort_values(by=[\"pos\"])\n",
    "chrom_csv = chrom_csv.reset_index(drop=True)\n",
    "chrom_csv.fillna('', inplace=True)\n",
    "chrom_csv[\"comments\"] = \"\"\n",
    "\n",
    "#Getting a list of all the relevant ensembl gene ids for this chromosome\n",
    "domain_ens_genes = (domain_chrom_data[\"ensembl_id\"]).unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 312,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "**protein_pos = 1\n",
      "**protein_pos = 2\n",
      "**protein_pos = 3\n",
      "**protein_pos = 4\n",
      "**protein_pos = 5\n",
      "**protein_pos = 6\n",
      "**protein_pos = 7\n",
      "**protein_pos = 8\n",
      "**protein_pos = 9\n",
      "**protein_pos = 10\n",
      "**protein_pos = 11\n",
      "**protein_pos = 12\n",
      "**protein_pos = 13\n",
      "**protein_pos = 14\n",
      "**protein_pos = 15\n",
      "**protein_pos = 16\n",
      "**protein_pos = 17\n",
      "**protein_pos = 18\n",
      "**protein_pos = 19\n",
      "**protein_pos = 20\n",
      "**protein_pos = 21\n",
      "**protein_pos = 22\n",
      "**protein_pos = 23\n",
      "**protein_pos = 24\n",
      "**protein_pos = 25\n",
      "**protein_pos = 26\n",
      "**protein_pos = 27\n",
      "**protein_pos = 28\n",
      "**protein_pos = 29\n",
      "**protein_pos = 30\n",
      "**protein_pos = 31\n",
      "**protein_pos = 32\n",
      "**protein_pos = 33\n",
      "**protein_pos = 34\n",
      "**protein_pos = 35\n",
      "**protein_pos = 36\n",
      "**protein_pos = 37\n",
      "**protein_pos = 38\n",
      "**protein_pos = 39\n",
      "**protein_pos = 40\n",
      "**protein_pos = 41\n",
      "**protein_pos = 42\n",
      "**protein_pos = 43\n",
      "**protein_pos = 44\n",
      "**protein_pos = 45\n",
      "**protein_pos = 46\n",
      "**protein_pos = 47\n",
      "**protein_pos = 48\n",
      "**protein_pos = 49\n",
      "**protein_pos = 50\n",
      "**protein_pos = 51\n",
      "**protein_pos = 52\n",
      "**protein_pos = 53\n",
      "**protein_pos = 54\n",
      "**protein_pos = 55\n",
      "**protein_pos = 56\n",
      "**protein_pos = 57\n",
      "**protein_pos = 58\n",
      "**protein_pos = 59\n",
      "**protein_pos = 60\n",
      "**protein_pos = 61\n",
      "**protein_pos = 62\n",
      "**protein_pos = 63\n",
      "**protein_pos = 64\n",
      "**protein_pos = 65\n",
      "**protein_pos = 66\n",
      "**protein_pos = 67\n",
      "**protein_pos = 68\n",
      "**protein_pos = 69\n",
      "**protein_pos = 70\n",
      "**protein_pos = 71\n",
      "**protein_pos = 72\n",
      "**protein_pos = 73\n",
      "**protein_pos = 74\n",
      "**protein_pos = 75\n",
      "**protein_pos = 76\n",
      "**protein_pos = 77\n",
      "**protein_pos = 78\n",
      "**protein_pos = 79\n",
      "**protein_pos = 80\n",
      "**protein_pos = 81\n",
      "**protein_pos = 82\n",
      "**protein_pos = 83\n",
      "**protein_pos = 84\n",
      "**protein_pos = 85\n",
      "**protein_pos = 86\n",
      "**protein_pos = 87\n",
      "**protein_pos = 88\n",
      "**protein_pos = 89\n",
      "**protein_pos = 90\n",
      "**protein_pos = 91\n",
      "**protein_pos = 92\n",
      "**protein_pos = 93\n",
      "**protein_pos = 94\n",
      "**protein_pos = 95\n",
      "**protein_pos = 96\n",
      "**protein_pos = 97\n",
      "**protein_pos = 98\n",
      "**protein_pos = 99\n",
      "**protein_pos = 100\n",
      "**protein_pos = 101\n",
      "**protein_pos = 102\n",
      "**protein_pos = 103\n",
      "**protein_pos = 104\n",
      "**protein_pos = 105\n",
      "**protein_pos = 106\n",
      "**protein_pos = 107\n",
      "**protein_pos = 108\n",
      "**protein_pos = 109\n",
      "**protein_pos = 110\n",
      "**protein_pos = 111\n",
      "**protein_pos = 112\n",
      "**protein_pos = 113\n",
      "**protein_pos = 114\n",
      "**protein_pos = 115\n",
      "**protein_pos = 116\n",
      "**protein_pos = 117\n",
      "**protein_pos = 118\n",
      "**protein_pos = 119\n",
      "**protein_pos = 120\n",
      "**protein_pos = 121\n",
      "**protein_pos = 122\n",
      "**protein_pos = 123\n",
      "**protein_pos = 124\n",
      "**protein_pos = 125\n",
      "**protein_pos = 126\n",
      "**protein_pos = 127\n",
      "**protein_pos = 128\n",
      "**protein_pos = 129\n",
      "**protein_pos = 130\n",
      "**protein_pos = 131\n",
      "**protein_pos = 132\n",
      "**protein_pos = 133\n",
      "**protein_pos = 134\n",
      "**protein_pos = 135\n",
      "**protein_pos = 136\n",
      "**protein_pos = 137\n",
      "**protein_pos = 138\n",
      "**protein_pos = 139\n",
      "**protein_pos = 140\n",
      "**protein_pos = 141\n",
      "**protein_pos = 142\n",
      "**protein_pos = 143\n",
      "**protein_pos = 144\n",
      "**protein_pos = 145\n",
      "**protein_pos = 146\n",
      "**protein_pos = 147\n",
      "**protein_pos = 148\n",
      "**protein_pos = 149\n",
      "**protein_pos = 150\n",
      "**protein_pos = 151\n",
      "**protein_pos = 152\n",
      "**protein_pos = 153\n",
      "**protein_pos = 154\n",
      "**protein_pos = 155\n",
      "**protein_pos = 156\n",
      "**protein_pos = 157\n",
      "**protein_pos = 158\n",
      "**protein_pos = 159\n",
      "**protein_pos = 160\n",
      "**protein_pos = 161\n",
      "**protein_pos = 162\n",
      "**protein_pos = 163\n",
      "**protein_pos = 164\n",
      "**protein_pos = 165\n",
      "**protein_pos = 166\n",
      "**protein_pos = 167\n",
      "**protein_pos = 168\n",
      "**protein_pos = 169\n",
      "**protein_pos = 170\n",
      "**protein_pos = 171\n",
      "**protein_pos = 172\n",
      "**protein_pos = 173\n",
      "**protein_pos = 174\n",
      "**protein_pos = 175\n",
      "**protein_pos = 176\n",
      "**protein_pos = 177\n",
      "**protein_pos = 178\n",
      "**protein_pos = 179\n",
      "**protein_pos = 180\n",
      "**protein_pos = 181\n",
      "**protein_pos = 182\n",
      "**protein_pos = 183\n",
      "**protein_pos = 184\n",
      "**protein_pos = 185\n",
      "**protein_pos = 186\n",
      "**protein_pos = 187\n",
      "**protein_pos = 188\n",
      "**protein_pos = 189\n",
      "**protein_pos = 190\n",
      "**protein_pos = 191\n",
      "**protein_pos = 192\n",
      "**protein_pos = 193\n",
      "**protein_pos = 194\n",
      "**protein_pos = 195\n",
      "**protein_pos = 196\n",
      "**protein_pos = 197\n",
      "**protein_pos = 198\n",
      "**protein_pos = 199\n",
      "**protein_pos = 200\n",
      "**protein_pos = 201\n",
      "**protein_pos = 202\n",
      "**protein_pos = 203\n",
      "**protein_pos = 204\n",
      "**protein_pos = 205\n",
      "**protein_pos = 206\n",
      "**protein_pos = 207\n",
      "**protein_pos = 208\n",
      "**protein_pos = 209\n",
      "**protein_pos = 210\n",
      "**protein_pos = 211\n",
      "**protein_pos = 212\n",
      "**protein_pos = 213\n",
      "**protein_pos = 214\n",
      "**protein_pos = 215\n",
      "**protein_pos = 216\n",
      "**protein_pos = 217\n",
      "**protein_pos = 218\n",
      "**protein_pos = 219\n",
      "**protein_pos = 220\n",
      "**protein_pos = 221\n",
      "**protein_pos = 222\n",
      "**protein_pos = 223\n",
      "**protein_pos = 224\n",
      "**protein_pos = 225\n",
      "**protein_pos = 226\n",
      "**protein_pos = 227\n",
      "**protein_pos = 228\n",
      "**protein_pos = 229\n",
      "**protein_pos = 230\n",
      "**protein_pos = 231\n",
      "**protein_pos = 232\n",
      "**protein_pos = 233\n",
      "**protein_pos = 234\n",
      "**protein_pos = 235\n",
      "**protein_pos = 236\n",
      "**protein_pos = 237\n",
      "**protein_pos = 238\n",
      "**protein_pos = 239\n",
      "**protein_pos = 240\n",
      "**protein_pos = 241\n",
      "**protein_pos = 242\n",
      "**protein_pos = 243\n",
      "**protein_pos = 244\n",
      "**protein_pos = 245\n",
      "**protein_pos = 246\n",
      "**protein_pos = 247\n",
      "**protein_pos = 248\n",
      "**protein_pos = 249\n",
      "**protein_pos = 250\n",
      "**protein_pos = 251\n",
      "**protein_pos = 252\n",
      "**protein_pos = 253\n",
      "**protein_pos = 254\n",
      "**protein_pos = 255\n",
      "**protein_pos = 256\n",
      "**protein_pos = 257\n",
      "**protein_pos = 258\n",
      "**protein_pos = 259\n",
      "**protein_pos = 260\n",
      "**protein_pos = 261\n",
      "**protein_pos = 262\n",
      "**protein_pos = 263\n",
      "**protein_pos = 264\n",
      "**protein_pos = 265\n",
      "**protein_pos = 266\n",
      "**protein_pos = 267\n",
      "**protein_pos = 268\n",
      "**protein_pos = 269\n",
      "**protein_pos = 270\n",
      "**protein_pos = 271\n",
      "**protein_pos = 272\n",
      "**protein_pos = 273\n",
      "**protein_pos = 274\n",
      "**protein_pos = 275\n",
      "**protein_pos = 276\n",
      "**protein_pos = 277\n",
      "**protein_pos = 278\n",
      "**protein_pos = 279\n",
      "**protein_pos = 280\n",
      "**protein_pos = 281\n",
      "**protein_pos = 282\n",
      "**protein_pos = 283\n",
      "**protein_pos = 284\n",
      "**protein_pos = 285\n",
      "**protein_pos = 286\n",
      "**protein_pos = 287\n",
      "**protein_pos = 288\n",
      "**protein_pos = 289\n",
      "**protein_pos = 290\n",
      "**protein_pos = 291\n",
      "**protein_pos = 292\n",
      "**protein_pos = 293\n",
      "**protein_pos = 294\n",
      "**protein_pos = 295\n",
      "**protein_pos = 296\n",
      "**protein_pos = 297\n",
      "**protein_pos = 298\n",
      "**protein_pos = 299\n",
      "**protein_pos = 300\n",
      "**protein_pos = 301\n",
      "**protein_pos = 302\n",
      "**protein_pos = 303\n",
      "**protein_pos = 304\n",
      "**protein_pos = 305\n",
      "**protein_pos = 306\n",
      "**protein_pos = 307\n",
      "**protein_pos = 308\n",
      "**protein_pos = 309\n",
      "**protein_pos = 310\n",
      "**protein_pos = 311\n",
      "**protein_pos = 312\n",
      "**protein_pos = 313\n",
      "**protein_pos = 314\n",
      "**protein_pos = 315\n",
      "**protein_pos = 316\n",
      "**protein_pos = 317\n",
      "**protein_pos = 318\n",
      "**protein_pos = 319\n",
      "**protein_pos = 320\n",
      "**protein_pos = 321\n",
      "**protein_pos = 322\n",
      "**protein_pos = 323\n",
      "**protein_pos = 324\n",
      "**protein_pos = 325\n",
      "**protein_pos = 326\n",
      "**protein_pos = 327\n",
      "**protein_pos = 328\n",
      "**protein_pos = 329\n",
      "**protein_pos = 330\n",
      "**protein_pos = 331\n",
      "**protein_pos = 332\n",
      "**protein_pos = 333\n",
      "**protein_pos = 334\n",
      "**protein_pos = 335\n",
      "**protein_pos = 336\n",
      "**protein_pos = 337\n",
      "**protein_pos = 338\n",
      "**protein_pos = 339\n",
      "**protein_pos = 340\n",
      "**protein_pos = 341\n",
      "**protein_pos = 342\n",
      "**protein_pos = 343\n",
      "**protein_pos = 344\n",
      "**protein_pos = 345\n",
      "**protein_pos = 346\n",
      "**protein_pos = 347\n",
      "**protein_pos = 348\n",
      "**protein_pos = 349\n",
      "**protein_pos = 350\n",
      "**protein_pos = 351\n",
      "**protein_pos = 352\n",
      "**protein_pos = 353\n",
      "**protein_pos = 354\n",
      "**protein_pos = 355\n",
      "**protein_pos = 356\n",
      "**protein_pos = 357\n",
      "**protein_pos = 358\n",
      "**protein_pos = 359\n",
      "**protein_pos = 360\n",
      "**protein_pos = 361\n",
      "**protein_pos = 362\n",
      "**protein_pos = 363\n",
      "**protein_pos = 364\n",
      "**protein_pos = 365\n",
      "**protein_pos = 366\n",
      "**protein_pos = 367\n",
      "**protein_pos = 368\n",
      "**protein_pos = 369\n",
      "**protein_pos = 370\n",
      "**protein_pos = 371\n",
      "**protein_pos = 372\n",
      "**protein_pos = 373\n",
      "**protein_pos = 374\n",
      "**protein_pos = 375\n",
      "**protein_pos = 376\n",
      "**protein_pos = 377\n",
      "**protein_pos = 378\n",
      "**protein_pos = 379\n",
      "**protein_pos = 380\n",
      "**protein_pos = 381\n",
      "**protein_pos = 382\n",
      "**protein_pos = 383\n",
      "**protein_pos = 384\n",
      "**protein_pos = 385\n",
      "**protein_pos = 386\n",
      "**protein_pos = 387\n",
      "**protein_pos = 388\n",
      "**protein_pos = 389\n",
      "**protein_pos = 390\n",
      "**protein_pos = 391\n",
      "**protein_pos = 392\n",
      "**protein_pos = 393\n",
      "**protein_pos = 394\n",
      "**protein_pos = 395\n",
      "**protein_pos = 396\n",
      "**protein_pos = 397\n",
      "**protein_pos = 398\n",
      "**protein_pos = 399\n",
      "**protein_pos = 400\n",
      "**protein_pos = 401\n",
      "**protein_pos = 402\n",
      "**protein_pos = 403\n",
      "**protein_pos = 404\n",
      "**protein_pos = 405\n",
      "**protein_pos = 406\n",
      "**protein_pos = 407\n",
      "**protein_pos = 408\n",
      "**protein_pos = 409\n",
      "**protein_pos = 410\n",
      "**protein_pos = 411\n",
      "**protein_pos = 412\n",
      "**protein_pos = 413\n",
      "**protein_pos = 414\n",
      "**protein_pos = 415\n",
      "**protein_pos = 416\n",
      "**protein_pos = 417\n",
      "**protein_pos = 418\n",
      "**protein_pos = 419\n",
      "**protein_pos = 420\n",
      "**protein_pos = 421\n",
      "**protein_pos = 422\n",
      "**protein_pos = 423\n",
      "**protein_pos = 424\n",
      "**protein_pos = 425\n",
      "**protein_pos = 426\n",
      "**protein_pos = 427\n",
      "**protein_pos = 428\n",
      "**protein_pos = 429\n",
      "**protein_pos = 430\n",
      "**protein_pos = 431\n",
      "**protein_pos = 432\n",
      "**protein_pos = 433\n",
      "**protein_pos = 434\n",
      "**protein_pos = 435\n",
      "**protein_pos = 436\n",
      "**protein_pos = 437\n",
      "**protein_pos = 438\n",
      "**protein_pos = 439\n",
      "**protein_pos = 440\n",
      "**protein_pos = 441\n",
      "**protein_pos = 442\n",
      "**protein_pos = 443\n",
      "**protein_pos = 444\n",
      "**protein_pos = 445\n",
      "**protein_pos = 446\n",
      "**protein_pos = 447\n",
      "**protein_pos = 448\n",
      "**protein_pos = 449\n",
      "**protein_pos = 450\n",
      "**protein_pos = 451\n",
      "**protein_pos = 452\n",
      "**protein_pos = 453\n",
      "**protein_pos = 454\n",
      "**protein_pos = 455\n",
      "**protein_pos = 456\n",
      "**protein_pos = 457\n",
      "**protein_pos = 458\n",
      "**protein_pos = 459\n",
      "**protein_pos = 460\n",
      "**protein_pos = 461\n",
      "**protein_pos = 462\n",
      "**protein_pos = 463\n",
      "**protein_pos = 464\n",
      "**protein_pos = 465\n",
      "**protein_pos = 466\n",
      "**protein_pos = 467\n",
      "**protein_pos = 468\n",
      "**protein_pos = 469\n",
      "**protein_pos = 470\n",
      "**protein_pos = 471\n",
      "**protein_pos = 472\n",
      "**protein_pos = 473\n",
      "**protein_pos = 474\n",
      "**protein_pos = 475\n",
      "**protein_pos = 476\n",
      "**protein_pos = 477\n",
      "**protein_pos = 478\n",
      "**protein_pos = 479\n",
      "**protein_pos = 480\n",
      "**protein_pos = 481\n",
      "**protein_pos = 482\n",
      "**protein_pos = 483\n",
      "**protein_pos = 484\n",
      "**protein_pos = 485\n",
      "**protein_pos = 486\n",
      "**protein_pos = 487\n",
      "**protein_pos = 488\n",
      "**protein_pos = 489\n",
      "**protein_pos = 490\n",
      "**protein_pos = 491\n",
      "**protein_pos = 492\n",
      "**protein_pos = 493\n",
      "**protein_pos = 494\n",
      "**protein_pos = 495\n",
      "**protein_pos = 496\n",
      "**protein_pos = 497\n",
      "**protein_pos = 498\n",
      "**protein_pos = 499\n",
      "**protein_pos = 500\n",
      "**protein_pos = 501\n",
      "**protein_pos = 502\n",
      "**protein_pos = 503\n",
      "**protein_pos = 504\n",
      "**protein_pos = 505\n",
      "**protein_pos = 506\n",
      "**protein_pos = 507\n",
      "**protein_pos = 508\n",
      "**protein_pos = 509\n",
      "**protein_pos = 510\n",
      "**protein_pos = 511\n",
      "**protein_pos = 512\n",
      "**protein_pos = 513\n",
      "**protein_pos = 514\n",
      "**protein_pos = 515\n",
      "**protein_pos = 516\n",
      "**protein_pos = 517\n",
      "**protein_pos = 518\n",
      "**protein_pos = 519\n",
      "**protein_pos = 520\n",
      "**protein_pos = 521\n",
      "**protein_pos = 522\n",
      "**protein_pos = 523\n",
      "**protein_pos = 524\n",
      "**protein_pos = 525\n",
      "**protein_pos = 526\n",
      "**protein_pos = 527\n",
      "**protein_pos = 528\n",
      "**protein_pos = 529\n",
      "**protein_pos = 530\n",
      "**protein_pos = 531\n",
      "**protein_pos = 532\n",
      "**protein_pos = 533\n",
      "**protein_pos = 534\n",
      "**protein_pos = 535\n",
      "**protein_pos = 536\n",
      "**protein_pos = 537\n",
      "**protein_pos = 538\n",
      "**protein_pos = 539\n",
      "**protein_pos = 540\n",
      "**protein_pos = 541\n",
      "**protein_pos = 542\n",
      "**protein_pos = 543\n",
      "**protein_pos = 544\n",
      "**protein_pos = 545\n",
      "**protein_pos = 546\n",
      "**protein_pos = 547\n",
      "**protein_pos = 548\n",
      "**protein_pos = 549\n",
      "**protein_pos = 550\n",
      "**protein_pos = 551\n",
      "**protein_pos = 552\n",
      "**protein_pos = 553\n",
      "**protein_pos = 554\n",
      "**protein_pos = 555\n",
      "**protein_pos = 556\n",
      "**protein_pos = 557\n",
      "**protein_pos = 558\n",
      "**protein_pos = 559\n",
      "**protein_pos = 560\n",
      "**protein_pos = 561\n",
      "**protein_pos = 562\n",
      "**protein_pos = 563\n",
      "**protein_pos = 564\n",
      "**protein_pos = 565\n",
      "**protein_pos = 566\n",
      "**protein_pos = 567\n",
      "**protein_pos = 568\n",
      "**protein_pos = 569\n",
      "**protein_pos = 570\n",
      "**protein_pos = 571\n",
      "**protein_pos = 572\n",
      "**protein_pos = 573\n",
      "**protein_pos = 574\n",
      "**protein_pos = 575\n",
      "**protein_pos = 576\n",
      "**protein_pos = 577\n",
      "**protein_pos = 578\n",
      "**protein_pos = 579\n",
      "**protein_pos = 580\n",
      "**protein_pos = 581\n",
      "**protein_pos = 582\n",
      "**protein_pos = 583\n",
      "**protein_pos = 584\n",
      "**protein_pos = 585\n",
      "**protein_pos = 586\n",
      "**protein_pos = 587\n",
      "**protein_pos = 588\n",
      "**protein_pos = 589\n",
      "**protein_pos = 590\n",
      "**protein_pos = 591\n",
      "**protein_pos = 592\n",
      "**protein_pos = 593\n",
      "**protein_pos = 594\n",
      "**protein_pos = 595\n",
      "**protein_pos = 596\n",
      "**protein_pos = 597\n",
      "**protein_pos = 598\n",
      "**protein_pos = 599\n",
      "**protein_pos = 600\n",
      "**protein_pos = 601\n",
      "**protein_pos = 602\n",
      "**protein_pos = 603\n",
      "**protein_pos = 604\n"
     ]
    }
   ],
   "source": [
    "ens_gene = \"ENSG00000158691\"\n",
    "\n",
    "#Filtering the domain data for this gene according to the canonical protein id\n",
    "canonic_prot = canonic_protein[ens_gene]\n",
    "domain_gene_table = domain_chrom_data[domain_chrom_data[\"prot_id\"] == canonic_prot]\n",
    "#Making sure that if two HMM-matches overlaps, the higher bit score will come first in the table\n",
    "domain_gene_table = domain_gene_table.sort_values(by=\"BitScore\", ascending=False)\n",
    "\n",
    "#Getting the chosen protein transcript id for the ExAC filtering\n",
    "canonic_transcript_id = domain_gene_table[\"transcript_id\"].tolist()[0]\n",
    "\n",
    "#Creating a table of the exons for this gene, according to the canonical protein\n",
    "chrom_raw_data = domain_gene_table[\"chromosome_id\"].unique()[0] #there should be only one element here\n",
    "if (len(domain_gene_table[\"chromosome_id\"].unique()) > 1):\n",
    "    print functionNameAsString+\" Error: \"+ens_gene+\": more than one chromosome raw data\" #sanity check\n",
    "exon_table = create_exon_pos_table(chrom_raw_data)\n",
    "\n",
    "#Filtering the chromosome data to the gene exons region\n",
    "exons_start_pos = min(exon_table[\"start_pos\"][0],exon_table[\"start_pos\"][len(exon_table)-1]) #in case of complelemt, the minimal position could be at the last row\n",
    "exons_end_pos = max(exon_table[\"end_pos\"][0],exon_table[\"end_pos\"][len(exon_table)-1]) #in case of complelemt, the maximal position could be at the first row\n",
    "chrom_gene_table = chrom_csv[chrom_csv[\"pos\"] >= int(exons_start_pos)][chrom_csv[\"pos\"] <= int(exons_end_pos)][chrom_csv[\"feature\"] == canonic_transcript_id]\n",
    "chrom_gene_table = chrom_gene_table.reset_index(drop=True)\n",
    "\n",
    "#Handling indels\n",
    "indels_table = table_editing(chrom_gene_table)\n",
    "\n",
    "#A counter for frameshifts inside the domain\n",
    "protein_frameshifts_cnt = 0\n",
    "#A counter for validation errors inside the domain\n",
    "protein_errors_cnt = 0\n",
    "\n",
    "#Iterating over the amino-acids of the protein\n",
    "prot_len = domain_gene_table[\"length\"].unique()[0]\n",
    "for protein_pos in range(1,prot_len+1):\n",
    "\n",
    "    #Trying to match HMM-state, and retreive the aa from HMMER results\n",
    "    (hmm_state, aa) = protein_pos_to_hmm_state_and_aa(protein_pos, domain_gene_table) #TODO: what happens when two matches overlap? maybe sort to the best bit score?\n",
    "\n",
    "    #If there's a match to HMM-state: find the corresponding codon bps chromosome positions\n",
    "    if (hmm_state > 0):\n",
    "        chrom_pos_list =find_chrom_bps(protein_pos, exon_table, chrom_raw_data)\n",
    "\n",
    "        #Analysis of the amino-acid MAF and realted data, returned in a dictionary\n",
    "        (info_dict, frameshift_cnt, errors_cnt) = calc_exac_maf_data(chrom_pos_list, chrom_gene_table, indels_table, protein_pos, aa, chrom_raw_data, chrom)\n",
    "\n",
    "        #Adding the dictionary to the HMM-state list\n",
    "        states_dict[hmm_state].append(info_dict)\n",
    "\n",
    "        #Adding the frameshifts to the global counter\n",
    "        protein_frameshifts_cnt += frameshift_cnt\n",
    "\n",
    "        #Adding the errors to the global counter\n",
    "        protein_errors_cnt += errors_cnt\n",
    "\n",
    "domain_ens_genes_frameshifts.append(protein_frameshifts_cnt)\n",
    "domain_ens_genes_errors.append(protein_errors_cnt)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
