{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>.container { width:100% !important; }</style>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#Import packages\n",
    "import gzip\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "from collections import defaultdict\n",
    "from IPython.core.display import HTML\n",
    "HTML(\"<style>.container { width:100% !important; }</style>\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "#Global vars\n",
    "#path = \"/Users/anat/Documents/Princeton/Research/ExAC/\" #Mac path\n",
    "path = \"/home/anat/Research/ExAC/\" #Linux path\n",
    "myFile = \"ExAC.r0.3.nonTCGA.sites.vep.vcf.gz\"\n",
    "\n",
    "#Positions in the VCF record\n",
    "CHROM = 0\n",
    "POS = 1\n",
    "ID = 2\n",
    "REF = 3\n",
    "ALT = 4\n",
    "QUAL = 5\n",
    "FILTER = 6\n",
    "INFO = 7\n",
    "\n",
    "#fields for extraction (it's important those names will match *exactly* the dictionary keys in \"data_dict\")\n",
    "headlines = [\"chrom\", \"pos\", \"id\", \"ref\", \"alt\", \"qual\", \"filter\", \"AC\", \"AC_Adj\", \"AF\", \"AN\", \"AN_Adj\", \"DP\", \"gene\", \"conseq\", \"prot_pos\",\n",
    "            \"amino_acids\", \"codons\", \"SWISSPROT\", \"SIFT\", \"domains\", \"clin_sig\"]\n",
    "\n",
    "#CSQ positions\n",
    "#http://useast.ensembl.org/info/docs/tools/vep/vep_formats.html?redirect=no\n",
    "GENE = 0        #Ensembl stable ID of affected gene\n",
    "CONSEQ = 3      #consequence type of this variant\n",
    "PROT_POS = 6    #relative position of amino acid in protein\n",
    "AMINO_ACIDS = 7 #the change. only given if the variant affects the protein-coding sequence\n",
    "CODONS = 8      #the alternative codons with the variant base in upper case\n",
    "SWISSPROT = 20  #UniProtKB/Swiss-Prot identifier of protein product\n",
    "SIFT = 23       #the SIFT prediction and/or score, with both given as prediction(score)\n",
    "DOMAINS = 27    #the source and identifer of any overlapping protein domains\n",
    "GMAF = 30       #Non-reference allele and frequency of existing variant in 1000 Genomes\n",
    "CLIN_SIG = 37   #Clinical significance of variant from dbSNP: http://varianttools.sourceforge.net/Annotation/DbSNP\n",
    "                #Variant Clinical Significance, 0 - unknown, 1 -\n",
    "                #untested, 2 - non-pathogenic, 3 - probable-non-pathogenic, 4 - probable-pathogenic, \n",
    "                #5 - pathogenic, 6 - drug-response, 7 - histocompatibility, 255 - other"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "#A function to find the nth position on a substring x in the string s\n",
    "def find_nth(s, x, n=0, overlap=False):\n",
    "    l = 1 if overlap else len(x)\n",
    "    i = -l\n",
    "    for c in xrange(n + 1):\n",
    "        i = s.find(x, i + l)\n",
    "        if i < 0:\n",
    "            break\n",
    "    return i"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "vcf_file = gzip.open(path+myFile,'r')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "#Process meta-data\n",
    "metadata_dict = {}\n",
    "data_flag = False\n",
    "for line in vcf_file:\n",
    "    if line[0:2] == \"##\":\n",
    "        #assign keys according to the format\n",
    "        key = line[2:line.index('=')]\n",
    "        if key == \"ALT\":\n",
    "            val = dict.fromkeys([\"ID\",\"Description\"])\n",
    "        elif key == \"FILTER\":\n",
    "            val = dict.fromkeys([\"ID\",\"Description\"])\n",
    "        elif key == \"FORMAT\":\n",
    "            val = dict.fromkeys([\"ID\",\"Number\",\"Type\",\"Description\"])\n",
    "        elif key == \"INFO\":\n",
    "            val = dict.fromkeys([\"ID\",\"Number\", \"Type\", \"Description\"])\n",
    "        elif key == \"contig\":\n",
    "            val = dict.fromkeys([\"ID\",\"length\"])\n",
    "        elif key == \"reference\":\n",
    "            val = dict.fromkeys([\"file\"])\n",
    "        #Not processing other metadata types\n",
    "        else:\n",
    "            continue\n",
    "        #fill in the data\n",
    "        for f in val.keys():\n",
    "            f_key = line.find(f)\n",
    "            f_beg = line.find(\"=\", f_key)\n",
    "            f_end = line.find(\",\", f_beg)\n",
    "            if (f_end < 0):\n",
    "                f_end = line.find(\">\")\n",
    "            val[f] = line[f_beg + 1:f_end]\n",
    "        #Adding to the metadata dictionary\n",
    "        if not metadata_dict.has_key(key):\n",
    "            metadata_dict[key] = [val]\n",
    "        else:\n",
    "            metadata_dict[key].append(val)\n",
    "            \n",
    "    #Processing the data starting the next line\n",
    "    elif line[0:6] == \"#CHROM\":\n",
    "        data_flag = True\n",
    "        break"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "#Arrange the INFO metadata to a data-frame\n",
    "info_df = pd.DataFrame(metadata_dict[\"INFO\"])\n",
    "info_df = info_df.sort_values(\"ID\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "#A function that saves the data dictionary to DataFrame and then to .csv\n",
    "def data_to_df_csv(data_dict, headlines, chrom_num):\n",
    "    #Creating a data_frame from all the parsed values of the chromosome\n",
    "    df = pd.DataFrame([data_dict[h] for h in headlines])\n",
    "    df = df.transpose()\n",
    "    df.columns = headlines\n",
    "    \n",
    "    #Saving the df to a file\n",
    "    df.to_csv(path+\"/parsed/parsed_chrom\"+chrom_num+\".csv\", sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "collapsed": false,
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "finished chromosome1\n",
      "finished chromosome2\n",
      "finished chromosome3\n",
      "finished chromosome4\n",
      "finished chromosome5\n",
      "finished chromosome6\n",
      "finished chromosome7\n",
      "finished chromosome8\n",
      "finished chromosome9\n",
      "finished chromosome10\n",
      "finished chromosome11\n",
      "finished chromosome12\n",
      "finished chromosome13\n",
      "finished chromosome14\n",
      "finished chromosome15\n",
      "finished chromosome16\n",
      "finished chromosome17\n",
      "finished chromosome18\n",
      "finished chromosome19\n",
      "finished chromosome20\n",
      "finished chromosome21\n",
      "finished chromosome22\n",
      "finished chromosomeX\n"
     ]
    }
   ],
   "source": [
    "#Process the data records of the vcf and save each chromosome to a seperate file\n",
    "chromosome_iter = '1'\n",
    "data_dict = defaultdict(list)\n",
    "for line in vcf_file:\n",
    "    line_parts = line.split(\"\\t\")\n",
    "    \n",
    "    #If the next line belongs to a different chromosome - saving to file\n",
    "    if line_parts[CHROM] != chromosome_iter:\n",
    "        data_to_df_csv(data_dict, headlines, chromosome_iter)\n",
    "        #Initializing the data dictionary\n",
    "        data_dict = defaultdict(list)\n",
    "        print \"finished chromosome\"+chromosome_iter\n",
    "        chromosome_iter = line_parts[CHROM]\n",
    "    \n",
    "    #Extracting Chromosome number\n",
    "    data_dict[\"chrom\"].append(line_parts[CHROM])\n",
    "    \n",
    "    #Extracting position\n",
    "    data_dict[\"pos\"].append(int(line_parts[POS]))\n",
    "    \n",
    "    #Extracting id\n",
    "    data_dict[\"id\"].append(line_parts[ID])\n",
    "    \n",
    "    #Extracting ref\n",
    "    data_dict[\"ref\"].append(line_parts[REF])\n",
    "    \n",
    "    #Extracting alt\n",
    "    data_dict[\"alt\"].append(line_parts[ALT])\n",
    "    \n",
    "    #Extracting quality\n",
    "    data_dict[\"qual\"].append(line_parts[QUAL])\n",
    "    \n",
    "    #Extracting filter\n",
    "    data_dict[\"filter\"].append(line_parts[FILTER])\n",
    "    \n",
    "    #Extracting fields from the info\n",
    "    info = line_parts[INFO]\n",
    "    \n",
    "    #AC = Allele Count\n",
    "    AC_beg = info.find(\"AC=\")\n",
    "    AC_end = info.find(\";\", AC_beg)\n",
    "    data_dict[\"AC\"].append(info[AC_beg+3:AC_end])\n",
    "    \n",
    "    #AC_adjusted = Adjusted Allele Count\n",
    "    AC_adj_beg = info.find(\"AC_Adj=\")\n",
    "    AC_adj_end = info.find(\";\", AC_adj_beg)\n",
    "    data_dict[\"AC_Adj\"].append(info[AC_adj_beg+7:AC_adj_end])\n",
    "    \n",
    "    #AF = Allele Frequency \n",
    "    AF_beg = info.find(\"AF=\")\n",
    "    AF_end = info.find(\";\", AF_beg)\n",
    "    data_dict[\"AF\"].append(info[AF_beg+3:AF_end])\n",
    "    \n",
    "    #AN = Allele Number\n",
    "    AN_beg = info.find(\"AN=\")\n",
    "    AN_end = info.find(\";\", AN_beg)\n",
    "    data_dict[\"AN\"].append(info[AN_beg+3:AN_end])\n",
    "    \n",
    "    #AN_adj = Adjusted Allele Number\n",
    "    AN_adj_beg = info.find(\"AN_Adj=\")\n",
    "    AN_adj_end = info.find(\";\", AN_adj_beg)\n",
    "    data_dict[\"AN_Adj\"].append(info[AN_adj_beg+7:AN_adj_end])\n",
    "    \n",
    "    #DP = \"Approximate read depth\n",
    "    DP_beg = info.find(\"DP=\")\n",
    "    DP_end = info.find(\";\", DP_beg)\n",
    "    data_dict[\"DP\"].append(info[DP_beg+3:DP_end])\n",
    "    \n",
    "    \n",
    "    #CSQ = Consequence type as predicted by VEP\n",
    "    CSQ_beg = info.find(\"CSQ=\")\n",
    "    CSQ = info[CSQ_beg+3:]\n",
    "    #Gene\n",
    "    gene_beg = find_nth(CSQ, \"|\", GENE)\n",
    "    gene_end = CSQ.find(\"|\", gene_beg+1)\n",
    "    data_dict[\"gene\"].append(CSQ[gene_beg+1:gene_end])\n",
    "    #Consequence\n",
    "    conseq_beg = find_nth(CSQ, \"|\", CONSEQ)\n",
    "    conseq_end = CSQ.find(\"|\", conseq_beg+1)\n",
    "    data_dict[\"conseq\"].append(CSQ[conseq_beg+1:conseq_end])\n",
    "    #Protein_pos\n",
    "    prot_pos_beg = find_nth(CSQ, \"|\", PROT_POS)\n",
    "    prot_pos_end = CSQ.find(\"|\", prot_pos_beg+1)\n",
    "    data_dict[\"prot_pos\"].append(CSQ[prot_pos_beg+1:prot_pos_end])\n",
    "    #Amino Acids\n",
    "    aa_beg = find_nth(CSQ, \"|\", AMINO_ACIDS)\n",
    "    aa_end = CSQ.find(\"|\", aa_beg+1)\n",
    "    data_dict[\"amino_acids\"].append(CSQ[aa_beg+1:aa_end])\n",
    "    #Codons\n",
    "    codons_beg = find_nth(CSQ, \"|\", CODONS)\n",
    "    codons_end = CSQ.find(\"|\", codons_beg+1)\n",
    "    data_dict[\"codons\"].append(CSQ[codons_beg+1:codons_end])\n",
    "    #Swissprot\n",
    "    swiss_beg = find_nth(CSQ, \"|\", SWISSPROT)\n",
    "    swiss_end = CSQ.find(\"|\", swiss_beg+1)\n",
    "    data_dict[\"SWISSPROT\"].append(CSQ[swiss_beg+1:swiss_end])\n",
    "    #SIFT\n",
    "    sift_beg = find_nth(CSQ, \"|\", SIFT)\n",
    "    sift_end = CSQ.find(\"|\", sift_beg+1)\n",
    "    data_dict[\"SIFT\"].append(CSQ[sift_beg+1:sift_end])\n",
    "    #Domains\n",
    "    domains_beg = find_nth(CSQ, \"|\", DOMAINS)\n",
    "    domains_end = CSQ.find(\"|\", domains_beg+1)\n",
    "    data_dict[\"domains\"].append(CSQ[domains_beg+1:domains_end])\n",
    "    #clin_sig\n",
    "    clin_sig_beg = find_nth(CSQ, \"|\", CLIN_SIG)\n",
    "    clin_sig_end = CSQ.find(\"|\", clin_sig_beg+1)\n",
    "    data_dict[\"clin_sig\"].append(CSQ[clin_sig_beg+1:clin_sig_end])\n",
    "\n",
    "#Saving the data of the last chromosome\n",
    "data_to_df_csv(data_dict, headlines, chromosome_iter)\n",
    "print \"finished chromosome\"+chromosome_iter"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "vcf_file.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 182,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Description</th>\n",
       "      <th>ID</th>\n",
       "      <th>Number</th>\n",
       "      <th>Type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>70</th>\n",
       "      <td>\"Consequence type as predicted by VEP. Format: Allele|Gene|Feature|Feature_type|Consequence|cDNA_position|CDS_position|Protein_position|Amino_acids|Codons|Existing_variation|ALLELE_NUM|DISTANCE|STRAND|SYMBOL|SYMBOL_SOURCE|HGNC_ID|BIOTYPE|CANONICAL|CCDS|ENSP|SWISSPROT|TREMBL|UNIPARC|SIFT|PolyPhen|EXON|INTRON|DOMAINS|HGVSc|HGVSp|GMAF|AFR_MAF|AMR_MAF|ASN_MAF|EUR_MAF|AA_MAF|EA_MAF|CLIN_SIG|SOMATIC|PUBMED|MOTIF_NAME|MOTIF_POS|HIGH_INF_POS|MOTIF_SCORE_CHANGE|LoF_info|LoF_flags|LoF_filter|LoF\"</td>\n",
       "      <td>CSQ</td>\n",
       "      <td>.</td>\n",
       "      <td>String</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                    Description  \\\n",
       "70  \"Consequence type as predicted by VEP. Format: Allele|Gene|Feature|Feature_type|Consequence|cDNA_position|CDS_position|Protein_position|Amino_acids|Codons|Existing_variation|ALLELE_NUM|DISTANCE|STRAND|SYMBOL|SYMBOL_SOURCE|HGNC_ID|BIOTYPE|CANONICAL|CCDS|ENSP|SWISSPROT|TREMBL|UNIPARC|SIFT|PolyPhen|EXON|INTRON|DOMAINS|HGVSc|HGVSp|GMAF|AFR_MAF|AMR_MAF|ASN_MAF|EUR_MAF|AA_MAF|EA_MAF|CLIN_SIG|SOMATIC|PUBMED|MOTIF_NAME|MOTIF_POS|HIGH_INF_POS|MOTIF_SCORE_CHANGE|LoF_info|LoF_flags|LoF_filter|LoF\"   \n",
       "\n",
       "     ID Number    Type  \n",
       "70  CSQ      .  String  "
      ]
     },
     "execution_count": 182,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "info_df[info_df['ID'] == \"CSQ\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 173,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Description</th>\n",
       "      <th>ID</th>\n",
       "      <th>Number</th>\n",
       "      <th>Type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>\"Allele count in genotypes</td>\n",
       "      <td>AC</td>\n",
       "      <td>A</td>\n",
       "      <td>Integer</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>\"African/African American Allele Counts\"</td>\n",
       "      <td>AC_AFR</td>\n",
       "      <td>A</td>\n",
       "      <td>Integer</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>\"American Allele Counts\"</td>\n",
       "      <td>AC_AMR</td>\n",
       "      <td>A</td>\n",
       "      <td>Integer</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>\"Adjusted Allele Counts\"</td>\n",
       "      <td>AC_Adj</td>\n",
       "      <td>A</td>\n",
       "      <td>Integer</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>\"East Asian Allele Counts\"</td>\n",
       "      <td>AC_EAS</td>\n",
       "      <td>A</td>\n",
       "      <td>Integer</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>\"Finnish Allele Counts\"</td>\n",
       "      <td>AC_FIN</td>\n",
       "      <td>A</td>\n",
       "      <td>Integer</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>\"Adjusted Hemizygous Counts\"</td>\n",
       "      <td>AC_Hemi</td>\n",
       "      <td>A</td>\n",
       "      <td>Integer</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>\"Adjusted Heterozygous Counts\"</td>\n",
       "      <td>AC_Het</td>\n",
       "      <td>A</td>\n",
       "      <td>Integer</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>\"Adjusted Homozygous Counts\"</td>\n",
       "      <td>AC_Hom</td>\n",
       "      <td>A</td>\n",
       "      <td>Integer</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>\"Non-Finnish European Allele Counts\"</td>\n",
       "      <td>AC_NFE</td>\n",
       "      <td>A</td>\n",
       "      <td>Integer</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>\"Other Allele Counts\"</td>\n",
       "      <td>AC_OTH</td>\n",
       "      <td>A</td>\n",
       "      <td>Integer</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>\"South Asian Allele Counts\"</td>\n",
       "      <td>AC_SAS</td>\n",
       "      <td>A</td>\n",
       "      <td>Integer</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>\"Allele Frequency</td>\n",
       "      <td>AF</td>\n",
       "      <td>A</td>\n",
       "      <td>Float</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>\"Total number of alleles in called genotypes\"</td>\n",
       "      <td>AN</td>\n",
       "      <td>1</td>\n",
       "      <td>Integer</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>\"African/African American Chromosome Count\"</td>\n",
       "      <td>AN_AFR</td>\n",
       "      <td>1</td>\n",
       "      <td>Integer</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>\"American Chromosome Count\"</td>\n",
       "      <td>AN_AMR</td>\n",
       "      <td>1</td>\n",
       "      <td>Integer</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>\"Adjusted Chromosome Count\"</td>\n",
       "      <td>AN_Adj</td>\n",
       "      <td>1</td>\n",
       "      <td>Integer</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>\"East Asian Chromosome Count\"</td>\n",
       "      <td>AN_EAS</td>\n",
       "      <td>1</td>\n",
       "      <td>Integer</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>\"Finnish Chromosome Count\"</td>\n",
       "      <td>AN_FIN</td>\n",
       "      <td>1</td>\n",
       "      <td>Integer</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>\"Non-Finnish European Chromosome Count\"</td>\n",
       "      <td>AN_NFE</td>\n",
       "      <td>1</td>\n",
       "      <td>Integer</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>\"Other Chromosome Count\"</td>\n",
       "      <td>AN_OTH</td>\n",
       "      <td>1</td>\n",
       "      <td>Integer</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>\"South Asian Chromosome Count\"</td>\n",
       "      <td>AN_SAS</td>\n",
       "      <td>1</td>\n",
       "      <td>Integer</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>\"Z-score from Wilcoxon rank sum test of Alt Vs...</td>\n",
       "      <td>BaseQRankSum</td>\n",
       "      <td>1</td>\n",
       "      <td>Float</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>\"Number of called chromosomes\"</td>\n",
       "      <td>CCC</td>\n",
       "      <td>1</td>\n",
       "      <td>Integer</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>70</th>\n",
       "      <td>\"Consequence type as predicted by VEP. Format:...</td>\n",
       "      <td>CSQ</td>\n",
       "      <td>.</td>\n",
       "      <td>String</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>\"Z-score From Wilcoxon rank sum test of Alt vs...</td>\n",
       "      <td>ClippingRankSum</td>\n",
       "      <td>1</td>\n",
       "      <td>Float</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>\"dbSNP Membership\"</td>\n",
       "      <td>DB</td>\n",
       "      <td>0</td>\n",
       "      <td>Flag</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>\"Approximate read depth; some reads may have b...</td>\n",
       "      <td>DP</td>\n",
       "      <td>1</td>\n",
       "      <td>Integer</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>68</th>\n",
       "      <td>\"Histogram for DP; Mids: 2.5|7.5|12.5|17.5|22....</td>\n",
       "      <td>DP_HIST</td>\n",
       "      <td>A</td>\n",
       "      <td>String</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>\"Were any of the samples downsampled?\"</td>\n",
       "      <td>DS</td>\n",
       "      <td>0</td>\n",
       "      <td>Flag</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>\"Non-Finnish European Hemizygous Counts\"</td>\n",
       "      <td>Hemi_NFE</td>\n",
       "      <td>A</td>\n",
       "      <td>Integer</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39</th>\n",
       "      <td>\"Other Hemizygous Counts\"</td>\n",
       "      <td>Hemi_OTH</td>\n",
       "      <td>A</td>\n",
       "      <td>Integer</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40</th>\n",
       "      <td>\"South Asian Hemizygous Counts\"</td>\n",
       "      <td>Hemi_SAS</td>\n",
       "      <td>A</td>\n",
       "      <td>Integer</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41</th>\n",
       "      <td>\"African/African American Heterozygous Counts\"</td>\n",
       "      <td>Het_AFR</td>\n",
       "      <td>A</td>\n",
       "      <td>Integer</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42</th>\n",
       "      <td>\"American Heterozygous Counts\"</td>\n",
       "      <td>Het_AMR</td>\n",
       "      <td>A</td>\n",
       "      <td>Integer</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43</th>\n",
       "      <td>\"East Asian Heterozygous Counts\"</td>\n",
       "      <td>Het_EAS</td>\n",
       "      <td>A</td>\n",
       "      <td>Integer</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44</th>\n",
       "      <td>\"Finnish Heterozygous Counts\"</td>\n",
       "      <td>Het_FIN</td>\n",
       "      <td>A</td>\n",
       "      <td>Integer</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>45</th>\n",
       "      <td>\"Non-Finnish European Heterozygous Counts\"</td>\n",
       "      <td>Het_NFE</td>\n",
       "      <td>A</td>\n",
       "      <td>Integer</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46</th>\n",
       "      <td>\"Other Heterozygous Counts\"</td>\n",
       "      <td>Het_OTH</td>\n",
       "      <td>A</td>\n",
       "      <td>Integer</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>47</th>\n",
       "      <td>\"South Asian Heterozygous Counts\"</td>\n",
       "      <td>Het_SAS</td>\n",
       "      <td>A</td>\n",
       "      <td>Integer</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48</th>\n",
       "      <td>\"African/African American Homozygous Counts\"</td>\n",
       "      <td>Hom_AFR</td>\n",
       "      <td>A</td>\n",
       "      <td>Integer</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>49</th>\n",
       "      <td>\"American Homozygous Counts\"</td>\n",
       "      <td>Hom_AMR</td>\n",
       "      <td>A</td>\n",
       "      <td>Integer</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50</th>\n",
       "      <td>\"East Asian Homozygous Counts\"</td>\n",
       "      <td>Hom_EAS</td>\n",
       "      <td>A</td>\n",
       "      <td>Integer</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>51</th>\n",
       "      <td>\"Finnish Homozygous Counts\"</td>\n",
       "      <td>Hom_FIN</td>\n",
       "      <td>A</td>\n",
       "      <td>Integer</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>52</th>\n",
       "      <td>\"Non-Finnish European Homozygous Counts\"</td>\n",
       "      <td>Hom_NFE</td>\n",
       "      <td>A</td>\n",
       "      <td>Integer</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>53</th>\n",
       "      <td>\"Other Homozygous Counts\"</td>\n",
       "      <td>Hom_OTH</td>\n",
       "      <td>A</td>\n",
       "      <td>Integer</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54</th>\n",
       "      <td>\"South Asian Homozygous Counts\"</td>\n",
       "      <td>Hom_SAS</td>\n",
       "      <td>A</td>\n",
       "      <td>Integer</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>55</th>\n",
       "      <td>\"Inbreeding coefficient as estimated from the ...</td>\n",
       "      <td>InbreedingCoeff</td>\n",
       "      <td>1</td>\n",
       "      <td>Float</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>56</th>\n",
       "      <td>\"Maximum likelihood expectation (MLE) for the ...</td>\n",
       "      <td>MLEAC</td>\n",
       "      <td>A</td>\n",
       "      <td>Integer</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>57</th>\n",
       "      <td>\"Maximum likelihood expectation (MLE) for the ...</td>\n",
       "      <td>MLEAF</td>\n",
       "      <td>A</td>\n",
       "      <td>Float</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58</th>\n",
       "      <td>\"RMS Mapping Quality\"</td>\n",
       "      <td>MQ</td>\n",
       "      <td>1</td>\n",
       "      <td>Float</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>59</th>\n",
       "      <td>\"Total Mapping Quality Zero Reads\"</td>\n",
       "      <td>MQ0</td>\n",
       "      <td>1</td>\n",
       "      <td>Integer</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60</th>\n",
       "      <td>\"Z-score From Wilcoxon rank sum test of Alt vs...</td>\n",
       "      <td>MQRankSum</td>\n",
       "      <td>1</td>\n",
       "      <td>Float</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>61</th>\n",
       "      <td>\"Number of no-called samples\"</td>\n",
       "      <td>NCC</td>\n",
       "      <td>1</td>\n",
       "      <td>Integer</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>62</th>\n",
       "      <td>\"This variant was used to build the negative t...</td>\n",
       "      <td>NEGATIVE_TRAIN_SITE</td>\n",
       "      <td>0</td>\n",
       "      <td>Flag</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>63</th>\n",
       "      <td>\"This variant was used to build the positive t...</td>\n",
       "      <td>POSITIVE_TRAIN_SITE</td>\n",
       "      <td>0</td>\n",
       "      <td>Flag</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>64</th>\n",
       "      <td>\"Variant Confidence/Quality by Depth\"</td>\n",
       "      <td>QD</td>\n",
       "      <td>1</td>\n",
       "      <td>Float</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>65</th>\n",
       "      <td>\"Z-score from Wilcoxon rank sum test of Alt vs...</td>\n",
       "      <td>ReadPosRankSum</td>\n",
       "      <td>1</td>\n",
       "      <td>Float</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>66</th>\n",
       "      <td>\"Log odds ratio of being a true variant versus...</td>\n",
       "      <td>VQSLOD</td>\n",
       "      <td>1</td>\n",
       "      <td>Float</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>67</th>\n",
       "      <td>\"The annotation which was the worst performing...</td>\n",
       "      <td>culprit</td>\n",
       "      <td>1</td>\n",
       "      <td>String</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>71 rows Ã— 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                          Description                   ID  \\\n",
       "0                          \"Allele count in genotypes                   AC   \n",
       "1            \"African/African American Allele Counts\"               AC_AFR   \n",
       "2                            \"American Allele Counts\"               AC_AMR   \n",
       "3                            \"Adjusted Allele Counts\"               AC_Adj   \n",
       "4                          \"East Asian Allele Counts\"               AC_EAS   \n",
       "5                             \"Finnish Allele Counts\"               AC_FIN   \n",
       "6                        \"Adjusted Hemizygous Counts\"              AC_Hemi   \n",
       "7                      \"Adjusted Heterozygous Counts\"               AC_Het   \n",
       "8                        \"Adjusted Homozygous Counts\"               AC_Hom   \n",
       "9                \"Non-Finnish European Allele Counts\"               AC_NFE   \n",
       "10                              \"Other Allele Counts\"               AC_OTH   \n",
       "11                        \"South Asian Allele Counts\"               AC_SAS   \n",
       "12                                  \"Allele Frequency                   AF   \n",
       "13      \"Total number of alleles in called genotypes\"                   AN   \n",
       "14        \"African/African American Chromosome Count\"               AN_AFR   \n",
       "15                        \"American Chromosome Count\"               AN_AMR   \n",
       "16                        \"Adjusted Chromosome Count\"               AN_Adj   \n",
       "17                      \"East Asian Chromosome Count\"               AN_EAS   \n",
       "18                         \"Finnish Chromosome Count\"               AN_FIN   \n",
       "19            \"Non-Finnish European Chromosome Count\"               AN_NFE   \n",
       "20                           \"Other Chromosome Count\"               AN_OTH   \n",
       "21                     \"South Asian Chromosome Count\"               AN_SAS   \n",
       "22  \"Z-score from Wilcoxon rank sum test of Alt Vs...         BaseQRankSum   \n",
       "23                     \"Number of called chromosomes\"                  CCC   \n",
       "70  \"Consequence type as predicted by VEP. Format:...                  CSQ   \n",
       "24  \"Z-score From Wilcoxon rank sum test of Alt vs...      ClippingRankSum   \n",
       "25                                 \"dbSNP Membership\"                   DB   \n",
       "26  \"Approximate read depth; some reads may have b...                   DP   \n",
       "68  \"Histogram for DP; Mids: 2.5|7.5|12.5|17.5|22....              DP_HIST   \n",
       "27             \"Were any of the samples downsampled?\"                   DS   \n",
       "..                                                ...                  ...   \n",
       "38           \"Non-Finnish European Hemizygous Counts\"             Hemi_NFE   \n",
       "39                          \"Other Hemizygous Counts\"             Hemi_OTH   \n",
       "40                    \"South Asian Hemizygous Counts\"             Hemi_SAS   \n",
       "41     \"African/African American Heterozygous Counts\"              Het_AFR   \n",
       "42                     \"American Heterozygous Counts\"              Het_AMR   \n",
       "43                   \"East Asian Heterozygous Counts\"              Het_EAS   \n",
       "44                      \"Finnish Heterozygous Counts\"              Het_FIN   \n",
       "45         \"Non-Finnish European Heterozygous Counts\"              Het_NFE   \n",
       "46                        \"Other Heterozygous Counts\"              Het_OTH   \n",
       "47                  \"South Asian Heterozygous Counts\"              Het_SAS   \n",
       "48       \"African/African American Homozygous Counts\"              Hom_AFR   \n",
       "49                       \"American Homozygous Counts\"              Hom_AMR   \n",
       "50                     \"East Asian Homozygous Counts\"              Hom_EAS   \n",
       "51                        \"Finnish Homozygous Counts\"              Hom_FIN   \n",
       "52           \"Non-Finnish European Homozygous Counts\"              Hom_NFE   \n",
       "53                          \"Other Homozygous Counts\"              Hom_OTH   \n",
       "54                    \"South Asian Homozygous Counts\"              Hom_SAS   \n",
       "55  \"Inbreeding coefficient as estimated from the ...      InbreedingCoeff   \n",
       "56  \"Maximum likelihood expectation (MLE) for the ...                MLEAC   \n",
       "57  \"Maximum likelihood expectation (MLE) for the ...                MLEAF   \n",
       "58                              \"RMS Mapping Quality\"                   MQ   \n",
       "59                 \"Total Mapping Quality Zero Reads\"                  MQ0   \n",
       "60  \"Z-score From Wilcoxon rank sum test of Alt vs...            MQRankSum   \n",
       "61                      \"Number of no-called samples\"                  NCC   \n",
       "62  \"This variant was used to build the negative t...  NEGATIVE_TRAIN_SITE   \n",
       "63  \"This variant was used to build the positive t...  POSITIVE_TRAIN_SITE   \n",
       "64              \"Variant Confidence/Quality by Depth\"                   QD   \n",
       "65  \"Z-score from Wilcoxon rank sum test of Alt vs...       ReadPosRankSum   \n",
       "66  \"Log odds ratio of being a true variant versus...               VQSLOD   \n",
       "67  \"The annotation which was the worst performing...              culprit   \n",
       "\n",
       "   Number     Type  \n",
       "0       A  Integer  \n",
       "1       A  Integer  \n",
       "2       A  Integer  \n",
       "3       A  Integer  \n",
       "4       A  Integer  \n",
       "5       A  Integer  \n",
       "6       A  Integer  \n",
       "7       A  Integer  \n",
       "8       A  Integer  \n",
       "9       A  Integer  \n",
       "10      A  Integer  \n",
       "11      A  Integer  \n",
       "12      A    Float  \n",
       "13      1  Integer  \n",
       "14      1  Integer  \n",
       "15      1  Integer  \n",
       "16      1  Integer  \n",
       "17      1  Integer  \n",
       "18      1  Integer  \n",
       "19      1  Integer  \n",
       "20      1  Integer  \n",
       "21      1  Integer  \n",
       "22      1    Float  \n",
       "23      1  Integer  \n",
       "70      .   String  \n",
       "24      1    Float  \n",
       "25      0     Flag  \n",
       "26      1  Integer  \n",
       "68      A   String  \n",
       "27      0     Flag  \n",
       "..    ...      ...  \n",
       "38      A  Integer  \n",
       "39      A  Integer  \n",
       "40      A  Integer  \n",
       "41      A  Integer  \n",
       "42      A  Integer  \n",
       "43      A  Integer  \n",
       "44      A  Integer  \n",
       "45      A  Integer  \n",
       "46      A  Integer  \n",
       "47      A  Integer  \n",
       "48      A  Integer  \n",
       "49      A  Integer  \n",
       "50      A  Integer  \n",
       "51      A  Integer  \n",
       "52      A  Integer  \n",
       "53      A  Integer  \n",
       "54      A  Integer  \n",
       "55      1    Float  \n",
       "56      A  Integer  \n",
       "57      A    Float  \n",
       "58      1    Float  \n",
       "59      1  Integer  \n",
       "60      1    Float  \n",
       "61      1  Integer  \n",
       "62      0     Flag  \n",
       "63      0     Flag  \n",
       "64      1    Float  \n",
       "65      1    Float  \n",
       "66      1    Float  \n",
       "67      1   String  \n",
       "\n",
       "[71 rows x 4 columns]"
      ]
     },
     "execution_count": 173,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "info_df = info_df.sort_values(\"ID\")\n",
    "info_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "pd.options.display.max_colwidth =600"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
