{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 03/04/2016 - Reading the parsed chromosome from ExAC and map the Esembl IDs to Hugo Symbols\n",
    "This is an old version of parsing the chromosome files that includes mapping to Hugo symbols.\n",
    "\n",
    "This mapping was needed for the first hmmer results obtained from Shilpa.\n",
    "\n",
    "The second hmmer results already contain Ensembl id, so such mapping isn't needed"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 95,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>.container { width:100% !important; }</style>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "execution_count": 95,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "from biomart import BiomartServer\n",
    "from IPython.core.display import HTML\n",
    "HTML(\"<style>.container { width:100% !important; }</style>\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 97,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "#Getting the list of chromosome files\n",
    "curr_dir = !pwd\n",
    "my_path = curr_dir[0]+\"/parsed/\"\n",
    "filename = \"parsed_chrom\"\n",
    "chrom_files = !ls $my_path"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 98,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "#Importing Biomart database\n",
    "server = BiomartServer(\"http://grch37.ensembl.org/biomart\")\n",
    "ens_genes = server.datasets[u'hsapiens_gene_ensembl']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 109,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "finished chrom10\n",
      "finished chrom11\n",
      "finished chrom12\n",
      "finished chrom13\n",
      "finished chrom14\n",
      "finished chrom15\n",
      "finished chrom16\n",
      "finished chrom17\n",
      "finished chrom18\n",
      "finished chrom19\n",
      "finished chrom1\n",
      "finished chrom20\n",
      "finished chrom21\n",
      "finished chrom22\n",
      "finished chrom2\n",
      "finished chrom3\n",
      "finished chrom4\n",
      "finished chrom5\n",
      "finished chrom6\n",
      "finished chrom7\n",
      "finished chrom8\n",
      "finished chrom9\n",
      "finished chromX\n",
      "finished chromY\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/anat/anaconda2/lib/python2.7/site-packages/IPython/core/interactiveshell.py:2723: DtypeWarning: Columns (16,17,18,19,20,21,22) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  interactivity=interactivity, compiler=compiler, result=result)\n"
     ]
    }
   ],
   "source": [
    "for filename in chrom_files:\n",
    "    #Reading the chromosomel file\n",
    "    chrom_csv = pd.read_csv(my_path+filename, sep='\\t', index_col=0)\n",
    "    chrom_csv.fillna('', inplace=True)\n",
    "    \n",
    "    #Getting all unique Esembl IDs\n",
    "    ensmbl_ids = ((chrom_csv[\"gene\"]).unique()).tolist()\n",
    "    \n",
    "    #Creating a dictionary of ensembl id mapped to hugo symbol\n",
    "    ens_hugo_dict = {}\n",
    "    not_found = []\n",
    "    for ens in ensmbl_ids:\n",
    "        if ens.startswith(\"ENSG\") == False:\n",
    "            not_found.append(ens)\n",
    "            continue\n",
    "\n",
    "        #Querying the Biomart database to get number of matches\n",
    "        num = ens_genes.count({\n",
    "            'filters': {\n",
    "                    'ensembl_gene_id': ens\n",
    "            },\n",
    "            'attributes': [\n",
    "                    'hgnc_symbol'\n",
    "            ]\n",
    "        })\n",
    "\n",
    "        #In case no match is found\n",
    "        if (num == 0):\n",
    "            not_found.append(ens)\n",
    "            continue\n",
    "\n",
    "        if (num > 1):\n",
    "            print \"more than 1 match for \"+ens\n",
    "            print num\n",
    "\n",
    "        #In case of at least one match: querying the Biomart database\n",
    "        response = ens_genes.search({\n",
    "            'filters': {\n",
    "                    'ensembl_gene_id': ens\n",
    "            },\n",
    "            'attributes': [\n",
    "                    'hgnc_symbol'\n",
    "            ]\n",
    "        })\n",
    "\n",
    "        #Reading the query result\n",
    "        for line in response.iter_lines():\n",
    "            line = line.decode('utf-8')\n",
    "            hugo_sym = (line.split(\"\\t\")[0]).encode('ascii','ignore')\n",
    "            break\n",
    "\n",
    "        #Updating the dictionary\n",
    "        ens_hugo_dict[ens] = hugo_sym\n",
    "        \n",
    "    #Create a mapping of Ensembl-Hugo and save to data-frame\n",
    "    ensmbl_col = []\n",
    "    hugo_col = []\n",
    "    for ens in ens_hugo_dict.keys():\n",
    "        ensmbl_col.append(ens)\n",
    "        hugo_col.append(ens_hugo_dict[ens])\n",
    "    mapping = pd.DataFrame({\"Ensembl_gene_id\": ensmbl_col, \"Hugo_symbol\": hugo_col})\n",
    "    \n",
    "    #Save to file\n",
    "    chrom_number = filename[filename.find(\"chrom\"):-4]\n",
    "    mapping.to_csv(curr_dir[0]+\"/ensembl-hugo_map/ids_map_\"+chrom_number+\".csv\", sep='\\t')\n",
    "    \n",
    "    print \"finished \"+chrom_number"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {
    "collapsed": false,
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "#Creating a dictionary of ensembl id mapped to hugo symbol\n",
    "ens_hugo_dict = {}\n",
    "not_found = 0\n",
    "for ens in ensmbl_ids:\n",
    "    if ens.startswith(\"ENS\") == False:\n",
    "        not_found += 1\n",
    "        continue\n",
    "    \n",
    "    #Querying the Biomart database to get number of matches\n",
    "    num = ens_genes.count({\n",
    "        'filters': {\n",
    "                'ensembl_gene_id': ens\n",
    "        },\n",
    "        'attributes': [\n",
    "                'hgnc_symbol'\n",
    "        ]\n",
    "    })\n",
    "    \n",
    "    #In case no match is found\n",
    "    if (num == 0):\n",
    "        print \"No match for \"+ens\n",
    "        not_found += 1\n",
    "        continue\n",
    "    \n",
    "    if (num > 1):\n",
    "        print \"more than 1 match for \"+ens\n",
    "        print num\n",
    "    \n",
    "    #In case of at least one match: querying the Biomart database\n",
    "    response = ens_genes.search({\n",
    "        'filters': {\n",
    "                'ensembl_gene_id': ens\n",
    "        },\n",
    "        'attributes': [\n",
    "                'hgnc_symbol'\n",
    "        ]\n",
    "    })\n",
    "    \n",
    "    #Reading the query result\n",
    "    for line in response.iter_lines():\n",
    "        line = line.decode('utf-8')\n",
    "        hugo_sym = (line.split(\"\\t\")[0]).encode('ascii','ignore')\n",
    "        break\n",
    "    \n",
    "    #Updating the dictionary\n",
    "    ens_hugo_dict[ens] = hugo_sym"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>pos</th>\n",
       "      <th>id</th>\n",
       "      <th>ref</th>\n",
       "      <th>alt</th>\n",
       "      <th>qual</th>\n",
       "      <th>filter</th>\n",
       "      <th>AC</th>\n",
       "      <th>AF</th>\n",
       "      <th>AN</th>\n",
       "      <th>gene</th>\n",
       "      <th>conseq</th>\n",
       "      <th>prot_pos</th>\n",
       "      <th>amino_acids</th>\n",
       "      <th>codons</th>\n",
       "      <th>SWISSPROT</th>\n",
       "      <th>SIFT</th>\n",
       "      <th>domains</th>\n",
       "      <th>clin_sig</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>263</th>\n",
       "      <td>1</td>\n",
       "      <td>664678</td>\n",
       "      <td>.</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>562.11</td>\n",
       "      <td>VQSRTrancheSNP99.60to99.80</td>\n",
       "      <td>8</td>\n",
       "      <td>7.672e-04</td>\n",
       "      <td>10428</td>\n",
       "      <td>2349</td>\n",
       "      <td>31</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0;GQ_HIST=51</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>477764</th>\n",
       "      <td>1</td>\n",
       "      <td>146253158</td>\n",
       "      <td>.</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>199.75</td>\n",
       "      <td>VQSRTrancheSNP99.80to99.90</td>\n",
       "      <td>2</td>\n",
       "      <td>5.377e-05</td>\n",
       "      <td>37196</td>\n",
       "      <td>2349</td>\n",
       "      <td>2299</td>\n",
       "      <td>456</td>\n",
       "      <td>231</td>\n",
       "      <td>134</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0;GQ_HIST=1472</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        chrom        pos id ref alt    qual                      filter AC  \\\n",
       "263         1     664678  .   G   A  562.11  VQSRTrancheSNP99.60to99.80  8   \n",
       "477764      1  146253158  .   C   G  199.75  VQSRTrancheSNP99.80to99.90  2   \n",
       "\n",
       "               AF     AN  gene conseq prot_pos amino_acids codons SWISSPROT  \\\n",
       "263     7.672e-04  10428  2349     31        1           0      0         2   \n",
       "477764  5.377e-05  37196  2349   2299      456         231    134         1   \n",
       "\n",
       "       SIFT domains        clin_sig  \n",
       "263       0       0    0;GQ_HIST=51  \n",
       "477764    0       0  0;GQ_HIST=1472  "
      ]
     },
     "execution_count": 47,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "chrom_csv[chrom_csv[\"gene\"] == \"2349\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "#Create a mapping of Ensembl-Hugo\n",
    "ensmbl_col = []\n",
    "hugo_col = []\n",
    "for ens in ens_hugo_dict.keys():\n",
    "    ensmbl_col.append(ens)\n",
    "    hugo_col.append(ens_hugo_dict[ens])\n",
    "mapping = pd.DataFrame({\"Ensembl_gene_id\": ensmbl_col, \"Hugo_symbol\": hugo_col})\n",
    "mapping.to_csv(curr_dir[0]+\"/ensembl-hugo_map/ids_map_1.csv\", sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Ensembl_gene_id</th>\n",
       "      <th>Hugo_symbol</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000064703</td>\n",
       "      <td>DDX20</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000159363</td>\n",
       "      <td>ATP13A2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000196482</td>\n",
       "      <td>ESRRG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000117215</td>\n",
       "      <td>PLA2G2D</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000086015</td>\n",
       "      <td>MAST2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>ENSG00000142675</td>\n",
       "      <td>CNKSR1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>ENSG00000142676</td>\n",
       "      <td>RPL11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>ENSG00000142677</td>\n",
       "      <td>IL22RA1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>ENSG00000215883</td>\n",
       "      <td>CYB5RL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>ENSG00000179743</td>\n",
       "      <td>FBLIM1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>ENSG00000163216</td>\n",
       "      <td>SPRR2D</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>ENSG00000143217</td>\n",
       "      <td>PVRL4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>ENSG00000156150</td>\n",
       "      <td>ALX3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>ENSG00000237276</td>\n",
       "      <td>ANO7P1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>ENSG00000024526</td>\n",
       "      <td>DEPDC1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>ENSG00000215864</td>\n",
       "      <td>NBPF7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>ENSG00000215860</td>\n",
       "      <td>PDZK1P1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>ENSG00000215861</td>\n",
       "      <td>PPIAL4B</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>ENSG00000175707</td>\n",
       "      <td>C1orf172</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>ENSG00000143079</td>\n",
       "      <td>CTTNBP2NL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>ENSG00000252947</td>\n",
       "      <td>SCARNA1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>ENSG00000143858</td>\n",
       "      <td>SYT2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>ENSG00000143851</td>\n",
       "      <td>PTPN7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>ENSG00000143850</td>\n",
       "      <td>PLEKHA6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>ENSG00000160766</td>\n",
       "      <td>GBAP1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>ENSG00000160767</td>\n",
       "      <td>FAM189B</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>ENSG00000198520</td>\n",
       "      <td>C1orf228</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>ENSG00000084628</td>\n",
       "      <td>NKAIN1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>ENSG00000183291</td>\n",
       "      <td>SH3GLB1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>ENSG00000185946</td>\n",
       "      <td>RNPC3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2202</th>\n",
       "      <td>ENSG00000182901</td>\n",
       "      <td>RGS7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2203</th>\n",
       "      <td>ENSG00000155380</td>\n",
       "      <td>SLC16A1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2204</th>\n",
       "      <td>ENSG00000228208</td>\n",
       "      <td>C1orf143</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2205</th>\n",
       "      <td>ENSG00000026751</td>\n",
       "      <td>SLAMF7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2206</th>\n",
       "      <td>ENSG00000117148</td>\n",
       "      <td>ACTL8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2207</th>\n",
       "      <td>ENSG00000224699</td>\n",
       "      <td>LAMTOR5-AS1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2208</th>\n",
       "      <td>ENSG00000081248</td>\n",
       "      <td>CACNA1S</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2209</th>\n",
       "      <td>ENSG00000143740</td>\n",
       "      <td>SNAP47</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2210</th>\n",
       "      <td>ENSG00000117143</td>\n",
       "      <td>UAP1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2211</th>\n",
       "      <td>ENSG00000120341</td>\n",
       "      <td>SEC16B</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2212</th>\n",
       "      <td>ENSG00000197147</td>\n",
       "      <td>LRRC8B</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2213</th>\n",
       "      <td>ENSG00000230124</td>\n",
       "      <td>LHX4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2214</th>\n",
       "      <td>ENSG00000197921</td>\n",
       "      <td>HES5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2215</th>\n",
       "      <td>ENSG00000117360</td>\n",
       "      <td>PRPF3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2216</th>\n",
       "      <td>ENSG00000117362</td>\n",
       "      <td>APH1A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2217</th>\n",
       "      <td>ENSG00000231437</td>\n",
       "      <td>KCND3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2218</th>\n",
       "      <td>ENSG00000235750</td>\n",
       "      <td>KIAA0040</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2219</th>\n",
       "      <td>ENSG00000197241</td>\n",
       "      <td>SLC2A7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2220</th>\n",
       "      <td>ENSG00000143457</td>\n",
       "      <td>GOLPH3L</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2221</th>\n",
       "      <td>ENSG00000134256</td>\n",
       "      <td>CD101</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2222</th>\n",
       "      <td>ENSG00000143107</td>\n",
       "      <td>FNDC7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2223</th>\n",
       "      <td>ENSG00000143106</td>\n",
       "      <td>PSMA5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2224</th>\n",
       "      <td>ENSG00000134253</td>\n",
       "      <td>TRIM45</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2225</th>\n",
       "      <td>ENSG00000143452</td>\n",
       "      <td>HORMAD1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2226</th>\n",
       "      <td>ENSG00000143450</td>\n",
       "      <td>OAZ3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2227</th>\n",
       "      <td>ENSG00000223416</td>\n",
       "      <td>RPS26P15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2228</th>\n",
       "      <td>ENSG00000097033</td>\n",
       "      <td>SH3GLB1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2229</th>\n",
       "      <td>ENSG00000134259</td>\n",
       "      <td>NGF</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2230</th>\n",
       "      <td>ENSG00000143458</td>\n",
       "      <td>GABPB2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2231</th>\n",
       "      <td>ENSG00000179914</td>\n",
       "      <td>ITLN1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2232 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      Ensembl_gene_id  Hugo_symbol\n",
       "0     ENSG00000064703        DDX20\n",
       "1     ENSG00000159363      ATP13A2\n",
       "2     ENSG00000196482        ESRRG\n",
       "3     ENSG00000117215      PLA2G2D\n",
       "4     ENSG00000086015        MAST2\n",
       "5     ENSG00000142675       CNKSR1\n",
       "6     ENSG00000142676        RPL11\n",
       "7     ENSG00000142677      IL22RA1\n",
       "8     ENSG00000215883       CYB5RL\n",
       "9     ENSG00000179743       FBLIM1\n",
       "10    ENSG00000163216       SPRR2D\n",
       "11    ENSG00000143217        PVRL4\n",
       "12    ENSG00000156150         ALX3\n",
       "13    ENSG00000237276       ANO7P1\n",
       "14    ENSG00000024526       DEPDC1\n",
       "15    ENSG00000215864        NBPF7\n",
       "16    ENSG00000215860      PDZK1P1\n",
       "17    ENSG00000215861      PPIAL4B\n",
       "18    ENSG00000175707     C1orf172\n",
       "19    ENSG00000143079    CTTNBP2NL\n",
       "20    ENSG00000252947      SCARNA1\n",
       "21    ENSG00000143858         SYT2\n",
       "22    ENSG00000143851        PTPN7\n",
       "23    ENSG00000143850      PLEKHA6\n",
       "24    ENSG00000160766        GBAP1\n",
       "25    ENSG00000160767      FAM189B\n",
       "26    ENSG00000198520     C1orf228\n",
       "27    ENSG00000084628       NKAIN1\n",
       "28    ENSG00000183291      SH3GLB1\n",
       "29    ENSG00000185946        RNPC3\n",
       "...               ...          ...\n",
       "2202  ENSG00000182901         RGS7\n",
       "2203  ENSG00000155380      SLC16A1\n",
       "2204  ENSG00000228208     C1orf143\n",
       "2205  ENSG00000026751       SLAMF7\n",
       "2206  ENSG00000117148        ACTL8\n",
       "2207  ENSG00000224699  LAMTOR5-AS1\n",
       "2208  ENSG00000081248      CACNA1S\n",
       "2209  ENSG00000143740       SNAP47\n",
       "2210  ENSG00000117143         UAP1\n",
       "2211  ENSG00000120341       SEC16B\n",
       "2212  ENSG00000197147       LRRC8B\n",
       "2213  ENSG00000230124         LHX4\n",
       "2214  ENSG00000197921         HES5\n",
       "2215  ENSG00000117360        PRPF3\n",
       "2216  ENSG00000117362        APH1A\n",
       "2217  ENSG00000231437        KCND3\n",
       "2218  ENSG00000235750     KIAA0040\n",
       "2219  ENSG00000197241       SLC2A7\n",
       "2220  ENSG00000143457      GOLPH3L\n",
       "2221  ENSG00000134256        CD101\n",
       "2222  ENSG00000143107        FNDC7\n",
       "2223  ENSG00000143106        PSMA5\n",
       "2224  ENSG00000134253       TRIM45\n",
       "2225  ENSG00000143452      HORMAD1\n",
       "2226  ENSG00000143450         OAZ3\n",
       "2227  ENSG00000223416     RPS26P15\n",
       "2228  ENSG00000097033      SH3GLB1\n",
       "2229  ENSG00000134259          NGF\n",
       "2230  ENSG00000143458       GABPB2\n",
       "2231  ENSG00000179914        ITLN1\n",
       "\n",
       "[2232 rows x 2 columns]"
      ]
     },
     "execution_count": 80,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mapping"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
