{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>.container { width:100% !important; }</style>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import math\n",
    "from collections import defaultdict\n",
    "from os import environ\n",
    "import pickle\n",
    "import fileinput\n",
    "import sys\n",
    "import datetime\n",
    "from mapping_func import create_exon_pos_table, find_chrom_bps, protein_pos_to_hmm_state_and_aa\n",
    "from calc_exac_freq_func import create_alt_codon, exac_validation_checks, retrieve_codon_seq, codon_table\n",
    "from indels_func import is_indel, table_editing, indel_type\n",
    "from entropy_func import JSD_background, JSD, SE, JSdiv\n",
    "from af_format_calc import format_af, calculate_af_adj\n",
    "from IPython.core.display import HTML\n",
    "HTML(\"<style>.container { width:100% !important; }</style>\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/anat/anaconda2/lib/python2.7/site-packages/IPython/config.py:13: ShimWarning: The `IPython.config` package has been deprecated. You should import from traitlets.config instead.\n",
      "  \"You should import from traitlets.config instead.\", ShimWarning)\n"
     ]
    }
   ],
   "source": [
    "#Trying to fix the problem of: \"The history saving thread hit an unexpected error\" when running on the server\n",
    "#as described here: https://github.com/ipython/ipython/issues/1342\n",
    "from IPython.config.loader import Config\n",
    "from IPython.core.interactiveshell import InteractiveShell\n",
    "\n",
    "cfg = Config()\n",
    "cfg.HistoryManager.hist_file = ':memory:'\n",
    "ip = InteractiveShell.instance(config=cfg)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "MHC_II_beta\n"
     ]
    }
   ],
   "source": [
    "#Getting path\n",
    "curr_dir = !pwd\n",
    "\n",
    "#Reading the list of filtered domains\n",
    "with open(curr_dir[0]+\"/../5.domains_stats/filtered10_list.pik\", 'rb') as handle:\n",
    "    filtered_domains_list10 = pickle.load(handle)\n",
    "filtered_domains_list10.sort()\n",
    "\n",
    "#Getting the domain index as environment variable called \"idx\"\n",
    "try:\n",
    "    domain_index = int(environ['idx'])\n",
    "except:\n",
    "    domain_index = 2\n",
    "\n",
    "domain_name = filtered_domains_list10[domain_index]\n",
    "\n",
    "print domain_name"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "in_path = curr_dir[0]+\"/../3.parse_HMMER/hmm_domains/pfam-v30/\"\n",
    "filename = domain_name+\".csv\"\n",
    "domain_data = pd.read_csv(in_path+filename, sep='\\t', index_col=0, dtype={\"chrom_num\": str})\n",
    "#Sort the domain data\n",
    "sorted_domain_data = domain_data.sort_values(by=[\"chrom_num\", \"gene\", \"TargetStart\"])\n",
    "sorted_domain_data = sorted_domain_data.reset_index(drop=True)\n",
    "\n",
    "#Get the canonic protein id for Zinc domain\n",
    "with open(curr_dir[0]+\"/../4.parse_Uniprot/domains_canonic_prot/pfam-v30/\"+domain_name+\"_canonic_prot.pik\", 'rb') as handle:\n",
    "    canonic_protein = pickle.load(handle)\n",
    "    \n",
    "#Get the filtered table of domains\n",
    "    \n",
    "chromosome_names = [\"1\", \"2\", \"3\", \"4\", \"5\", \"6\", \"7\", \"8\", \"9\", \"10\", \"11\", \"12\", \"13\", \"14\", \"15\", \"16\", \"17\", \"18\", \"19\", \"20\", \"21\", \"22\", \"X\", \"Y\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Leaving outside 108 out of 124\n"
     ]
    }
   ],
   "source": [
    "#Counting how many domains instances are excluded because of strange chrom names.\n",
    "chrom_names_list = sorted_domain_data[\"chrom_num\"].tolist()\n",
    "strange_chrom_sum = 0\n",
    "for name in chrom_names_list:\n",
    "    if (name not in chromosome_names):\n",
    "        strange_chrom_sum += 1\n",
    "print \"Leaving outside \"+str(strange_chrom_sum)+\" out of \"+str(len(chrom_names_list))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "def stop_codon_notation(aa):\n",
    "    \"\"\"Coverting the HMMER 'x' char to '*' for stop codon notation unifirmity\"\"\"\n",
    "    if (aa == 'X'):\n",
    "        return \"*\"\n",
    "    else:\n",
    "        return aa"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "def validate_HMMER_hg19_codon(bp_ref, aa, chrom_pos_list, protein_pos, hmm_state):\n",
    "    \"\"\"Validation: checking that the returned codon sequence from hg19 match the HMMER amino-acid\"\"\"\n",
    "\n",
    "    #For error logging\n",
    "    functionNameAsString = sys._getframe().f_code.co_name\n",
    "    \n",
    "    translated_aa = codon_table[bp_ref.upper()]\n",
    "    if (translated_aa != aa):\n",
    "        print \"chrom_pos_list = \"+str(chrom_pos_list)+\" protein_pos = \"+str(protein_pos)+\" hmm_state = \"+str(hmm_state)\n",
    "        print functionNameAsString+\" Error: hg19 codon sequence retrieved \"+bp_ref.upper()+\"=\"+translated_aa+\" doesn't match HMMER amino-acid \"+aa"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def validate_exac_aa(exac_prot_data, exac_alt_aa, alt_aa, chrom_pos):\n",
    "    \"\"\"Validation: checking if the calculated aa matches the ExAC aa\"\"\"\n",
    "    #For error logging\n",
    "    functionNameAsString = sys._getframe().f_code.co_name\n",
    "    \n",
    "    if (exac_prot_data and exac_alt_aa != alt_aa):\n",
    "        print functionNameAsString+\" \"+ str(chrom_pos)+\" Error: the ExAC alt aa \"+exac_alt_aa+\" doesn't match my alt aa calculation \"+alt_aa"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def change_ref_aa(res_dict, alterations_af_dict, alterations_af_adj_dict, aa, aa_sum, aa_adj_sum, bp_af_dict, bp_af_adj_dict):\n",
    "    \"\"\"Changing the ref aa and updating all relevant dictionaries\"\"\"\n",
    "    \n",
    "    #Adding the refrence allele to the alterations dicts\n",
    "    old_ref = res_dict[\"aa_ref\"]\n",
    "    sum_of_all_alt = sum(sum(alterations_af_dict.values(), []))\n",
    "    sum_of_all_alt_adj = sum(sum(alterations_af_adj_dict.values(), []))\n",
    "    alterations_af_dict[old_ref] = [1 - sum_of_all_alt]\n",
    "    alterations_af_adj_dict[old_ref] = [1 - sum_of_all_alt_adj]\n",
    "\n",
    "    #Updating the aa to be the ref\n",
    "    res_dict[\"aa_ref\"] = aa\n",
    "\n",
    "    #Finding the codon that codes aa with highest frequency and update bp_ref\n",
    "    old_bp_ref = res_dict[\"bp_ref\"]\n",
    "    max_af = 0\n",
    "    for codon in (bp_af_adj_dict.keys()):\n",
    "        if (codon_table[codon.upper()] == aa):\n",
    "            if (bp_af_adj_dict[codon] >= max_af):\n",
    "                max_af = bp_af_adj_dict[codon]\n",
    "                res_dict[\"bp_ref\"] = codon\n",
    "                \n",
    "    #Adding the ref bp to the bp dicts\n",
    "    bp_af_dict[old_bp_ref] = res_dict[\"af\"]\n",
    "    bp_af_adj_dict[old_bp_ref] = res_dict[\"af_adj\"]\n",
    "    \n",
    "    #Updating the Frequencies of ref\n",
    "    res_dict[\"af\"] =(1 - aa_sum)\n",
    "    res_dict[\"af_adj\"] = (1 - aa_adj_sum)\n",
    "\n",
    "    #Deleting from the alterations dict\n",
    "    del alterations_af_dict[aa]\n",
    "    del alterations_af_adj_dict[aa]\n",
    "    \n",
    "    #Deleting from the bps dict\n",
    "    del bp_af_dict[res_dict[\"bp_ref\"]]\n",
    "    del bp_af_adj_dict[res_dict[\"bp_ref\"]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def change_syn_ref_bp(new_bp_ref, old_bp_ref, res_dict, bp_af_dict, bp_af_adj_dict):\n",
    "    \n",
    "    #Updating bp_ref with the high freq. codon\n",
    "    res_dict[\"bp_ref\"] = new_bp_ref\n",
    "    \n",
    "    #Add the old ref to the bp dicts with its freq.\n",
    "    bp_freq_adj_sum = sum(bp_af_adj_dict.values())\n",
    "    bp_freq_sum = sum(bp_af_dict.values())\n",
    "    bp_af_adj_dict[old_bp_ref] = (1 - bp_freq_adj_sum)\n",
    "    bp_af_dict[old_bp_ref] = (1 - bp_freq_sum)\n",
    "    \n",
    "    #Delete the new ref from the bp dicts\n",
    "    del bp_af_adj_dict[new_bp_ref]\n",
    "    del bp_af_dict[new_bp_ref]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "#A function that return a dict with the MAF info for the protein position and corresponding chromosomal location\n",
    "def calc_exac_maf_data(chrom_pos_list, chrom_gene_table, indels_table, protein_pos, aa, chrom_raw_data, chrom, hmm_state):\n",
    "    \n",
    "    #For error logging\n",
    "    functionNameAsString = sys._getframe().f_code.co_name\n",
    "    \n",
    "    #Initializing the results dictionary\n",
    "    res_dict = {}\n",
    "    res_dict[\"chrom\"] = chrom\n",
    "    res_dict[\"chrom_pos\"] = chrom_pos_list\n",
    "    res_dict[\"prot_pos\"] = protein_pos\n",
    "    res_dict[\"aa_ref\"] = stop_codon_notation(aa)\n",
    "    res_dict[\"bp_ref\"] = retrieve_codon_seq(chrom_pos_list, chrom_raw_data, chrom)\n",
    "    #an counters lists\n",
    "    an_dict = {k: [] for k in [\"an\", \"an_adj\", \"an_afr\", \"an_amr\", \"an_eas\", \"an_fin\", \"an_nfe\", \"an_oth\", \"an_sas\"]}\n",
    "    res_dict.update(an_dict)\n",
    "    #ac counters lists\n",
    "    ac_dict = {k: [] for k in [\"ac_adj\", \"ac_afr\", \"ac_amr\", \"ac_eas\", \"ac_fin\", \"ac_het\", \"ac_hom\", \"ac_nfe\", \"ac_oth\", \"ac_sas\"]}\n",
    "    res_dict.update(ac_dict)\n",
    "    res_dict[\"SIFT\"] = []\n",
    "    res_dict[\"PolyPhen\"] = []\n",
    "    res_dict[\"clin_sig\"] = []\n",
    "    res_dict[\"SwissProt\"] = []\n",
    "    res_dict[\"ens_prot\"] = []\n",
    "    \n",
    "    #Initializing more variables\n",
    "    indels_cnt = 0\n",
    "    errors_cnt = 0\n",
    "    filter_cnt = 0\n",
    "    inframe_ids = []\n",
    "    alterations_af_dict = defaultdict(list)\n",
    "    alterations_af_adj_dict = defaultdict(list)\n",
    "    bp_af_dict = dict()\n",
    "    bp_af_adj_dict = dict()\n",
    "    bp_list = [] #Will not update when changing ref aa/bp, has to correspond to the population prob.\n",
    "    \n",
    "    #Validation: checking that the returned codon sequence from hg19 match the HMMER amino-acid\n",
    "    validate_HMMER_hg19_codon(res_dict[\"bp_ref\"], aa, chrom_pos_list, protein_pos, hmm_state)\n",
    "    \n",
    "    #Going over all 3 codon positions\n",
    "    for i in range(len(chrom_pos_list)):\n",
    "        chrom_pos = chrom_pos_list[i]\n",
    "        alt_codon_pos = i\n",
    "            \n",
    "        #Retreiving relevant ExAC entry\n",
    "        chrom_alter_table = chrom_gene_table[chrom_gene_table[\"pos\"] == chrom_pos]\n",
    "        chrom_alter_table = chrom_alter_table.reset_index(drop=True)\n",
    "                \n",
    "        if (chrom_alter_table.shape[0] == 0):\n",
    "            #No ExAC entry for this chromosome position - not adding alteration data\n",
    "            continue\n",
    "        \n",
    "        else:\n",
    "            #In case there are several alterations for that position, iterating\n",
    "            for index, line in chrom_alter_table.iterrows():\n",
    "                chrom_alter = line\n",
    "                \n",
    "                #Skipping alterations that were filtered out by VQSR\n",
    "                if (chrom_alter[\"filter\"] != \"PASS\"):\n",
    "                    filter_cnt += 1\n",
    "                    continue\n",
    "                \n",
    "                #Extracting ref and alt\n",
    "                exac_ref_bp = chrom_alter[\"ref\"]\n",
    "                exac_alt_bp = chrom_alter[\"alt\"]\n",
    "                \n",
    "                #Check if indel - skip (we assume the whole protein may not function and don't add those to the MAF count)\n",
    "                if (is_indel(exac_ref_bp, exac_alt_bp, chrom_alter) != indel_type.NO_INDEL):\n",
    "                    indels_cnt += 1\n",
    "                    continue\n",
    "                \n",
    "                #Perform validation checks (comparing ExAC and HMMER data)\n",
    "                (exac_prot_data, exac_alt_aa, exac_alt_codon, errors) = exac_validation_checks(chrom_alter, protein_pos, aa, alt_codon_pos, chrom_pos, res_dict[\"bp_ref\"])\n",
    "                if (errors):\n",
    "                    errors_cnt += 1\n",
    "\n",
    "                #Extracting ExAC allele frequency data\n",
    "                af = chrom_alter[\"AF\"]\n",
    "                an = int(chrom_alter[\"AN\"])\n",
    "                an_adj = int(chrom_alter[\"AN_Adj\"])\n",
    "                ac_adj = chrom_alter[\"AC_Adj\"]\n",
    "                \n",
    "                #Calculating the alteration relevant data\n",
    "                alt_codon = create_alt_codon (exac_ref_bp, exac_alt_bp, res_dict[\"bp_ref\"], alt_codon_pos, chrom_raw_data)   \n",
    "                if (len(alt_codon) != 3):\n",
    "                    continue \n",
    "                else:\n",
    "                    alt_aa = codon_table[alt_codon.upper()]\n",
    "                \n",
    "                #Validation: ExAC aa match the calculated alt\n",
    "                validate_exac_aa(exac_prot_data, exac_alt_aa, alt_aa, chrom_pos)\n",
    "\n",
    "                #Calculating the allele frequency adjusted\n",
    "                af_adj = calculate_af_adj(an_adj, ac_adj)\n",
    "                \n",
    "                #Saving the bp with the frequency (for both syn and nonsyn)\n",
    "                bp_af_dict[alt_codon] = af\n",
    "                bp_af_adj_dict[alt_codon] = af_adj\n",
    "                bp_list.append(alt_codon)\n",
    "                \n",
    "                res_dict[\"an\"].append(an)\n",
    "                res_dict[\"an_adj\"].append(chrom_alter[\"AN_Adj\"])\n",
    "                res_dict[\"an_afr\"].append(chrom_alter[\"AN_AFR\"])\n",
    "                res_dict[\"an_amr\"].append(chrom_alter[\"AN_AMR\"])\n",
    "                res_dict[\"an_eas\"].append(chrom_alter[\"AN_EAS\"])\n",
    "                res_dict[\"an_fin\"].append(chrom_alter[\"AN_FIN\"])\n",
    "                res_dict[\"an_nfe\"].append(chrom_alter[\"AN_NFE\"])\n",
    "                res_dict[\"an_oth\"].append(chrom_alter[\"AN_OTH\"])\n",
    "                res_dict[\"an_sas\"].append(chrom_alter[\"AN_SAS\"])\n",
    "                \n",
    "                res_dict[\"ac_adj\"].append(chrom_alter[\"AC_Adj\"])\n",
    "                res_dict[\"ac_afr\"].append(chrom_alter[\"AC_AFR\"])\n",
    "                res_dict[\"ac_amr\"].append(chrom_alter[\"AC_AMR\"])\n",
    "                res_dict[\"ac_eas\"].append(chrom_alter[\"AC_EAS\"])\n",
    "                res_dict[\"ac_fin\"].append(chrom_alter[\"AC_FIN\"])\n",
    "                res_dict[\"ac_het\"].append(chrom_alter[\"AC_Het\"])\n",
    "                res_dict[\"ac_hom\"].append(chrom_alter[\"AC_Hom\"])\n",
    "                res_dict[\"ac_nfe\"].append(chrom_alter[\"AC_NFE\"])\n",
    "                res_dict[\"ac_oth\"].append(chrom_alter[\"AC_OTH\"])\n",
    "                res_dict[\"ac_sas\"].append(chrom_alter[\"AC_SAS\"])\n",
    "                \n",
    "                #Non-synonymous(!!!) - logging the alteration in the dictionary\n",
    "                if (alt_aa != res_dict[\"aa_ref\"]):\n",
    "                    alterations_af_dict[alt_aa].append(float(af))\n",
    "                    alterations_af_adj_dict[alt_aa].append(af_adj)\n",
    "                    \n",
    "                #Saving the SIFT and PolyPhen scores (for both syn and nonsyn)\n",
    "                res_dict[\"SIFT\"].append(chrom_alter[\"SIFT\"])\n",
    "                res_dict[\"PolyPhen\"].append(chrom_alter[\"PolyPhen\"])\n",
    "                res_dict[\"clin_sig\"].append(chrom_alter[\"clin_sig\"])\n",
    "                \n",
    "                #Saving SwissProt id and Ensembl prot id\n",
    "                res_dict[\"SwissProt\"].append(chrom_alter[\"SWISSPROT\"])\n",
    "                res_dict[\"ens_prot\"].append(chrom_alter[\"ENSP\"])\n",
    "                \n",
    "\n",
    "    #Calculating the overall MAF from the alteration dicts\n",
    "    res_dict[\"af\"] = 0\n",
    "    res_dict[\"af_adj\"] = 0\n",
    "    \n",
    "    res_dict[\"aa_ref_orig\"] = res_dict[\"aa_ref\"]\n",
    "    for aa in alterations_af_dict.keys():\n",
    "        aa_sum = sum(alterations_af_dict[aa])\n",
    "        aa_adj_sum = sum(alterations_af_adj_dict[aa])\n",
    "        \n",
    "        #Checking if any alteration is above 0.5, and changing the ref accordingly\n",
    "        if (aa_sum > 0.5):\n",
    "            \n",
    "            #Update all relevent information to switching a ref aa\n",
    "            change_ref_aa(res_dict, alterations_af_dict, alterations_af_adj_dict, aa, aa_sum, aa_adj_sum, bp_af_dict, bp_af_adj_dict)\n",
    "            break\n",
    "        else:\n",
    "            res_dict[\"af\"] += aa_sum\n",
    "            res_dict[\"af_adj\"] += aa_adj_sum\n",
    "        \n",
    "        #Fix the AF format\n",
    "        res_dict[\"af\"] = format_af(res_dict[\"af\"])\n",
    "        res_dict[\"af_adj\"] = format_af(res_dict[\"af_adj\"])\n",
    "    \n",
    "    #Checking if any syn aleration bp is above 0.5 (nonsyn were already checked), and changing ref bp accordingly\n",
    "    for codon in bp_af_adj_dict.keys():\n",
    "        if ((codon_table[codon] == res_dict[\"aa_ref\"]) and (bp_af_adj_dict[codon] > 0.5)):\n",
    "            new_bp_ref = codon\n",
    "            old_bp_ref = res_dict[\"bp_ref\"]\n",
    "            change_syn_ref_bp(new_bp_ref, old_bp_ref, res_dict, bp_af_dict, bp_af_adj_dict)\n",
    "            \n",
    "        \n",
    "    res_dict[\"alterations_af_adj_dict\"] = alterations_af_adj_dict\n",
    "    res_dict[\"bp_af_dict\"] = bp_af_dict\n",
    "    res_dict[\"bp_af_adj_dict\"] = bp_af_adj_dict\n",
    "    res_dict[\"bp_list\"] = bp_list\n",
    "    \n",
    "    return (res_dict, indels_cnt, errors_cnt, filter_cnt)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Starting....\n",
      "Finished chromosome 1\n",
      "Finished chromosome 2\n",
      "Finished chromosome 3\n",
      "Finished chromosome 4\n",
      "Finished chromosome 5\n",
      "Finished protein ENSG00000179344.12\n",
      "Finished protein ENSG00000196126.6\n",
      "Finished protein ENSG00000198502.5\n",
      "Finished protein ENSG00000223865.6\n",
      "Finished protein ENSG00000232629.4\n",
      "Finished protein ENSG00000241106.2\n",
      "Finished protein ENSG00000242574.4\n",
      "Finished chromosome 6\n",
      "Finished chromosome 7\n",
      "Finished chromosome 8\n",
      "Finished chromosome 9\n",
      "Finished chromosome 10\n",
      "Finished chromosome 11\n",
      "Finished chromosome 12\n",
      "Finished chromosome 13\n",
      "Finished chromosome 14\n",
      "Finished chromosome 15\n",
      "Finished chromosome 16\n",
      "Finished chromosome 17\n",
      "Finished chromosome 18\n",
      "Finished chromosome 19\n",
      "Finished chromosome 20\n",
      "Finished chromosome 21\n",
      "Finished chromosome 22\n",
      "Finished chromosome X\n",
      "Finished chromosome Y\n",
      "CPU times: user 4min 45s, sys: 3min 33s, total: 8min 19s\n",
      "Wall time: 11min 3s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "\n",
    "chrom_path = curr_dir[0]+\"/../1.parse_ExAC/parsed/\"\n",
    "chrom_filename = \"parsed_chrom\"\n",
    "states_dict = defaultdict(list)\n",
    "print \"Starting....\"\n",
    "\n",
    "#For error logging\n",
    "functionNameAsString = sys._getframe().f_code.co_name\n",
    "\n",
    "#A list of all the ens genes\n",
    "domain_ens_genes_all = []\n",
    "\n",
    "#A list to count indels per gene\n",
    "domain_ens_genes_indels = []\n",
    "\n",
    "#A list to count validation errors per gene\n",
    "domain_ens_genes_errors = []\n",
    "\n",
    "#A list to count ExAC filtered-out per gene\n",
    "domain_ens_genes_filter = []\n",
    "\n",
    "for chrom in chromosome_names:\n",
    "    \n",
    "    #Filtering the domain data relevant to this chromosome\n",
    "    domain_chrom_data = sorted_domain_data[sorted_domain_data[\"chrom_num\"] == chrom]\n",
    "    \n",
    "    #Loading the ExAC parsed data of this chromosome\n",
    "    fields = [\"chrom\", \"pos\", \"ref\", \"alt\", \"filter\", \"AC\", \"AC_AFR\", \"AC_AMR\", \"AC_Adj\", \"AC_EAS\", \"AC_FIN\", \"AC_Het\", \"AC_Hom\", \"AC_NFE\", \"AC_OTH\", \"AC_SAS\", \n",
    "              \"AF\", \"AN\", \"AN_AFR\", \"AN_AMR\", \"AN_Adj\", \"AN_EAS\", \"AN_FIN\", \"AN_NFE\", \"AN_OTH\", \"AN_SAS\", \"gene\", \"feature\", \"feature_type\", \"conseq\",\n",
    "              \"prot_pos\", \"amino_acids\", \"codons\", \"strand\", \"ENSP\", \"SWISSPROT\", \"exon\", \"intron\", \"domains\", \"SIFT\", \"PolyPhen\",\"clin_sig\"]\n",
    "    chrom_csv = pd.read_csv(chrom_path+chrom_filename+chrom+\".csv\", sep='\\t', index_col=0, usecols=fields,\n",
    "                           dtype={\"AC\": int, \"AC_AFR\": int, \"AC_AMR\": int, \"AC_Adj\": int, \"AC_EAS\": int,\n",
    "                               \"AC_FIN\": int, \"AC_Het\": int, \"AC_Hom\": int, \"AC_NFE\": int, \"AC_OTH\": int, \"AC_SAS\": int, \"AF\": float, \"AN\": int, \"AN_AFR\": int,\n",
    "                               \"AN_AMR\": int, \"AN_Adj\": int, \"AN_EAS\": int, \"AN_FIN\": int, \"AN_NFE\": int, \"AN_OTH\": int, \"AN_SAS\": int, \"prot_pos\": str})\n",
    "    chrom_csv = chrom_csv.sort_values(by=[\"pos\"])\n",
    "    chrom_csv = chrom_csv.reset_index(drop=True)\n",
    "    chrom_csv.fillna('', inplace=True)\n",
    "    chrom_csv[\"comments\"] = \"\"\n",
    "    \n",
    "    #Getting a list of all the relevant ensembl gene ids for this chromosome\n",
    "    domain_ens_genes = (domain_chrom_data[\"gene\"]).unique()\n",
    "    domain_ens_genes_all.extend(domain_ens_genes)\n",
    "    \n",
    "    #For each ensembl gene in the domain data - finding all the ExAC alterations\n",
    "    for ens_gene in domain_ens_genes:\n",
    "        \n",
    "        #Filtering the domain data for this gene according to the canonical protein id\n",
    "        canonic_prot = canonic_protein[ens_gene]\n",
    "        canonic_prot_t = canonic_prot[:canonic_prot.find(\".\")] #Trimming the \".#\" at the end\n",
    "        domain_gene_table = domain_chrom_data[domain_chrom_data[\"prot\"] == canonic_prot]\n",
    "        #Making sure that if two HMM-matches overlaps, the higher bit score will come first in the table\n",
    "        domain_gene_table = domain_gene_table.sort_values(by=\"BitScore\", ascending=False)\n",
    "        domain_gene_name = domain_gene_table[\"hugoSymbol\"].unique()[0]\n",
    "        if (len(domain_gene_table[\"hugoSymbol\"].unique()) > 1):\n",
    "            print functionNameAsString+\" Error: \"+ens_gene+\": more than one Hugo symbol\"  #sanity check\n",
    "        \n",
    "        #Creating a table of the exons for this gene, according to the canonical protein\n",
    "        chrom_raw_data = domain_gene_table[\"chromosome\"].unique()[0] #there should be only one element here\n",
    "        if (len(domain_gene_table[\"chromosome\"].unique()) > 1):\n",
    "            print functionNameAsString+\" Error: \"+ens_gene+\": more than one chromosome raw data\" #sanity check\n",
    "        targetid = domain_gene_table[\"#TargetID\"].unique()[0]\n",
    "        exon_table = create_exon_pos_table(chrom_raw_data, targetid)\n",
    "        \n",
    "        #Filtering the chromosome data to the gene exons region\n",
    "        exons_start_pos = min(exon_table[\"start_pos\"][0],exon_table[\"start_pos\"][len(exon_table)-1]) #in case of complelemt, the minimal position could be at the last row\n",
    "        exons_end_pos = max(exon_table[\"end_pos\"][0],exon_table[\"end_pos\"][len(exon_table)-1]) #in case of complelemt, the maximal position could be at the first row\n",
    "        chrom_gene_table = chrom_csv[chrom_csv[\"pos\"] >= int(exons_start_pos)][chrom_csv[\"pos\"] <= int(exons_end_pos)][chrom_csv[\"ENSP\"] == canonic_prot_t]\n",
    "        chrom_gene_table = chrom_gene_table.reset_index(drop=True)\n",
    "        \n",
    "        #Handling indels\n",
    "        indels_table = table_editing(chrom_gene_table)\n",
    "        \n",
    "        #A counter for indels inside the domain\n",
    "        protein_indels_cnt = 0\n",
    "        #A counter for validation errors inside the domain\n",
    "        protein_errors_cnt = 0\n",
    "        #A counter for ExAC filter-out inside the domain\n",
    "        protein_filter_cnt = 0\n",
    "        \n",
    "        #Iterating over the amino-acids of the protein\n",
    "        prot_len = int(domain_gene_table[\"length\"].unique()[0])\n",
    "        \n",
    "        for protein_pos in range(1,prot_len+1):\n",
    "    \n",
    "            #Trying to match HMM-state, and retreive the aa from HMMER results\n",
    "            (hmm_state, aa) = protein_pos_to_hmm_state_and_aa(protein_pos, domain_gene_table) #TODO: what happens when two matches overlap? maybe sort to the best bit score?\n",
    "                \n",
    "            #If there's a match to HMM-state: find the corresponding codon bps chromosome positions\n",
    "            if (hmm_state > 0):\n",
    "                chrom_pos_list =find_chrom_bps(protein_pos, exon_table, chrom_raw_data)\n",
    "                \n",
    "                #Analysis of the amino-acid MAF and realted data, returned in a dictionary\n",
    "                (info_dict, indels_cnt, errors_cnt, filter_cnt) = calc_exac_maf_data(chrom_pos_list, chrom_gene_table, indels_table, protein_pos, aa, chrom_raw_data, chrom, hmm_state)\n",
    "                info_dict[\"ens_gene\"] = ens_gene\n",
    "                \n",
    "                #Adding the dictionary to the HMM-state list\n",
    "                states_dict[hmm_state].append(info_dict)\n",
    "                \n",
    "                #Adding the indels to the global counter\n",
    "                protein_indels_cnt += indels_cnt\n",
    "                \n",
    "                #Adding the errors to the global counter\n",
    "                protein_errors_cnt += errors_cnt\n",
    "                \n",
    "                #Adding the filtered to the global counter\n",
    "                protein_filter_cnt += filter_cnt\n",
    "        \n",
    "        domain_ens_genes_indels.append(protein_indels_cnt)\n",
    "        domain_ens_genes_errors.append(protein_errors_cnt)\n",
    "        domain_ens_genes_filter.append(protein_filter_cnt)\n",
    "        print \"Finished protein \"+ens_gene \n",
    "                                \n",
    "    print \"Finished chromosome \"+chrom\n",
    "\n",
    "!mkdir -p domains_states_dicts/pfam-v30/$domain_name\n",
    "with open(curr_dir[0]+\"/domains_states_dicts/pfam-v30/\"+domain_name+\"/\"+domain_name+\"_hmm_states_dict_\"+datetime.date.today().strftime(\"%m.%d.%y\")+\".pik\", 'wb') as handle:\n",
    "    pickle.dump(states_dict, handle, protocol=pickle.HIGHEST_PROTOCOL)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[{'PolyPhen': [],\n",
       "  'SIFT': [],\n",
       "  'SwissProt': [],\n",
       "  'aa_ref': 'P',\n",
       "  'aa_ref_orig': 'P',\n",
       "  'ac_adj': [],\n",
       "  'ac_afr': [],\n",
       "  'ac_amr': [],\n",
       "  'ac_eas': [],\n",
       "  'ac_fin': [],\n",
       "  'ac_het': [],\n",
       "  'ac_hom': [],\n",
       "  'ac_nfe': [],\n",
       "  'ac_oth': [],\n",
       "  'ac_sas': [],\n",
       "  'af': 0,\n",
       "  'af_adj': 0,\n",
       "  'alterations_af_adj_dict': defaultdict(list, {}),\n",
       "  'an': [],\n",
       "  'an_adj': [],\n",
       "  'an_afr': [],\n",
       "  'an_amr': [],\n",
       "  'an_eas': [],\n",
       "  'an_fin': [],\n",
       "  'an_nfe': [],\n",
       "  'an_oth': [],\n",
       "  'an_sas': [],\n",
       "  'bp_af_adj_dict': {},\n",
       "  'bp_af_dict': {},\n",
       "  'bp_list': [],\n",
       "  'bp_ref': 'CCA',\n",
       "  'chrom': '1',\n",
       "  'chrom_pos': (51913765, 51913764, 51913763),\n",
       "  'clin_sig': [],\n",
       "  'ens_gene': 'ENSG00000085832.12',\n",
       "  'ens_prot': [],\n",
       "  'prot_pos': 202},\n",
       " {'PolyPhen': ['probably_damaging(1)', ''],\n",
       "  'SIFT': ['deleterious(0)', ''],\n",
       "  'SwissProt': ['EPS15_HUMAN', 'EPS15_HUMAN'],\n",
       "  'aa_ref': 'P',\n",
       "  'aa_ref_orig': 'P',\n",
       "  'ac_adj': [1, 1],\n",
       "  'ac_afr': [0, 0],\n",
       "  'ac_amr': [0, 0],\n",
       "  'ac_eas': [0, 0],\n",
       "  'ac_fin': [0, 0],\n",
       "  'ac_het': [1, 1],\n",
       "  'ac_hom': [0, 0],\n",
       "  'ac_nfe': [1, 0],\n",
       "  'ac_oth': [0, 0],\n",
       "  'ac_sas': [0, 1],\n",
       "  'af': 8.236e-06,\n",
       "  'af_adj': 8.241e-06,\n",
       "  'alterations_af_adj_dict': defaultdict(list, {'S': [8.241e-06]}),\n",
       "  'an': [121412, 121412],\n",
       "  'an_adj': [121342, 121344],\n",
       "  'an_afr': [10406, 10406],\n",
       "  'an_amr': [11556, 11558],\n",
       "  'an_eas': [8638, 8640],\n",
       "  'an_fin': [6612, 6612],\n",
       "  'an_nfe': [66716, 66714],\n",
       "  'an_oth': [908, 908],\n",
       "  'an_sas': [16506, 16506],\n",
       "  'bp_af_adj_dict': {'CCC': 8.241e-06, 'TCT': 8.241e-06},\n",
       "  'bp_af_dict': {'CCC': 8.236e-06, 'TCT': 8.236e-06},\n",
       "  'bp_list': ['TCT', 'CCC'],\n",
       "  'bp_ref': 'CCT',\n",
       "  'chrom': '1',\n",
       "  'chrom_pos': (51910617, 51910616, 51910615),\n",
       "  'clin_sig': ['', ''],\n",
       "  'ens_gene': 'ENSG00000085832.12',\n",
       "  'ens_prot': ['ENSP00000360798', 'ENSP00000360798'],\n",
       "  'prot_pos': 300},\n",
       " {'PolyPhen': [],\n",
       "  'SIFT': [],\n",
       "  'SwissProt': [],\n",
       "  'aa_ref': 'P',\n",
       "  'aa_ref_orig': 'P',\n",
       "  'ac_adj': [],\n",
       "  'ac_afr': [],\n",
       "  'ac_amr': [],\n",
       "  'ac_eas': [],\n",
       "  'ac_fin': [],\n",
       "  'ac_het': [],\n",
       "  'ac_hom': [],\n",
       "  'ac_nfe': [],\n",
       "  'ac_oth': [],\n",
       "  'ac_sas': [],\n",
       "  'af': 0,\n",
       "  'af_adj': 0,\n",
       "  'alterations_af_adj_dict': defaultdict(list, {}),\n",
       "  'an': [],\n",
       "  'an_adj': [],\n",
       "  'an_afr': [],\n",
       "  'an_amr': [],\n",
       "  'an_eas': [],\n",
       "  'an_fin': [],\n",
       "  'an_nfe': [],\n",
       "  'an_oth': [],\n",
       "  'an_sas': [],\n",
       "  'bp_af_adj_dict': {},\n",
       "  'bp_af_dict': {},\n",
       "  'bp_list': [],\n",
       "  'bp_ref': 'CCC',\n",
       "  'chrom': '2',\n",
       "  'chrom_pos': (31489514, 31489515, 31489516),\n",
       "  'clin_sig': [],\n",
       "  'ens_gene': 'ENSG00000013016.10',\n",
       "  'ens_prot': [],\n",
       "  'prot_pos': 518},\n",
       " {'PolyPhen': [],\n",
       "  'SIFT': [],\n",
       "  'SwissProt': [],\n",
       "  'aa_ref': 'P',\n",
       "  'aa_ref_orig': 'P',\n",
       "  'ac_adj': [],\n",
       "  'ac_afr': [],\n",
       "  'ac_amr': [],\n",
       "  'ac_eas': [],\n",
       "  'ac_fin': [],\n",
       "  'ac_het': [],\n",
       "  'ac_hom': [],\n",
       "  'ac_nfe': [],\n",
       "  'ac_oth': [],\n",
       "  'ac_sas': [],\n",
       "  'af': 0,\n",
       "  'af_adj': 0,\n",
       "  'alterations_af_adj_dict': defaultdict(list, {}),\n",
       "  'an': [],\n",
       "  'an_adj': [],\n",
       "  'an_afr': [],\n",
       "  'an_amr': [],\n",
       "  'an_eas': [],\n",
       "  'an_fin': [],\n",
       "  'an_nfe': [],\n",
       "  'an_oth': [],\n",
       "  'an_sas': [],\n",
       "  'bp_af_adj_dict': {},\n",
       "  'bp_af_dict': {},\n",
       "  'bp_list': [],\n",
       "  'bp_ref': 'CCT',\n",
       "  'chrom': '2',\n",
       "  'chrom_pos': (24535147, 24535146, 24535145),\n",
       "  'clin_sig': [],\n",
       "  'ens_gene': 'ENSG00000198399.10',\n",
       "  'ens_prot': [],\n",
       "  'prot_pos': 96},\n",
       " {'PolyPhen': [''],\n",
       "  'SIFT': [''],\n",
       "  'SwissProt': ['ITSN2_HUMAN'],\n",
       "  'aa_ref': 'P',\n",
       "  'aa_ref_orig': 'P',\n",
       "  'ac_adj': [2],\n",
       "  'ac_afr': [0],\n",
       "  'ac_amr': [1],\n",
       "  'ac_eas': [0],\n",
       "  'ac_fin': [0],\n",
       "  'ac_het': [2],\n",
       "  'ac_hom': [0],\n",
       "  'ac_nfe': [1],\n",
       "  'ac_oth': [0],\n",
       "  'ac_sas': [0],\n",
       "  'af': 0,\n",
       "  'af_adj': 0,\n",
       "  'alterations_af_adj_dict': defaultdict(list, {}),\n",
       "  'an': [121412],\n",
       "  'an_adj': [118198],\n",
       "  'an_afr': [10172],\n",
       "  'an_amr': [11316],\n",
       "  'an_eas': [8494],\n",
       "  'an_fin': [6554],\n",
       "  'an_nfe': [64904],\n",
       "  'an_oth': [888],\n",
       "  'an_sas': [15870],\n",
       "  'bp_af_adj_dict': {'CCG': 1.692e-05},\n",
       "  'bp_af_dict': {'CCG': 1.647e-05},\n",
       "  'bp_list': ['CCG'],\n",
       "  'bp_ref': 'CCA',\n",
       "  'chrom': '2',\n",
       "  'chrom_pos': (24524874, 24524873, 24524872),\n",
       "  'clin_sig': [''],\n",
       "  'ens_gene': 'ENSG00000198399.10',\n",
       "  'ens_prot': ['ENSP00000347244'],\n",
       "  'prot_pos': 319},\n",
       " {'PolyPhen': [],\n",
       "  'SIFT': [],\n",
       "  'SwissProt': [],\n",
       "  'aa_ref': 'P',\n",
       "  'aa_ref_orig': 'P',\n",
       "  'ac_adj': [],\n",
       "  'ac_afr': [],\n",
       "  'ac_amr': [],\n",
       "  'ac_eas': [],\n",
       "  'ac_fin': [],\n",
       "  'ac_het': [],\n",
       "  'ac_hom': [],\n",
       "  'ac_nfe': [],\n",
       "  'ac_oth': [],\n",
       "  'ac_sas': [],\n",
       "  'af': 0,\n",
       "  'af_adj': 0,\n",
       "  'alterations_af_adj_dict': defaultdict(list, {}),\n",
       "  'an': [],\n",
       "  'an_adj': [],\n",
       "  'an_afr': [],\n",
       "  'an_amr': [],\n",
       "  'an_eas': [],\n",
       "  'an_fin': [],\n",
       "  'an_nfe': [],\n",
       "  'an_oth': [],\n",
       "  'an_sas': [],\n",
       "  'bp_af_adj_dict': {},\n",
       "  'bp_af_dict': {},\n",
       "  'bp_list': [],\n",
       "  'bp_ref': 'CCA',\n",
       "  'chrom': '6',\n",
       "  'chrom_pos': (139262529, 139262528, 139262527),\n",
       "  'clin_sig': [],\n",
       "  'ens_gene': 'ENSG00000135597.14',\n",
       "  'ens_prot': [],\n",
       "  'prot_pos': 360},\n",
       " {'PolyPhen': [''],\n",
       "  'SIFT': [''],\n",
       "  'SwissProt': ['EHD1_HUMAN'],\n",
       "  'aa_ref': 'P',\n",
       "  'aa_ref_orig': 'P',\n",
       "  'ac_adj': [120170],\n",
       "  'ac_afr': [10252],\n",
       "  'ac_amr': [11496],\n",
       "  'ac_eas': [8586],\n",
       "  'ac_fin': [6550],\n",
       "  'ac_het': [2],\n",
       "  'ac_hom': [60084],\n",
       "  'ac_nfe': [65909],\n",
       "  'ac_oth': [896],\n",
       "  'ac_sas': [16481],\n",
       "  'af': 0,\n",
       "  'af_adj': 0,\n",
       "  'alterations_af_adj_dict': defaultdict(list, {}),\n",
       "  'an': [121286],\n",
       "  'an_adj': [120172],\n",
       "  'an_afr': [10252],\n",
       "  'an_amr': [11496],\n",
       "  'an_eas': [8586],\n",
       "  'an_fin': [6550],\n",
       "  'an_nfe': [65910],\n",
       "  'an_oth': [896],\n",
       "  'an_sas': [16482],\n",
       "  'bp_af_adj_dict': {'CCT': 0.0},\n",
       "  'bp_af_dict': {'CCT': 0.0},\n",
       "  'bp_list': ['CCC'],\n",
       "  'bp_ref': 'CCC',\n",
       "  'chrom': '11',\n",
       "  'chrom_pos': (64621858, 64621857, 64621856),\n",
       "  'clin_sig': [''],\n",
       "  'ens_gene': 'ENSG00000110047.13',\n",
       "  'ens_prot': ['ENSP00000320516'],\n",
       "  'prot_pos': 518},\n",
       " {'PolyPhen': [],\n",
       "  'SIFT': [],\n",
       "  'SwissProt': [],\n",
       "  'aa_ref': 'P',\n",
       "  'aa_ref_orig': 'P',\n",
       "  'ac_adj': [],\n",
       "  'ac_afr': [],\n",
       "  'ac_amr': [],\n",
       "  'ac_eas': [],\n",
       "  'ac_fin': [],\n",
       "  'ac_het': [],\n",
       "  'ac_hom': [],\n",
       "  'ac_nfe': [],\n",
       "  'ac_oth': [],\n",
       "  'ac_sas': [],\n",
       "  'af': 0,\n",
       "  'af_adj': 0,\n",
       "  'alterations_af_adj_dict': defaultdict(list, {}),\n",
       "  'an': [],\n",
       "  'an_adj': [],\n",
       "  'an_afr': [],\n",
       "  'an_amr': [],\n",
       "  'an_eas': [],\n",
       "  'an_fin': [],\n",
       "  'an_nfe': [],\n",
       "  'an_oth': [],\n",
       "  'an_sas': [],\n",
       "  'bp_af_adj_dict': {},\n",
       "  'bp_af_dict': {},\n",
       "  'bp_list': [],\n",
       "  'bp_ref': 'CCC',\n",
       "  'chrom': '15',\n",
       "  'chrom_pos': (42192908, 42192907, 42192906),\n",
       "  'clin_sig': [],\n",
       "  'ens_gene': 'ENSG00000103966.5',\n",
       "  'ens_prot': [],\n",
       "  'prot_pos': 521},\n",
       " {'PolyPhen': [''],\n",
       "  'SIFT': [''],\n",
       "  'SwissProt': ['EHD2_HUMAN'],\n",
       "  'aa_ref': 'P',\n",
       "  'aa_ref_orig': 'P',\n",
       "  'ac_adj': [1],\n",
       "  'ac_afr': [0],\n",
       "  'ac_amr': [0],\n",
       "  'ac_eas': [1],\n",
       "  'ac_fin': [0],\n",
       "  'ac_het': [1],\n",
       "  'ac_hom': [0],\n",
       "  'ac_nfe': [0],\n",
       "  'ac_oth': [0],\n",
       "  'ac_sas': [0],\n",
       "  'af': 0,\n",
       "  'af_adj': 0,\n",
       "  'alterations_af_adj_dict': defaultdict(list, {}),\n",
       "  'an': [120840],\n",
       "  'an_adj': [100688],\n",
       "  'an_afr': [8140],\n",
       "  'an_amr': [9730],\n",
       "  'an_eas': [7196],\n",
       "  'an_fin': [4628],\n",
       "  'an_nfe': [55136],\n",
       "  'an_oth': [714],\n",
       "  'an_sas': [15144],\n",
       "  'bp_af_adj_dict': {'CCT': 9.932e-06},\n",
       "  'bp_af_dict': {'CCT': 8.275e-06},\n",
       "  'bp_list': ['CCT'],\n",
       "  'bp_ref': 'CCC',\n",
       "  'chrom': '19',\n",
       "  'chrom_pos': (48244624, 48244625, 48244626),\n",
       "  'clin_sig': [''],\n",
       "  'ens_gene': 'ENSG00000024422.7',\n",
       "  'ens_prot': ['ENSP00000263277'],\n",
       "  'prot_pos': 523},\n",
       " {'PolyPhen': [],\n",
       "  'SIFT': [],\n",
       "  'SwissProt': [],\n",
       "  'aa_ref': 'P',\n",
       "  'aa_ref_orig': 'P',\n",
       "  'ac_adj': [],\n",
       "  'ac_afr': [],\n",
       "  'ac_amr': [],\n",
       "  'ac_eas': [],\n",
       "  'ac_fin': [],\n",
       "  'ac_het': [],\n",
       "  'ac_hom': [],\n",
       "  'ac_nfe': [],\n",
       "  'ac_oth': [],\n",
       "  'ac_sas': [],\n",
       "  'af': 0,\n",
       "  'af_adj': 0,\n",
       "  'alterations_af_adj_dict': defaultdict(list, {}),\n",
       "  'an': [],\n",
       "  'an_adj': [],\n",
       "  'an_afr': [],\n",
       "  'an_amr': [],\n",
       "  'an_eas': [],\n",
       "  'an_fin': [],\n",
       "  'an_nfe': [],\n",
       "  'an_oth': [],\n",
       "  'an_sas': [],\n",
       "  'bp_af_adj_dict': {},\n",
       "  'bp_af_dict': {},\n",
       "  'bp_list': [],\n",
       "  'bp_ref': 'CCC',\n",
       "  'chrom': '19',\n",
       "  'chrom_pos': (16536085, 16536084, 16536083),\n",
       "  'clin_sig': [],\n",
       "  'ens_gene': 'ENSG00000127527.9',\n",
       "  'ens_prot': [],\n",
       "  'prot_pos': 201},\n",
       " {'PolyPhen': [],\n",
       "  'SIFT': [],\n",
       "  'SwissProt': [],\n",
       "  'aa_ref': 'G',\n",
       "  'aa_ref_orig': 'G',\n",
       "  'ac_adj': [],\n",
       "  'ac_afr': [],\n",
       "  'ac_amr': [],\n",
       "  'ac_eas': [],\n",
       "  'ac_fin': [],\n",
       "  'ac_het': [],\n",
       "  'ac_hom': [],\n",
       "  'ac_nfe': [],\n",
       "  'ac_oth': [],\n",
       "  'ac_sas': [],\n",
       "  'af': 0,\n",
       "  'af_adj': 0,\n",
       "  'alterations_af_adj_dict': defaultdict(list, {}),\n",
       "  'an': [],\n",
       "  'an_adj': [],\n",
       "  'an_afr': [],\n",
       "  'an_amr': [],\n",
       "  'an_eas': [],\n",
       "  'an_fin': [],\n",
       "  'an_nfe': [],\n",
       "  'an_oth': [],\n",
       "  'an_sas': [],\n",
       "  'bp_af_adj_dict': {},\n",
       "  'bp_af_dict': {},\n",
       "  'bp_list': [],\n",
       "  'bp_ref': 'GGA',\n",
       "  'chrom': '21',\n",
       "  'chrom_pos': (35107434, 35107435, 35107436),\n",
       "  'clin_sig': [],\n",
       "  'ens_gene': 'ENSG00000205726.9',\n",
       "  'ens_prot': [],\n",
       "  'prot_pos': 91},\n",
       " {'PolyPhen': [],\n",
       "  'SIFT': [],\n",
       "  'SwissProt': [],\n",
       "  'aa_ref': 'P',\n",
       "  'aa_ref_orig': 'P',\n",
       "  'ac_adj': [],\n",
       "  'ac_afr': [],\n",
       "  'ac_amr': [],\n",
       "  'ac_eas': [],\n",
       "  'ac_fin': [],\n",
       "  'ac_het': [],\n",
       "  'ac_hom': [],\n",
       "  'ac_nfe': [],\n",
       "  'ac_oth': [],\n",
       "  'ac_sas': [],\n",
       "  'af': 0,\n",
       "  'af_adj': 0,\n",
       "  'alterations_af_adj_dict': defaultdict(list, {}),\n",
       "  'an': [],\n",
       "  'an_adj': [],\n",
       "  'an_afr': [],\n",
       "  'an_amr': [],\n",
       "  'an_eas': [],\n",
       "  'an_fin': [],\n",
       "  'an_nfe': [],\n",
       "  'an_oth': [],\n",
       "  'an_sas': [],\n",
       "  'bp_af_adj_dict': {},\n",
       "  'bp_af_dict': {},\n",
       "  'bp_list': [],\n",
       "  'bp_ref': 'CCA',\n",
       "  'chrom': '21',\n",
       "  'chrom_pos': (35138276, 35138277, 35138278),\n",
       "  'clin_sig': [],\n",
       "  'ens_gene': 'ENSG00000205726.9',\n",
       "  'ens_prot': [],\n",
       "  'prot_pos': 296},\n",
       " {'PolyPhen': [''],\n",
       "  'SIFT': [''],\n",
       "  'SwissProt': ['REPS2_HUMAN'],\n",
       "  'aa_ref': 'P',\n",
       "  'aa_ref_orig': 'P',\n",
       "  'ac_adj': [1],\n",
       "  'ac_afr': [1],\n",
       "  'ac_amr': [0],\n",
       "  'ac_eas': [0],\n",
       "  'ac_fin': [0],\n",
       "  'ac_het': [0],\n",
       "  'ac_hom': [0],\n",
       "  'ac_nfe': [0],\n",
       "  'ac_oth': [0],\n",
       "  'ac_sas': [0],\n",
       "  'af': 0,\n",
       "  'af_adj': 0,\n",
       "  'alterations_af_adj_dict': defaultdict(list, {}),\n",
       "  'an': [121412],\n",
       "  'an_adj': [86105],\n",
       "  'an_afr': [8393],\n",
       "  'an_amr': [9194],\n",
       "  'an_eas': [6561],\n",
       "  'an_fin': [4424],\n",
       "  'an_nfe': [47141],\n",
       "  'an_oth': [622],\n",
       "  'an_sas': [9770],\n",
       "  'bp_af_adj_dict': {'CCC': 1.161e-05},\n",
       "  'bp_af_dict': {'CCC': 1.647e-05},\n",
       "  'bp_list': ['CCC'],\n",
       "  'bp_ref': 'CCT',\n",
       "  'chrom': 'X',\n",
       "  'chrom_pos': (17073028, 17073029, 17073030),\n",
       "  'clin_sig': [''],\n",
       "  'ens_gene': 'ENSG00000169891.13',\n",
       "  'ens_prot': ['ENSP00000349824'],\n",
       "  'prot_pos': 357}]"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "states_dict[84]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sum(domain_ens_genes_errors)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sum(domain_ens_genes_indels)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "9"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sum(domain_ens_genes_filter)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "121412"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "max(chrom_gene_table[\"AN\"])"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
